<!DOCTYPE html><html lang="en-us" >

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  
  
  
    <meta name="generator" content="Wowchemy 5.0.0-beta.3 for Hugo" />
  

  

  
  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Jelmer Poelstra" />

  
  
  
    
  
  <meta name="description" content="Now that we&#39;ve gotten a feel for creating plots, we&#39;ll look at how they can be arranged to include multiple plots in a single figure." />

  
  <link rel="alternate" hreflang="en-us" href="https://biodash.github.io/codeclub/s02e08_multiple_plots/" />

  







  




  
  
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
  

  
  
  
    <meta name="theme-color" content="#3f51b5" />
  

  
  
    
    <script src="/js/mathjax-config.js"></script>
  

  
  
  
  
    
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha256-FMvZuGapsJLjouA6k7Eo2lusoAX9i0ShlWFG6qt7SLc=" crossorigin="anonymous">

    
    

    
    
    
      
    
    
      
      
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-light" media="print" onload="this.media='all'">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-dark" media="print" onload="this.media='all'" disabled>
      
    

    
    
    

    

    
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      
      

      
      

      
    
      
      

      
      

      
    
  

  
  
  
    
      
      <link rel="preload" as="style" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap">
      <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media="print" onload="this.media='all'">
    
  

  
  
  
  
  
  <link rel="stylesheet" href="/css/wowchemy.80f9a276c6100914c616a06e45ba8d09.css" />

  




<script async src="https://www.googletagmanager.com/gtag/js?id=G-9F9CQ7P0DX"></script>
<script>
  window.dataLayer = window.dataLayer || [];

  function gtag() {
      dataLayer.push(arguments);
  }

  function trackOutboundLink(url, target) {
    gtag('event', 'click', {
         'event_category': 'outbound',
         'event_label': url,
         'transport_type': 'beacon',
         'event_callback': function () {
           if (target !== '_blank') {
             document.location = url;
           }
         }
    });
    console.debug("Outbound link clicked: " + url);
  }

  function onClickCallback(event) {
    if ((event.target.tagName !== 'A') || (event.target.host === window.location.host)) {
      return;
    }
    trackOutboundLink(event.target, event.target.getAttribute('target'));  
  }

  gtag('js', new Date());
  gtag('config', 'G-9F9CQ7P0DX', {});
  gtag('set', {'cookie_flags': 'SameSite=None;Secure'});

  
  document.addEventListener('click', onClickCallback, false);
</script>


  

  

  




  
  
  

  

  
    <link rel="manifest" href="/index.webmanifest" />
  

  <link rel="icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_32x32_fill_lanczos_center_2.png" />
  <link rel="apple-touch-icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_180x180_fill_lanczos_center_2.png" />

  <link rel="canonical" href="https://biodash.github.io/codeclub/s02e08_multiple_plots/" />

  
  
  
  
  
  
  
  
    
    
  
  
  <meta property="twitter:card" content="summary" />
  
  <meta property="og:site_name" content="BioDASH" />
  <meta property="og:url" content="https://biodash.github.io/codeclub/s02e08_multiple_plots/" />
  <meta property="og:title" content="Code Club S02E08: Combining Plots | BioDASH" />
  <meta property="og:description" content="Now that we&#39;ve gotten a feel for creating plots, we&#39;ll look at how they can be arranged to include multiple plots in a single figure." /><meta property="og:image" content="https://biodash.github.io/media/logo_hu6407506e646a281542ecb299640680ce_37988_300x300_fit_lanczos_2.png" />
    <meta property="twitter:image" content="https://biodash.github.io/media/logo_hu6407506e646a281542ecb299640680ce_37988_300x300_fit_lanczos_2.png" /><meta property="og:locale" content="en-us" />
  
    
      <meta
        property="article:published_time"
        content="2021-10-19T00:00:00&#43;00:00"
      />
    
    <meta property="article:modified_time" content="2021-10-26T14:42:15-04:00">
  

  



  

  

  <link rel='stylesheet' href='/css/highlight-light.css' title='hl-light'>
<link rel='stylesheet' href='/css/highlight-dark.css' title='hl-dark' disabled>



  <title>Code Club S02E08: Combining Plots | BioDASH</title>
</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper   " data-wc-page-id="803100179d0db591e042f5c7dc41eb7e" >

  
  
  
  
  
  
  
  
  <script src="/js/wowchemy-init.min.7a224883d792570fd34d30c7134eace3.js"></script>

  


<aside class="search-modal" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#" aria-label="Close"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control"
        aria-label="Search...">
        
      </div>

      
      

      

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>



  <div class="page-header">
    











  


<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container-xl">

    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Code Club</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/codeclub-schedule/"><span>- <b>Schedule</b></span></a>
            
              <a class="dropdown-item" href="/codeclub-schedule/#previous-semesters"><span>- Past Sessions</span></a>
            
              <a class="dropdown-item" href="/codeclub-setup/"><span>- Computer Setup</span></a>
            
              <a class="dropdown-item" href="/"><span>------------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-suggest/"><span>- Suggest a Topic</span></a>
            
              <a class="dropdown-item" href="/"><span>-----------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-about/"><span>- <b>About Code Club & Join Us!</b></span></a>
            
          </div>
        </li>

        
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Tutorials & Resources</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/events/2023-02_aws/"><span>AWS Workshop (Feb '23)</span></a>
            
              <a class="dropdown-item" href="/"><span>---------------------</span></a>
            
              <a class="dropdown-item" href="/tutorials/r-resources-tips/"><span>- R Resources and Tips</span></a>
            
              <a class="dropdown-item" href="/"><span>--------------------</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/2020-12-microbiomics-workshop/"><span>- Metabarcoding workshop (Fall '20)</span></a>
            
              <a class="dropdown-item" href="/tutorials/2021-01_rnaseq/"><span>- RNA-seq intro meetings (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/pracs-sp21/"><span>- Computing Course (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/cl-workshop-22/"><span>- Command-line, shell scripting and OSC Workshop (Summer '22)</span></a>
            
              <a class="dropdown-item" href="/tutorials/2023-04_aggenomics/presentation.html"><span>- Ag Genomics Lecture: Command-line computing (Apr '23)</span></a>
            
          </div>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/support/"><span>Bioinformatics Support</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/about/"><span>About</span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="https://github.com/biodash/biodash.github.io" target="_blank" rel="noopener"><i class='fab fa-github'></i><span></span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/index.xml" target="_blank" rel="noopener"><i class='fas fa-rss'></i><span></span></a>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">

      
      
        
      
        
      
        
      
        
      

      
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      
      <li class="nav-item dropdown theme-dropdown">
        <a href="#" class="nav-link" data-toggle="dropdown" aria-haspopup="true" aria-label="Display preferences">
          <i class="fas fa-moon" aria-hidden="true"></i>
        </a>
        <div class="dropdown-menu">
          <a href="#" class="dropdown-item js-set-theme-light">
            <span>Light</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-dark">
            <span>Dark</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-auto">
            <span>Automatic</span>
          </a>
        </div>
      </li>
      

      
      

    </ul>

  </div>
</nav>


  </div>

  <div class="page-body">
    
<div class="container-fluid docs">
    <div class="row flex-xl-nowrap">
        <div class="d-none d-xl-block col-xl-2 docs-toc">
            <ul class="nav toc-top">
                <li><a href="#" id="back_to_top" class="docs-toc-title">Contents</a></li>
            </ul>
            <nav id="TableOfContents">
  <ul>
    <li><a href="#learning-objectives">Learning objectives</a></li>
    <li><a href="#1----intro">1 &ndash; Intro</a></li>
    <li><a href="#2----faceting">2 &ndash; Faceting</a>
      <ul>
        <li><a href="#breakout-rooms-faceting">Breakout Rooms: Faceting</a></li>
      </ul>
    </li>
  </ul>
</nav>
        </div>
        <main class="col-12 col-md-0 col-xl-10 py-md-3 pl-md-5 docs-content" role="main">
            

            <article class="article">

                












  

  
  
  
<div class="article-container pt-3">
  <h1>Code Club S02E08: Combining Plots</h1>

  

  
    



<div class="article-metadata">

  
  
  
  
  <div>
    

  <span >
      <a href="/authors/mike-sovic/">Mike Sovic</a></span>
  </div>
  
  

  
  <span class="article-date">
    
    
    Oct 26, 2021
  </span>
  

  

  

  
  
  
  
  

  
  

</div>

    





  
</div>



                <div class="article-container">
                    <div class="article-style">
                        <br>
<h2 id="learning-objectives">Learning objectives</h2>
<blockquote>
<ul>
<li>Continue to practice creating plots with ggplot</li>
<li>Use faceting to divide a plot into multiple panels according to some variable.</li>
<li>Arrange multiple plots of different types on a single figure.</li>
</ul>
</blockquote>
<br>
<hr>
<h2 id="1----intro">1 &ndash; Intro</h2>
<p>We&rsquo;ll continue with our theme on plotting by exploring some options for arranging multiple plots on a single figure. A couple scenarios where you might want to do this&hellip;</p>
<p>1.) You create a plot that needs to be subdivided according to some variable, possibly because accounting for that variable is important for the interpretation, or maybe there&rsquo;s just too much on one plot and it helps to split the data up according to some factor.</p>
<p>2.) You have a series of different plots that all address some related question, maybe each in a slightly different way, and you want to present them all in one figure.</p>
<p>We&rsquo;ll take a couple approaches during this and next week&rsquo;s sessions to deal with these two scenarios. Today we&rsquo;ll look at some <em>ggplot</em> functions like <code>facet_wrap()</code> and <code>facet_grid()</code> that allow us to easily deal with scenario 1. Then in the next session we&rsquo;ll try a separate package, <em>patchwork</em>, that offers one good option for scenario 2.</p>
<p>Like in previous sessions, we&rsquo;ll use some packages from the <em>tidyverse</em> and also the <em>palmerpenguins</em> dataset. If you haven&rsquo;t installed either of those yet, you can do so with the following commands. If you installed them previously, you can just run the latter of the commands (<a href="https://rdrr.io/r/base/library.html" target="_blank" rel="noopener"><code>library()</code></a>) to load them for the current session.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"tidyverse"</span><span class='o'>)</span>
<span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"palmerpenguins"</span><span class='o'>)</span>
</code></pre>
</div>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://allisonhorst.github.io/palmerpenguins/'>palmerpenguins</a></span><span class='o'>)</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='o'>)</span>
</code></pre>
</div>
<p>And now let&rsquo;s preview/explore the penguins dataset just to remind ourselves of what&rsquo;s in there&hellip;</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='o'>(</span><span class='nv'>penguins</span><span class='o'>)</span>

<span class='c'>#&gt; <span style='color: #555555;'># A tibble: 6 x 8</span></span>
<span class='c'>#&gt;   species island bill_length_mm bill_depth_mm flipper_length_… body_mass_g sex  </span>
<span class='c'>#&gt;   <span style='color: #555555;font-style: italic;'>&lt;fct&gt;</span><span>   </span><span style='color: #555555;font-style: italic;'>&lt;fct&gt;</span><span>           </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>         </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>            </span><span style='color: #555555;font-style: italic;'>&lt;int&gt;</span><span>       </span><span style='color: #555555;font-style: italic;'>&lt;int&gt;</span><span> </span><span style='color: #555555;font-style: italic;'>&lt;fct&gt;</span></span>
<span class='c'>#&gt; <span style='color: #555555;'>1</span><span> Adelie  Torge…           39.1          18.7              181        </span><span style='text-decoration: underline;'>3</span><span>750 male </span></span>
<span class='c'>#&gt; <span style='color: #555555;'>2</span><span> Adelie  Torge…           39.5          17.4              186        </span><span style='text-decoration: underline;'>3</span><span>800 fema…</span></span>
<span class='c'>#&gt; <span style='color: #555555;'>3</span><span> Adelie  Torge…           40.3          18                195        </span><span style='text-decoration: underline;'>3</span><span>250 fema…</span></span>
<span class='c'>#&gt; <span style='color: #555555;'>4</span><span> Adelie  Torge…           </span><span style='color: #BB0000;'>NA</span><span>            </span><span style='color: #BB0000;'>NA</span><span>                 </span><span style='color: #BB0000;'>NA</span><span>          </span><span style='color: #BB0000;'>NA</span><span> </span><span style='color: #BB0000;'>NA</span><span>   </span></span>
<span class='c'>#&gt; <span style='color: #555555;'>5</span><span> Adelie  Torge…           36.7          19.3              193        </span><span style='text-decoration: underline;'>3</span><span>450 fema…</span></span>
<span class='c'>#&gt; <span style='color: #555555;'>6</span><span> Adelie  Torge…           39.3          20.6              190        </span><span style='text-decoration: underline;'>3</span><span>650 male </span></span>
<span class='c'>#&gt; <span style='color: #555555;'># … with 1 more variable: year &lt;int&gt;</span></span>

<span class='nf'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='o'>(</span><span class='nv'>penguins</span><span class='o'>)</span>

<span class='c'>#&gt;       species          island    bill_length_mm  bill_depth_mm  </span>
<span class='c'>#&gt;  Adelie   :152   Biscoe   :168   Min.   :32.10   Min.   :13.10  </span>
<span class='c'>#&gt;  Chinstrap: 68   Dream    :124   1st Qu.:39.23   1st Qu.:15.60  </span>
<span class='c'>#&gt;  Gentoo   :124   Torgersen: 52   Median :44.45   Median :17.30  </span>
<span class='c'>#&gt;                                  Mean   :43.92   Mean   :17.15  </span>
<span class='c'>#&gt;                                  3rd Qu.:48.50   3rd Qu.:18.70  </span>
<span class='c'>#&gt;                                  Max.   :59.60   Max.   :21.50  </span>
<span class='c'>#&gt;                                  NA's   :2       NA's   :2      </span>
<span class='c'>#&gt;  flipper_length_mm  body_mass_g       sex           year     </span>
<span class='c'>#&gt;  Min.   :172.0     Min.   :2700   female:165   Min.   :2007  </span>
<span class='c'>#&gt;  1st Qu.:190.0     1st Qu.:3550   male  :168   1st Qu.:2007  </span>
<span class='c'>#&gt;  Median :197.0     Median :4050   NA's  : 11   Median :2008  </span>
<span class='c'>#&gt;  Mean   :200.9     Mean   :4202                Mean   :2008  </span>
<span class='c'>#&gt;  3rd Qu.:213.0     3rd Qu.:4750                3rd Qu.:2009  </span>
<span class='c'>#&gt;  Max.   :231.0     Max.   :6300                Max.   :2009  </span>
<span class='c'>#&gt;  NA's   :2         NA's   :2</span>
</code></pre>
</div>
<h2 id="2----faceting">2 &ndash; Faceting</h2>
<p>Let&rsquo;s start by revisiting some plots Michael Broe created in his intro to ggplot a couple sessions ago. He was using the plots to investigate whether a relationship exists between the variables <em>bill length</em> and <em>bill depth</em> in these penguins. A scatterplot with a line of best fit from <em>ggplot</em>&hellip;</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span>method <span class='o'>=</span> <span class='s'>"lm"</span><span class='o'>)</span>

<span class='c'>#&gt; `geom_smooth()` using formula 'y ~ x'</span>

<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_smooth).</span>

<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>

</code></pre>
<p><img src="figs/unnamed-chunk-4-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>As Michael pointed out previously, mapping an additional aesthetic (color) to the variable <em>species</em> helps us see a relationship a little more clearly&hellip;</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span>, color <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span>method <span class='o'>=</span> <span class='s'>"lm"</span><span class='o'>)</span>

<span class='c'>#&gt; `geom_smooth()` using formula 'y ~ x'</span>

<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_smooth).</span>

<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>

</code></pre>
<p><img src="figs/unnamed-chunk-5-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>The color aesthetic partitions the data according to some variable (in this case, species), and here helps add important information to the visualization. An alternative might be to plot the data in separate panels, with each corresponding to a different species. We can do that with either of two functions from ggplot, <code>facet_wrap()</code> or <code>facet_grid()</code>. Let&rsquo;s start with <code>facet_wrap()</code>. This is added as an additional layer to the plot, and indicates one or more variables that will be used to split the data into separate panels. I&rsquo;ll facet here by species.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span>method <span class='o'>=</span> <span class='s'>"lm"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>facet_wrap</span><span class='o'>(</span><span class='s'>"species"</span><span class='o'>)</span>

<span class='c'>#&gt; `geom_smooth()` using formula 'y ~ x'</span>

<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_smooth).</span>

<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>

</code></pre>
<p><img src="figs/unnamed-chunk-6-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>The effect here is similar to what we did with adding a color aesthetic to the <em>species</em> variable earlier - it allows us to evaluate the relationship between bill length and bill depth for each species separately.</p>
<hr>
<h3 id="breakout-rooms-faceting">Breakout Rooms: Faceting</h3>
<h4 id="exercise-1-analyze-adelie-penguins-by-island">Exercise 1: Analyze Adelie Penguins By Island</h4>
<div class="puzzle">
<div>
<p>Try analyzing the relationship between bill length and bill depth for just the Adelie penguins (the only species with observations from each of the three islands). For this species, try faceting by island. Does the relationship seem to be consistent across all islands?</p>
<details>
<summary>
<b>Hint</b> (click here)
</summary>
<p><br> Use <a href="https://rdrr.io/r/stats/filter.html" target="_blank" rel="noopener"><code>filter()</code></a> to select out Adelie penguins, then create a plot similar to the one in the example, but facet on <em>island</em> instead of <em>species</em> <br></p>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span>
  <span class='nf'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='o'>(</span><span class='nv'>species</span> <span class='o'>==</span> <span class='s'>"Adelie"</span><span class='o'>)</span> <span class='o'>%&gt;%</span>
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span>method <span class='o'>=</span> <span class='s'>"lm"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>facet_wrap</span><span class='o'>(</span><span class='s'>"island"</span><span class='o'>)</span>

<span class='c'>#&gt; `geom_smooth()` using formula 'y ~ x'</span>

<span class='c'>#&gt; Warning: Removed 1 rows containing non-finite values (stat_smooth).</span>

<span class='c'>#&gt; Warning: Removed 1 rows containing missing values (geom_point).</span>

</code></pre>
<p><img src="figs/unnamed-chunk-7-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<h4 id="exercise-2a-multiple-facets">Exercise 2a: Multiple Facets</h4>
<div class="puzzle">
<div>
<p>Now building on the plot you just created for Adelie Penguins, what if you wanted to facet on not just <em>island</em>, but a combination of <em>island</em> and <em>sex</em>? Give it a try.</p>
<details>
<summary>
<b>Hint</b> (click here)
</summary>
<p><br> <code>facet_wrap()</code> accepts a character vector of column names. Use the c() function to provide a vector with two column names. <br></p>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='o'>(</span><span class='nv'>species</span> <span class='o'>==</span> <span class='s'>"Adelie"</span><span class='o'>)</span> <span class='o'>%&gt;%</span>
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span>method <span class='o'>=</span> <span class='s'>"lm"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>facet_wrap</span><span class='o'>(</span><span class='nf'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='o'>(</span><span class='s'>"island"</span>, <span class='s'>"sex"</span><span class='o'>)</span><span class='o'>)</span>

<span class='c'>#&gt; `geom_smooth()` using formula 'y ~ x'</span>

<span class='c'>#&gt; Warning: Removed 1 rows containing non-finite values (stat_smooth).</span>

<span class='c'>#&gt; Warning: Removed 1 rows containing missing values (geom_point).</span>

</code></pre>
<p><img src="figs/unnamed-chunk-8-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<h4 id="exercise-2b-multiple-facets">Exercise 2b: Multiple Facets</h4>
<div class="puzzle">
<div>
<p>There are some facets coming through in that last plot that are based on NA&rsquo;s. Try getting rid of all observations that include missing data before creating the plot.</p>
<details>
<summary>
<b>Hint</b> (click here)
</summary>
<p><br> Use the <code>drop_na()</code> function to remove observations with NA before calling ggplot. <br></p>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>drop_na</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>%&gt;%</span>
  <span class='nf'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='o'>(</span><span class='nv'>species</span> <span class='o'>==</span> <span class='s'>"Adelie"</span><span class='o'>)</span> <span class='o'>%&gt;%</span>
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span>method <span class='o'>=</span> <span class='s'>"lm"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>facet_wrap</span><span class='o'>(</span><span class='nf'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='o'>(</span><span class='s'>"island"</span>, <span class='s'>"sex"</span><span class='o'>)</span><span class='o'>)</span>

<span class='c'>#&gt; `geom_smooth()` using formula 'y ~ x'</span>

</code></pre>
<p><img src="figs/unnamed-chunk-9-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<h4 id="exercise-3-axis-scales">Exercise 3: Axis Scales</h4>
<div class="puzzle">
<div>
<p>Now let&rsquo;s go back to the full dataset where we faceted by species. The code we used (with the <code>drop_na</code> function added), along with its associated plot, are below&hellip;</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>drop_na</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>%&gt;%</span>
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span>method <span class='o'>=</span> <span class='s'>"lm"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>facet_wrap</span><span class='o'>(</span><span class='s'>"species"</span><span class='o'>)</span>

<span class='c'>#&gt; `geom_smooth()` using formula 'y ~ x'</span>

</code></pre>
<p><img src="figs/unnamed-chunk-10-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Use the help page for <code>facet_wrap</code> to look in to the <em>scales</em> option. Try changing the value of this option to see what effect it has on the plot.</p>
<details>
<summary>
<b>Hint 1</b> (click here)
</summary>
<p><br> Use <code>?facet_wrap</code> to get the help page for the function, and find information about the <em>scales</em> option. <br></p>
</details>
<details>
<summary>
<b>Hint 2</b> (click here)
</summary>
<p><br> Within the <code>facet_wrap()</code> function, set scales = &ldquo;free_y&rdquo;. <br></p>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span>method <span class='o'>=</span> <span class='s'>"lm"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>facet_wrap</span><span class='o'>(</span><span class='s'>"species"</span>, scales <span class='o'>=</span> <span class='s'>"free"</span><span class='o'>)</span>

<span class='c'>#&gt; `geom_smooth()` using formula 'y ~ x'</span>

<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_smooth).</span>

<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>

</code></pre>
<p><img src="figs/unnamed-chunk-11-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<br>
<hr>
<p>In next week&rsquo;s session, we&rsquo;ll use <code>facet_grid()</code>, which has some similarities to <code>facet_wrap()</code>, and then check out the <em>patchwork</em> package, which gives you more control over how multiple plots are combined in a single figure.</p>

                    </div>

                    <br>

                    








<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://biodash.github.io/codeclub/s02e08_multiple_plots/&amp;text=Code%20Club%20S02E08:%20Combining%20Plots" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://biodash.github.io/codeclub/s02e08_multiple_plots/&amp;t=Code%20Club%20S02E08:%20Combining%20Plots" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=Code%20Club%20S02E08:%20Combining%20Plots&amp;body=https://biodash.github.io/codeclub/s02e08_multiple_plots/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://biodash.github.io/codeclub/s02e08_multiple_plots/&amp;title=Code%20Club%20S02E08:%20Combining%20Plots" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="whatsapp://send?text=Code%20Club%20S02E08:%20Combining%20Plots%20https://biodash.github.io/codeclub/s02e08_multiple_plots/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://biodash.github.io/codeclub/s02e08_multiple_plots/&amp;title=Code%20Club%20S02E08:%20Combining%20Plots" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>











  
  
    



  
  
  
  
  
  <div class="media author-card content-widget-hr">
    
      
      <a href="/authors/mike-sovic/"><img class="avatar mr-3 avatar-circle" src="/authors/mike-sovic/avatar_hu1ab2c0735ccc2e414b344a1bc08b8162_24652_270x270_fill_q90_lanczos_center.jpg" alt="Mike Sovic"></a>
    

    <div class="media-body">
      <h5 class="card-title"><a href="/authors/mike-sovic/">Mike Sovic</a></h5>
      <h6 class="card-subtitle">Bioinformatician at CAPS</h6>
      
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
    <li>
      <a href="mailto:sovic.1@osu.edu" >
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/mikesovic" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://www.youtube.com/channel/UC2dB6jDTbqzlTM6edzfBSGQ" target="_blank" rel="noopener">
        <i class="fab fa-youtube"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>


  
















                    <br>

                    <hr class="separator">

                    <div class="clearfix">
                        
                        
                        <span class="previous-link">
                            <a href="https://biodash.github.io/codeclub/s02e07_ggplot2_part2/">← S02E07: Intro to ggplot2 (part 2)</a>
                        </span>
                        
                        
                        
                        
                        <span class="next-link">
                            <a href="https://biodash.github.io/codeclub/s02e09_multiple_plots_part2/">Code Club S02E09: Combining Plots - Part 2 →</a>
                        </span>
                        
                        
                    </div>

                    <hr class="separator">

                </div>

            </article>

            
        </main>
    </div>
</div>

  </div>

  <div class="page-footer">
    
    
    <div class="container">
      <footer class="site-footer">
    
    <p class="powered-by">
      <a href="/license/">LICENSE: CC-BY-SA <br> <i class="fab fa-creative-commons fa-2x"></i><i class="fab fa-creative-commons-by fa-2x"></i><i class="fab fa-creative-commons-sa fa-2x"></i> </a><br>
    </p>
    
</footer>
    </div>
    
  </div>

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/instant.page/5.1.0/instantpage.min.js" integrity="sha512-1+qUtKoh9XZW7j+6LhRMAyOrgSQKenQ4mluTR+cvxXjP1Z54RxZuzstR/H9kgPXQsVB8IW7DMDFUJpzLjvhGSQ==" crossorigin="anonymous"></script>

      
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      

      
      

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/highlight.min.js" integrity="sha512-TDKKr+IvoqZnPzc3l35hdjpHD0m+b2EC2SrLEgKDRWpxf2rFCxemkgvJ5kfU48ip+Y+m2XVKyOCD85ybtlZDmw==" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/languages/r.min.js"></script>
        
      

    

    
    
    
      <script async defer src="https://maps.googleapis.com/maps/api/js?key="></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/gmaps.js/0.4.25/gmaps.min.js" integrity="sha256-7vjlAeb8OaTrCXZkCNun9djzuB2owUsaO72kXaFDBJs=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
      
      <script id="search-hit-fuse-template" type="text/x-template">
        <div class="search-hit" id="summary-{{key}}">
          <div class="search-hit-content">
            <div class="search-hit-name">
              <a href="{{relpermalink}}">{{title}}</a>
              <div class="article-metadata search-hit-type">{{type}}</div>
              <p class="search-hit-description">{{snippet}}</p>
            </div>
          </div>
        </div>
      </script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
    

    
    <script src="/js/bootstrap.bundle.min.6aed84840afc03ab4d5750157f69c120.js"></script>

    
    
      
      
      
      
      
      
      
    

    
    
    
    
    
    
    
    
      
      
    
    
      
    
    <script src="/en/js/wowchemy.min.2f16af175a3ec51b41b36bb4934b1cf7.js"></script>

    






</body>
</html>
