<!DOCTYPE html><html lang="en-us" >

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  
  
  
    <meta name="generator" content="Wowchemy 5.0.0-beta.3 for Hugo" />
  

  

  
  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Jelmer Poelstra" />

  
  
  
    
  
  <meta name="description" content="In the first session of Code Club, we&#39;ll make sure that everyone is properly set up, create an RStudio Project, and start working with some data from the Great Backyard Bird Count." />

  
  <link rel="alternate" hreflang="en-us" href="https://biodash.github.io/codeclub/01_backyard-birds/" />

  







  




  
  
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
  

  
  
  
    <meta name="theme-color" content="#3f51b5" />
  

  
  
    
    <script src="/js/mathjax-config.js"></script>
  

  
  
  
  
    
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha256-FMvZuGapsJLjouA6k7Eo2lusoAX9i0ShlWFG6qt7SLc=" crossorigin="anonymous">

    
    

    
    
    
      
    
    
      
      
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-light" media="print" onload="this.media='all'">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-dark" media="print" onload="this.media='all'" disabled>
      
    

    
    
    

    

    
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      
      

      
      

      
    
      
      

      
      

      
    
  

  
  
  
    
      
      <link rel="preload" as="style" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap">
      <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media="print" onload="this.media='all'">
    
  

  
  
  
  
  
  <link rel="stylesheet" href="/css/wowchemy.80f9a276c6100914c616a06e45ba8d09.css" />

  




<script async src="https://www.googletagmanager.com/gtag/js?id=G-9F9CQ7P0DX"></script>
<script>
  window.dataLayer = window.dataLayer || [];

  function gtag() {
      dataLayer.push(arguments);
  }

  function trackOutboundLink(url, target) {
    gtag('event', 'click', {
         'event_category': 'outbound',
         'event_label': url,
         'transport_type': 'beacon',
         'event_callback': function () {
           if (target !== '_blank') {
             document.location = url;
           }
         }
    });
    console.debug("Outbound link clicked: " + url);
  }

  function onClickCallback(event) {
    if ((event.target.tagName !== 'A') || (event.target.host === window.location.host)) {
      return;
    }
    trackOutboundLink(event.target, event.target.getAttribute('target'));  
  }

  gtag('js', new Date());
  gtag('config', 'G-9F9CQ7P0DX', {});
  gtag('set', {'cookie_flags': 'SameSite=None;Secure'});

  
  document.addEventListener('click', onClickCallback, false);
</script>


  

  

  




  
  
  

  

  
    <link rel="manifest" href="/index.webmanifest" />
  

  <link rel="icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_32x32_fill_lanczos_center_2.png" />
  <link rel="apple-touch-icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_180x180_fill_lanczos_center_2.png" />

  <link rel="canonical" href="https://biodash.github.io/codeclub/01_backyard-birds/" />

  
  
  
  
  
  
  
  
    
  
  
  <meta property="twitter:card" content="summary_large_image" />
  
  <meta property="og:site_name" content="BioDASH" />
  <meta property="og:url" content="https://biodash.github.io/codeclub/01_backyard-birds/" />
  <meta property="og:title" content="Session 1: Backyard Birds | BioDASH" />
  <meta property="og:description" content="In the first session of Code Club, we&#39;ll make sure that everyone is properly set up, create an RStudio Project, and start working with some data from the Great Backyard Bird Count." /><meta property="og:image" content="https://biodash.github.io/codeclub/01_backyard-birds/featured.jpg" />
    <meta property="twitter:image" content="https://biodash.github.io/codeclub/01_backyard-birds/featured.jpg" /><meta property="og:locale" content="en-us" />
  
    
      <meta
        property="article:published_time"
        content="2020-11-04T00:00:00&#43;00:00"
      />
    
    <meta property="article:modified_time" content="2020-11-21T09:45:15-05:00">
  

  



  

  

  <link rel='stylesheet' href='/css/highlight-light.css' title='hl-light'>
<link rel='stylesheet' href='/css/highlight-dark.css' title='hl-dark' disabled>



  <title>Session 1: Backyard Birds | BioDASH</title>
</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper   " data-wc-page-id="5b8cbf273e25c833a3e7f4cd2615654b" >

  
  
  
  
  
  
  
  
  <script src="/js/wowchemy-init.min.7a224883d792570fd34d30c7134eace3.js"></script>

  


<aside class="search-modal" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#" aria-label="Close"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control"
        aria-label="Search...">
        
      </div>

      
      

      

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>



  <div class="page-header">
    











  


<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container-xl">

    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Code Club</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/codeclub-schedule/"><span>- <b>Schedule</b></span></a>
            
              <a class="dropdown-item" href="/codeclub-schedule/#previous-semesters"><span>- Past Sessions</span></a>
            
              <a class="dropdown-item" href="/codeclub-setup/"><span>- Computer Setup</span></a>
            
              <a class="dropdown-item" href="/"><span>------------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-suggest/"><span>- Suggest a Topic</span></a>
            
              <a class="dropdown-item" href="/"><span>-----------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-about/"><span>- <b>About Code Club & Join Us!</b></span></a>
            
          </div>
        </li>

        
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Tutorials & Resources</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/events/2023-02_aws/"><span>AWS Workshop (Feb '23)</span></a>
            
              <a class="dropdown-item" href="/"><span>---------------------</span></a>
            
              <a class="dropdown-item" href="/tutorials/r-resources-tips/"><span>- R Resources and Tips</span></a>
            
              <a class="dropdown-item" href="/"><span>--------------------</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/2020-12-microbiomics-workshop/"><span>- Metabarcoding workshop (Fall '20)</span></a>
            
              <a class="dropdown-item" href="/tutorials/2021-01_rnaseq/"><span>- RNA-seq intro meetings (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/pracs-sp21/"><span>- Computing Course (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/cl-workshop-22/"><span>- Command-line, shell scripting and OSC Workshop (Summer '22)</span></a>
            
              <a class="dropdown-item" href="/tutorials/2023-04_aggenomics/presentation.html"><span>- Ag Genomics Lecture: Command-line computing (Apr '23)</span></a>
            
          </div>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/support/"><span>Bioinformatics Support</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/about/"><span>About</span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="https://github.com/biodash/biodash.github.io" target="_blank" rel="noopener"><i class='fab fa-github'></i><span></span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/index.xml" target="_blank" rel="noopener"><i class='fas fa-rss'></i><span></span></a>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">

      
      
        
      
        
      
        
      
        
      

      
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      
      <li class="nav-item dropdown theme-dropdown">
        <a href="#" class="nav-link" data-toggle="dropdown" aria-haspopup="true" aria-label="Display preferences">
          <i class="fas fa-moon" aria-hidden="true"></i>
        </a>
        <div class="dropdown-menu">
          <a href="#" class="dropdown-item js-set-theme-light">
            <span>Light</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-dark">
            <span>Dark</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-auto">
            <span>Automatic</span>
          </a>
        </div>
      </li>
      

      
      

    </ul>

  </div>
</nav>


  </div>

  <div class="page-body">
    
<div class="container-fluid docs">
    <div class="row flex-xl-nowrap">
        <div class="d-none d-xl-block col-xl-2 docs-toc">
            <ul class="nav toc-top">
                <li><a href="#" id="back_to_top" class="docs-toc-title">Contents</a></li>
            </ul>
            <nav id="TableOfContents">
  <ul>
    <li><a href="#prep-homework">Prep homework</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#slides">Slides</a></li>
    <li><a href="#1---create-an-rstudio-project">1 - Create an RStudio Project</a></li>
    <li><a href="#2---orienting-ourselves">2 - Orienting ourselves</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#3---getting-our-dataset">3 - Getting our dataset</a>
      <ul>
        <li><a href="#the-great-backyard-bird-count">The Great Backyard Bird Count</a></li>
      </ul>
    </li>
    <li><a href="#4---exploring-backyard-birds">4 - Exploring backyard birds</a>
      <ul>
        <li><a href="#exercise-1">Exercise 1</a></li>
      </ul>
    </li>
    <li><a href="#bonus-material">Bonus material</a>
      <ul>
        <li><a href="#readr-options-for-challenging-files"><code>readr</code> options for challenging files</a></li>
        <li><a href="#exercise-2-optional">Exercise 2 (Optional)</a></li>
        <li><a href="#other-options-for-reading-tabular-data">Other options for reading tabular data</a></li>
      </ul>
    </li>
  </ul>
</nav>
        </div>
        <main class="col-12 col-md-0 col-xl-10 py-md-3 pl-md-5 docs-content" role="main">
            

            <article class="article">

                




















  
  
    
  


<div class="article-container pt-3">
  <h1>Session 1: Backyard Birds</h1>

  
  <p class="page-subtitle">Starting with RStudio Projects and reading in our data</p>
  

  



<div class="article-metadata">

  
  
  
  
  <div>
    

  <span >
      <a href="/authors/admin/">Jelmer Poelstra</a></span>
  </div>
  
  

  
  <span class="article-date">
    
    
    Nov 21, 2020
  </span>
  

  

  

  
  
  
  
  

  
  

</div>

  





</div>


<div class="article-header article-container featured-image-wrapper mt-4 mb-4" style="max-width: 720px; max-height: 480px;">
  <div style="position: relative">
    <img src="/codeclub/01_backyard-birds/featured_hu5614a25c5db1a711bf5d1af486bb9cdf_890326_720x0_resize_q90_lanczos.jpg" alt="" class="featured-image">
    <span class="article-header-caption">Red-breasted Nuthatch, <em>Sitta canadensis</em></span>
  </div>
</div>



                <div class="article-container">
                    <div class="article-style">
                        <p><br> <br> <br></p>
<hr>
<h2 id="prep-homework">Prep homework</h2>
<h4 id="basic-computer-setup">Basic computer setup</h4>
<p>If you didn&rsquo;t already do this, please follow the <a href="/codeclub-setup/">Code Club Computer Setup</a> instructions.</p>
<h4 id="test-if-it-works">Test if it works</h4>
<p>Please open RStudio locally or <a href="/codeclub-setup/#osc-run-rstudio">start an OSC RStudio Server session</a>.</p>
<p><strong>Nov 19 addition: If you&rsquo;re working locally, test if you can load the <em>tidyverse</em> package with <a href="http://tidyverse.tidyverse.org" target="_blank" rel="noopener"><code>library(&quot;tidyverse&quot;)</code></a> inside R.</strong> (If you haven&rsquo;t installed the <em>tidyverse</em> yet, please go to the <a href="/codeclub-setup/#install-tidy">Code Club Computer Setup</a> instructions.)</p>
<p>If you have not used RStudio before, take a moment to explore what&rsquo;s in the panels and tabs. (It may help to check out <a href="https://www.youtube.com/watch?v=ByxF3xjN2JQ&amp;list=PLxhIMi78eQegFm3XqsylVa-Lm7nfiUshe&amp;t=2m15s" target="_blank" rel="noopener">Mike Sovic&rsquo;s 1-minute intro to the RStudio interface</a> or <a href="https://fast.wistia.net/embed/iframe/520zbd3tij?videoFoam=true" target="_blank" rel="noopener">RStudio&rsquo;s 3-minute intro</a>.)</p>
<p>If you&rsquo;re able to do so, please open RStudio again a bit before Code Club starts &ndash; and in case you run into issues, please join the Zoom call early and we&rsquo;ll troubleshoot.</p>
<h4 id="new-to-r">New to R?</h4>
<p>If you&rsquo;re completely new to R, it will be useful to have a look at some of the resources listed on our <a href="/codeclub-novice/">New to R?</a> page prior to Code Club.</p>
<br>
<hr>
<h2 id="slides">Slides</h2>
<p>On Friday, we started with a couple of <a href="/slides/CC01/">introductory slides</a>.</p>
<br>
<hr>
<hr>
<h2 id="1---create-an-rstudio-project">1 - Create an RStudio Project</h2>
<p>Projects are an RStudio-specific concept that create a special file (<code>.Rproj</code>), primarily to designate a directory as the working directory for everything within it. We recommend <em>creating exactly one separate Project for each research project</em> with an R component &ndash; and for things like Code Club.</p>
<br>
<div class="alert alert-note">
<div>
<p><strong>Why use Projects?</strong></p>
<p>In brief, Projects help you to organize your work and to make it more portable.</p>
<ul>
<li>
<p>They record which scripts (and R Markdown files) are open in RStudio, and will reopen all of those when you reopen the project. This becomes quite handy, say, when you work on three different projects, each of which uses a number of scripts.</p>
</li>
<li>
<p>When using Projects, you generally don&rsquo;t have to manually set your working directory, and can use <em>relative file paths</em> to refer to files within the project. This way, even if you move the project directory, or copy it to a different computer, the same paths will still work. (This would not be the case if you used <a href="https://rdrr.io/r/base/getwd.html" target="_blank" rel="noopener"><code>setwd()</code></a> which will generally require you to use an absolute path, e.g. <a href="https://rdrr.io/r/base/getwd.html" target="_blank" rel="noopener"><code>setwd(&quot;C:/Users/Jelmer/Documents/&quot;)</code></a>.)</p>
</li>
<li>
<p>Projects encourage you to organize research projects inside self-contained directories, rather than with files spread around your computer. This can save you a lot of headaches and increases reproducibility. And because R will restart whenever you switch Projects, there is no risk of unwanted cross-talk between your projects.</p>
</li>
</ul>
</div>
</div>
<br>
<p><strong>Let&rsquo;s create an RStudio Project for Code Club:</strong></p>
<ul>
<li>
<p>Open RStudio locally or <a href="/codeclub-setup/#osc-run-rstudio">start an OSC RStudio Server session</a>.<br>
(<em>If you&rsquo;re at OSC</em>, you should see a file <code>0_CODECLUB.md</code> that&rsquo;s open in your top-left panel. You can ignore/close this file.)</p>
</li>
<li>
<p><em>If you&rsquo;re working locally</em>, create a directory wherever you like on your computer for all things Code Club. You can do this in R using <a href="https://rdrr.io/r/base/files2.html" target="_blank" rel="noopener"><code>dir.create(&quot;path/to/your/dir&quot;)</code></a>, or outside of R.<br>
(<em>If you&rsquo;re at OSC</em>, skip this step because you&rsquo;re automatically inside a Code Club-specific, personal directory.)</p>
</li>
<li>
<p>Click <code>File</code> (top menu bar) &gt; <code>New Project</code>, and then select <code>Existing Directory</code>.</p>
<ul>
<li>
<p><em>If you&rsquo;re working locally</em>, select the Code Club directory that you created in the previous step.</p>
</li>
<li>
<p><em>If you&rsquo;re working at OSC</em>, keep the default choice &ldquo;<a href="https://rdrr.io/r/base/tilde.html" target="_blank" rel="noopener"><code>~</code></a>&rdquo; (i.e., <em>home</em>), which is the directory you started in when entering the RStudio Server session.</p>
</li>
</ul>
</li>
<li>
<p>After RStudio automatically reloads, you should see the file ending in <code>.Rproj</code> in the RStudio <code>Files</code> tab in the lower right pane, and you will have the Project open. All done for now!</p>
</li>
</ul>
<p>(For future Code Club sessions: RStudio will by default reopen the most recently used Project, and therefore, OSC users will have the Project automatically opened. If you&rsquo;re working locally and are also using other Projects, you can open this Project with <code>File</code> &gt; <code>Open Project</code> inside RStudio, or by clicking the <code>.Rproj</code> file in your file browser, which will open RStudio <em>and</em> the Project.)</p>
<br>
<hr>
<h2 id="2---orienting-ourselves">2 - Orienting ourselves</h2>
<h4 id="where-are-we">Where are we?</h4>
<p>We don&rsquo;t need to set our working directory, because our newly created Project is open, and therefore, our working directory is the directory that contains the <code>.Rproj</code> file.</p>
<p>To see where you are, type or copy into the console (bottom left):</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># Print the working directory:</span>
<span class='nf'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span><span class='o'>(</span><span class='o'>)</span>

<span class='c'># List the files in your current directory:</span>
<span class='nf'><a href='https://rdrr.io/r/base/list.files.html'>dir</a></span><span class='o'>(</span><span class='o'>)</span>          <span class='c'># This should print at least the `.RProj` file.</span>
</code></pre>
</div>
<h4 id="create-directories">Create directories</h4>
<p>Create two new directories &ndash; one for this session, and one for a dataset that we will download shortly (and will be reusing across sessions):</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># Dir for Code Club Session 1:</span>
<span class='nf'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='o'>(</span><span class='s'>"S01"</span><span class='o'>)</span>

<span class='c'># Dir for our bird data:</span>
<span class='c'># ("recursive" to create two levels at once.)</span>
<span class='nf'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='o'>(</span><span class='s'>"data/birds/"</span>, recursive <span class='o'>=</span> <span class='kc'>TRUE</span><span class='o'>)</span>
</code></pre>
</div>
<h4 id="create-a-script">Create a script</h4>
<p>To keep a record of what we are doing, and to easily modify and rerun earlier commands, we&rsquo;ll want to save our commands in a script and execute them from there, rather than typing our commands directly in the console.</p>
<ul>
<li>
<p>Click <code>File</code> (top menu bar) &gt; <code>New File</code> &gt; <code>R script</code>.</p>
</li>
<li>
<p>Save the script (<code>File</code> &gt; <code>Save</code>) as <code>S01.R</code> inside your <code>S01</code> directory.</p>
</li>
</ul>
<h4 id="first-line-of-the-script">First line of the script</h4>
<p>We will now load the core set of 8 <em>tidyverse</em> packages all at once. To do so, type/copy the command below on the first line of the script, and then <strong>execute it</strong> by clicking <code>Run</code> (top right of script pane) or by pressing <code>Ctrl Enter</code> (Windows/Linux, this should also work in your browser) or <code>⌘ Enter</code> (Mac).</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># If you're working locally, and did not install it yet:</span>
<span class='c'># install.packages("tidyverse")</span>

<span class='c'># Load the tidyverse (meta)package:</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='http://tidyverse.tidyverse.org'>tidyverse</a></span><span class='o'>)</span>

<span class='c'>#&gt; ── <span style='font-weight: bold;'>Attaching packages</span><span> ─────────────────────────────────────── tidyverse 1.3.0 ──</span></span>

<span class='c'>#&gt; <span style='color: #00BB00;'>✔</span><span> </span><span style='color: #0000BB;'>ggplot2</span><span> 3.3.2     </span><span style='color: #00BB00;'>✔</span><span> </span><span style='color: #0000BB;'>purrr  </span><span> 0.3.4</span></span>
<span class='c'>#&gt; <span style='color: #00BB00;'>✔</span><span> </span><span style='color: #0000BB;'>tibble </span><span> 3.0.4     </span><span style='color: #00BB00;'>✔</span><span> </span><span style='color: #0000BB;'>dplyr  </span><span> 1.0.2</span></span>
<span class='c'>#&gt; <span style='color: #00BB00;'>✔</span><span> </span><span style='color: #0000BB;'>tidyr  </span><span> 1.1.2     </span><span style='color: #00BB00;'>✔</span><span> </span><span style='color: #0000BB;'>stringr</span><span> 1.4.0</span></span>
<span class='c'>#&gt; <span style='color: #00BB00;'>✔</span><span> </span><span style='color: #0000BB;'>readr  </span><span> 1.3.1     </span><span style='color: #00BB00;'>✔</span><span> </span><span style='color: #0000BB;'>forcats</span><span> 0.5.0</span></span>

<span class='c'>#&gt; ── <span style='font-weight: bold;'>Conflicts</span><span> ────────────────────────────────────────── tidyverse_conflicts() ──</span></span>
<span class='c'>#&gt; <span style='color: #BB0000;'>✖</span><span> </span><span style='color: #0000BB;'>dplyr</span><span>::</span><span style='color: #00BB00;'>filter()</span><span> masks </span><span style='color: #0000BB;'>stats</span><span>::filter()</span></span>
<span class='c'>#&gt; <span style='color: #BB0000;'>✖</span><span> </span><span style='color: #0000BB;'>dplyr</span><span>::</span><span style='color: #00BB00;'>lag()</span><span>    masks </span><span style='color: #0000BB;'>stats</span><span>::lag()</span></span>
</code></pre>
</div>
<p>If this worked, you should get the same output as shown in the code block above: it attached 8 packages, and it warns that some of its functions are now &ldquo;masking&rdquo; base R functions.</p>
<div class="alert alert-note">
<div>
<p>The <em>tidyverse</em> is a very popular and useful ecosystem of R packages for data analysis, which we will be using a lot in Code Club.</p>
<p>When we refer to &ldquo;<em>base R</em>&rdquo; as opposed to the <em>tidyverse</em>, we mean functions that are loaded in R by default (without loading a package), and that can perform similar operations in a different way.</p>
</div>
</div>
<br>
<hr>
<h2 id="3---getting-our-dataset">3 - Getting our dataset</h2>
<p>We downloaded a Great Backyard Bird Count (GBBC) <a href="https://www.gbif.org/dataset/82cb293c-f762-11e1-a439-00145eb45e9a" target="_blank" rel="noopener">dataset</a> from the <a href="https://www.gbif.org/" target="_blank" rel="noopener">Global Biodiversity Information Facility (GBIF)</a>. Because the file was 3.1 GB large, we selected only the records from Ohio and removed some uninformative columns. We also added columns with English names and the breeding range for each species. We&rsquo;ll download the resulting much smaller file (41.5 MB) from our Github repo.</p>
<div class="alert alert-note">
<div>
<h3 id="the-great-backyard-bird-count">The Great Backyard Bird Count</h3>
<p align="center">
<img src=GBBC_screenshot.png width="95%">
</p>
<p>The <a href="https://gbbc.birdcount.org/" target="_blank" rel="noopener">GBBC</a> is an annual citizen science event where everyone is encouraged to to identify and count birds in their backyard &ndash; or anywhere else &ndash; for at least 15 minutes, and report their sightings online. Since 2013, it is a global event, but it has been organized in the US and Canada since 1998.</p>
</div>
</div>
<h4 id="download-the-data">Download the data</h4>
<p>Let&rsquo;s download the dataset using the <a href="https://rdrr.io/r/utils/download.file.html" target="_blank" rel="noopener"><code>download.file()</code></a> function:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># The URL to our file:</span>
<span class='nv'>birds_file_url</span> <span class='o'>&lt;-</span> <span class='s'>"https://raw.githubusercontent.com/biodash/biodash.github.io/master/assets/data/birds/backyard-birds_Ohio.tsv"</span>

<span class='c'># The path to the file we want to download to:</span>
<span class='nv'>birds_file</span> <span class='o'>&lt;-</span> <span class='s'>"data/birds/backyard-birds_Ohio.tsv"</span>

<span class='c'># Download:</span>
<span class='nf'><a href='https://rdrr.io/r/utils/download.file.html'>download.file</a></span><span class='o'>(</span>url <span class='o'>=</span> <span class='nv'>birds_file_url</span>, destfile <span class='o'>=</span> <span class='nv'>birds_file</span><span class='o'>)</span>
</code></pre>
</div>
<h4 id="read-the-data">Read the data</h4>
<p>Now, let&rsquo;s read the file into R. The <code>.tsv</code> extension (&ldquo;tab-separated values&rdquo;) tells us this is a plain text file in which columns are separated by tabs, so we will use a convenience function from the <em>readr</em> package (which is loaded as part of the core set <em>tidyverse</em> packages) for exactly this type of file:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># Read the data:</span>
<span class='nv'>birds</span> <span class='o'>&lt;-</span> <span class='nf'>read_tsv</span><span class='o'>(</span>file <span class='o'>=</span> <span class='nv'>birds_file</span><span class='o'>)</span>

<span class='c'>#&gt; Parsed with column specification:</span>
<span class='c'>#&gt; cols(</span>
<span class='c'>#&gt;   class = <span style='color: #BB0000;'>col_character()</span><span>,</span></span>
<span class='c'>#&gt;   order = <span style='color: #BB0000;'>col_character()</span><span>,</span></span>
<span class='c'>#&gt;   family = <span style='color: #BB0000;'>col_character()</span><span>,</span></span>
<span class='c'>#&gt;   genus = <span style='color: #BB0000;'>col_character()</span><span>,</span></span>
<span class='c'>#&gt;   species = <span style='color: #BB0000;'>col_character()</span><span>,</span></span>
<span class='c'>#&gt;   locality = <span style='color: #BB0000;'>col_character()</span><span>,</span></span>
<span class='c'>#&gt;   stateProvince = <span style='color: #BB0000;'>col_character()</span><span>,</span></span>
<span class='c'>#&gt;   decimalLatitude = <span style='color: #00BB00;'>col_double()</span><span>,</span></span>
<span class='c'>#&gt;   decimalLongitude = <span style='color: #00BB00;'>col_double()</span><span>,</span></span>
<span class='c'>#&gt;   eventDate = <span style='color: #0000BB;'>col_datetime(format = "")</span><span>,</span></span>
<span class='c'>#&gt;   species_en = <span style='color: #BB0000;'>col_character()</span><span>,</span></span>
<span class='c'>#&gt;   range = <span style='color: #BB0000;'>col_character()</span></span>
<span class='c'>#&gt; )</span>
</code></pre>
</div>
<p>Done! We have now read our data into a <em>tibble</em>, which is a type of data frame (formally a <em>data.frame</em>): R&rsquo;s object class to deal with tabular data wherein each column can contain a different type of data (numeric, characters/strings, etc).</p>
<br>
<hr>
<h2 id="4---exploring-backyard-birds">4 - Exploring backyard birds</h2>
<div class="puzzle">
<h3 id="exercise-1">Exercise 1</h3>
<p><strong>What&rsquo;s in the dataset?</strong></p>
<ul>
<li>
<p>Explore the dataset using some functions and methods you may know to get a quick overview of data(frames), and try to understand what you see. What does a single row represent, and what is in each column? (Be sure to check out the hints below at some point, especially if you&rsquo;re stuck.)</p>
</li>
<li>
<p>Pay attention to the data types (e.g., &ldquo;character&rdquo; or <code>chr</code>) of the different columns, which several of these functions print. The output of our <code>read_tsv()</code> command also printed this information &ndash; this function parsed our columns as the types we see now. Were all the columns parsed correctly?</p>
</li>
<li>
<p>How many rows and how many columns does the dataset have?</p>
</li>
<li>
<p>What are some questions you would like to explore with this dataset? We&rsquo;ll collect some of these and try to answer them in later sessions. If your group has sufficient R skills already, you are also welcome to go ahead and try to answer one or more of these questions.</p>
</li>
</ul>
<details>
<summary> Hints (click here) </summary> <br>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># Type an object's name to print it to screen:</span>
<span class='nv'>birds</span>
<span class='c'># Same as above, but explicitly calling print():</span>
<span class='nf'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='o'>(</span><span class='nv'>birds</span><span class='o'>)</span>   

<span class='c'># For column-wise information (short for "structure"):</span>
<span class='nf'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='o'>(</span><span class='nv'>birds</span><span class='o'>)</span>
<span class='c'># tidyverse version of str():</span>
<span class='nf'>glimpse</span><span class='o'>(</span><span class='nv'>birds</span><span class='o'>)</span>

<span class='c'># In RStudio, open object in a separate tab:</span>
<span class='nf'><a href='https://rdrr.io/r/utils/View.html'>View</a></span><span class='o'>(</span><span class='nv'>birds</span><span class='o'>)</span>
</code></pre>
</div>
<ul>
<li>
<p>Note that in R, <code>dbl</code> (for &ldquo;double&rdquo;) and <code>num</code> (for &ldquo;numeric&rdquo;) are both used, and almost interchangeably so, for floating point numbers. (Integers are a separate type that are simply called &ldquo;integers&rdquo; and abbreviated as <code>int</code>, but we have no integer columns in this dataset.)</p>
</li>
<li>
<p><code>read_tsv()</code> parsed our date as a &ldquo;date-time&rdquo; (<code>dttm</code> or <code>POSIXct</code> for short), which contains both a date and a time. In our case, it looks like the time is always &ldquo;00:00:00&rdquo; and thus doesn&rsquo;t provide any information.</p>
</li>
</ul>
</details>
<br>
<details>
<summary> Solutions (click here) </summary> <br>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># Just printing the glimpse() output,</span>
<span class='c'># which will show the number of rows and columns:</span>
<span class='nf'>glimpse</span><span class='o'>(</span><span class='nv'>birds</span><span class='o'>)</span>

<span class='c'>#&gt; Rows: 311,441</span>
<span class='c'>#&gt; Columns: 12</span>
<span class='c'>#&gt; $ class            <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> "Aves", "Aves", "Aves", "Aves", "Aves", "Aves", "Ave…</span></span>
<span class='c'>#&gt; $ order            <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> "Passeriformes", "Passeriformes", "Passeriformes", "…</span></span>
<span class='c'>#&gt; $ family           <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> "Corvidae", "Corvidae", "Corvidae", "Corvidae", "Cor…</span></span>
<span class='c'>#&gt; $ genus            <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> "Cyanocitta", "Cyanocitta", "Cyanocitta", "Cyanocitt…</span></span>
<span class='c'>#&gt; $ species          <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> "Cyanocitta cristata", "Cyanocitta cristata", "Cyano…</span></span>
<span class='c'>#&gt; $ locality         <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> "44805 Ashland", "45244 Cincinnati", "44132 Euclid",…</span></span>
<span class='c'>#&gt; $ stateProvince    <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> "Ohio", "Ohio", "Ohio", "Ohio", "Ohio", "Ohio", "Ohi…</span></span>
<span class='c'>#&gt; $ decimalLatitude  <span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span> 40.86166, 39.10666, 41.60768, 39.24236, 39.28207, 41…</span></span>
<span class='c'>#&gt; $ decimalLongitude <span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span> -82.31558, -84.32972, -81.50085, -84.35545, -84.4688…</span></span>
<span class='c'>#&gt; $ eventDate        <span style='color: #555555;font-style: italic;'>&lt;dttm&gt;</span><span> 2007-02-16, 2007-02-17, 2007-02-17, 2007-02-19, 200…</span></span>
<span class='c'>#&gt; $ species_en       <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> "Blue Jay", "Blue Jay", "Blue Jay", "Blue Jay", "Blu…</span></span>
<span class='c'>#&gt; $ range            <span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …</span></span>
</code></pre>
</div>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># You can also check the number of rows and columns directly using:</span>
<span class='nf'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='o'>(</span><span class='nv'>birds</span><span class='o'>)</span>          <span class='c'># Will return the number of rows and columns</span>

<span class='c'>#&gt; [1] 311441     12</span>


<span class='nf'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='o'>(</span><span class='nv'>birds</span><span class='o'>)</span>         <span class='c'># Will return the number of rows</span>

<span class='c'>#&gt; [1] 311441</span>

<span class='nf'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='o'>(</span><span class='nv'>birds</span><span class='o'>)</span>         <span class='c'># Will return the number of columns</span>

<span class='c'>#&gt; [1] 12</span>
</code></pre>
</div>
</details>
</div>
<p><br> <br></p>
<hr>
<hr>
<h2 id="bonus-material">Bonus material</h2>
<p>If your breakout group is done with Exercise 1, you can have a look at the bonus material below which includes another exercise. You can also have a look at this as homework. Or not at all!</p>
<br>
<div class="alert alert-note">
<div>
<h3 id="readr-options-for-challenging-files"><code>readr</code> options for challenging files</h3>
<p>Earlier, we successfully read in our file without specifying any arguments other than the file name to the <code>read_tsv()</code> function, i.e. with all the default options. It is not always this easy!</p>
<p>Some options for more complex cases:</p>
<ul>
<li>
<p>The more general counterpart of this function is <code>read_delim()</code>, which allows you to specify the delimiter using the <code>sep</code> argument, e.g. <code>delim=&quot;\t&quot;</code> for tabs.</p>
</li>
<li>
<p>There are also arguments to these functions for when you need to skip lines, when you don&rsquo;t have column headers, when you need to specify the column types of some or all the columns, and so forth &ndash; see this example:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>my_df</span> <span class='o'>&lt;-</span> <span class='nf'>read_delim</span><span class='o'>(</span>
  file <span class='o'>=</span> <span class='s'>"file.txt"</span>,
  delim <span class='o'>=</span> <span class='s'>"\t"</span>,             <span class='c'># Specify tab as delimiter</span>
  col_names <span class='o'>=</span> <span class='kc'>FALSE</span>,        <span class='c'># First line is not a header</span>
  skip <span class='o'>=</span> <span class='m'>3</span>,                 <span class='c'># Skip the first three lines</span>
  comment <span class='o'>=</span> <span class='s'>"#"</span>,            <span class='c'># Skip any line beginning with a "#"</span>
  col_types <span class='o'>=</span> <span class='nf'>cols</span><span class='o'>(</span>         <span class='c'># Specify column types</span>
    col1 <span class='o'>=</span> <span class='nf'>col_character</span><span class='o'>(</span><span class='o'>)</span>, <span class='c'># ..We only need to specify columns for </span>
    col2 <span class='o'>=</span> <span class='nf'>col_double</span><span class='o'>(</span><span class='o'>)</span>     <span class='c'># ..which we need non-automatic typing</span>
    <span class='o'>)</span>
  <span class='o'>)</span>
</code></pre>
</div>
</li>
</ul>
</div>
</div>
<br>
<div class="puzzle">
<h3 id="exercise-2-optional">Exercise 2 (Optional)</h3>
<p><strong>Read this file!</strong></p>
<p>Try to read the following file into R, which is a modified and much smaller version of the bird dataset.</p>
<p>Make the function parse the &ldquo;order&rdquo; column as a factor, and the &ldquo;year&rdquo;, &ldquo;month&rdquo;, and &ldquo;day&rdquo; columns as whatever you think is sensible.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># Download and read the file:</span>
<span class='nv'>birds2_file_url</span> <span class='o'>&lt;-</span> <span class='s'>"https://raw.githubusercontent.com/biodash/biodash.github.io/master/assets/data/birds/backyard-birds_read-challenge.txt"</span>
<span class='nv'>birds2_file</span> <span class='o'>&lt;-</span> <span class='s'>"data/birds/backyard-birds_read-challenge.txt"</span>
<span class='nf'><a href='https://rdrr.io/r/utils/download.file.html'>download.file</a></span><span class='o'>(</span>url <span class='o'>=</span> <span class='nv'>birds2_file_url</span>, destfile <span class='o'>=</span> <span class='nv'>birds2_file</span><span class='o'>)</span>
</code></pre>
</div>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># Your turn!</span>
<span class='nv'>birds2</span> <span class='o'>&lt;-</span> <span class='nv'>read_</span>    <span class='c'># Complete the command</span>
</code></pre>
</div>
<details>
<summary> Hints (click here) </summary> <br>
<ul>
<li>
<p>The file is saved as <code>.txt</code>, so the delimiter is not obvious &ndash; first have a look at it (open it in RStudio, a text editor, or the terminal) to determine the delimiter. Then, use <code>read_delim()</code> with manual specification of the delimiter using the <code>delim</code> argument, or use a specialized convenience function.</p>
</li>
<li>
<p>Besides a leading line with no data, there is another problematic line further down. You will need both the <code>skip</code> and <code>comment</code> arguments to circumvent these.</p>
</li>
<li>
<p>Note that <em>readr</em> erroneously parses <code>month</code> as a character column if you don&rsquo;t manually specify its type.</p>
</li>
<li>
<p>Note that you can also use a succinct column type specification like <code>col_types = &quot;fc&quot;</code>, which would parse, for a two-column file, the first column as a factor and the second as a character &ndash; type e.g. <code>?read_tsv</code> for details.</p>
</li>
</ul>
</details>
<br>
<details>
<summary> Bare solution (click here) </summary> <br>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># With succint column type specification:</span>
<span class='nv'>birds2</span> <span class='o'>&lt;-</span> <span class='nf'>read_csv</span><span class='o'>(</span>
  file <span class='o'>=</span> <span class='nv'>birds2_file</span>,
  skip <span class='o'>=</span> <span class='m'>1</span>,
  comment <span class='o'>=</span> <span class='s'>"$"</span>,
  col_types <span class='o'>=</span> <span class='s'>"fcdiii"</span>
  <span class='o'>)</span>

<span class='c'># With long column type specification:</span>
<span class='nv'>birds2</span> <span class='o'>&lt;-</span> <span class='nf'>read_csv</span><span class='o'>(</span>
  file <span class='o'>=</span> <span class='nv'>birds2_file</span>,
  skip <span class='o'>=</span> <span class='m'>1</span>,
  comment <span class='o'>=</span> <span class='s'>"$"</span>,
  col_types <span class='o'>=</span> <span class='nf'>cols</span><span class='o'>(</span>
    order <span class='o'>=</span> <span class='nf'>col_factor</span><span class='o'>(</span><span class='o'>)</span>,
    year <span class='o'>=</span>  <span class='nf'>col_integer</span><span class='o'>(</span><span class='o'>)</span>,
    month <span class='o'>=</span> <span class='nf'>col_integer</span><span class='o'>(</span><span class='o'>)</span>,
    day <span class='o'>=</span> <span class='nf'>col_integer</span><span class='o'>(</span><span class='o'>)</span>
    <span class='o'>)</span>
  <span class='o'>)</span>
</code></pre>
</div>
</details>
<br>
<details>
<summary> Solution with explanations (click here) </summary> <br>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># With succinct column type specification:</span>
<span class='nv'>birds2</span> <span class='o'>&lt;-</span> <span class='nf'>read_csv</span><span class='o'>(</span>     <span class='c'># `read_csv()`: file is comma-delimited</span>
  file <span class='o'>=</span> <span class='nv'>birds2_file</span>,
  skip <span class='o'>=</span> <span class='m'>1</span>,             <span class='c'># First line is not part of the dataframe</span>
  comment <span class='o'>=</span> <span class='s'>"$"</span>,        <span class='c'># Line 228 is a comment that starts with `$`</span>
  col_types <span class='o'>=</span> <span class='s'>"fcdiii"</span>  <span class='c'># "f" for factor, "c" for character,</span>
  <span class='o'>)</span>                     <span class='c'># .."d" for double (=numeric),</span>
                        <span class='c'># .."i" for integer.</span>

<span class='c'># With long column type specification:</span>
<span class='nv'>birds2</span> <span class='o'>&lt;-</span> <span class='nf'>read_csv</span><span class='o'>(</span>
  file <span class='o'>=</span> <span class='nv'>birds2_file</span>,
  skip <span class='o'>=</span> <span class='m'>1</span>,
  comment <span class='o'>=</span> <span class='s'>"$"</span>,
  col_types <span class='o'>=</span> <span class='nf'>cols</span><span class='o'>(</span>        <span class='c'># We can omit columns for which we</span>
    order <span class='o'>=</span> <span class='nf'>col_factor</span><span class='o'>(</span><span class='o'>)</span>,  <span class='c'># ..accept the automatic parsing,</span>
    year <span class='o'>=</span>  <span class='nf'>col_integer</span><span class='o'>(</span><span class='o'>)</span>, <span class='c'># ..when using the long specification. </span>
    month <span class='o'>=</span> <span class='nf'>col_integer</span><span class='o'>(</span><span class='o'>)</span>,
    day <span class='o'>=</span> <span class='nf'>col_integer</span><span class='o'>(</span><span class='o'>)</span>
    <span class='o'>)</span>
  <span class='o'>)</span>
</code></pre>
</div>
</details>
<br>
</div>
<br>
<div class="alert alert-note">
<div>
<h3 id="other-options-for-reading-tabular-data">Other options for reading tabular data</h3>
<p>There are also functions in <em>base R</em> that read tabular data, such as <a href="https://rdrr.io/r/utils/read.table.html" target="_blank" rel="noopener"><code>read.table()</code></a> and <a href="https://rdrr.io/r/utils/read.table.html" target="_blank" rel="noopener"><code>read.delim()</code></a>.</p>
<p>These are generally slower than the <em>readr</em> functions, and have less sensible default options to their arguments. Particularly relevant is how columns with characters (strings) are parsed &ndash; until R 4.0, which was released earlier this year, base R&rsquo;s default behavior was to parse them as <strong>factors</strong>, and this is generally not desirable<sup id="fnref:1"><a href="#fn:1" class="footnote-ref" role="doc-noteref">1</a></sup>. <em>readr</em> functions will never convert columns with strings to factors.</p>
<p>If speed is important, such as when reading in very large files (~ 100s of MBs or larger), you should consider using the <code>fread()</code> function from the <em>data.table</em> package.</p>
<p>Finally, some examples of reading other types of files:</p>
<ul>
<li>Read excel files directly using the <em>readxl</em> package.</li>
<li>Read Google Sheets directly from the web using the <em>googlesheets4</em> package.</li>
<li>Read non-tabular data using the base R <a href="https://rdrr.io/r/base/readLines.html" target="_blank" rel="noopener"><code>readLines()</code></a> function.</li>
</ul>
</div>
</div>
<p><br> <br> <br> <br></p>
<section class="footnotes" role="doc-endnotes">
<hr>
<ol>
<li id="fn:1" role="doc-endnote">
<p>You can check which version of R you are running by typing <a href="https://rdrr.io/r/utils/sessionInfo.html" target="_blank" rel="noopener"><code>sessionInfo()</code></a>. You can also check directly how strings are read by default with <a href="https://rdrr.io/r/base/data.frame.html" target="_blank" rel="noopener"><code>default.stringsAsFactors()</code></a>. To avoid conversion to factors, specify <code>stringsAsFactors = FALSE</code> in your <a href="https://rdrr.io/r/utils/read.table.html" target="_blank" rel="noopener"><code>read.table()</code></a> / <a href="https://rdrr.io/r/utils/read.table.html" target="_blank" rel="noopener"><code>read.delim()</code></a> function call. <a href="#fnref:1" class="footnote-backref" role="doc-backlink">&#x21a9;&#xfe0e;</a></p>
</li>
</ol>
</section>

                    </div>

                    <br>

                    






<div class="article-tags">
  
  <a class="badge badge-light" href="/tags/codeclub/">codeclub</a>
  
  <a class="badge badge-light" href="/tags/backyard-birds/">backyard-birds</a>
  
</div>



<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://biodash.github.io/codeclub/01_backyard-birds/&amp;text=Session%201:%20Backyard%20Birds" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://biodash.github.io/codeclub/01_backyard-birds/&amp;t=Session%201:%20Backyard%20Birds" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=Session%201:%20Backyard%20Birds&amp;body=https://biodash.github.io/codeclub/01_backyard-birds/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://biodash.github.io/codeclub/01_backyard-birds/&amp;title=Session%201:%20Backyard%20Birds" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="whatsapp://send?text=Session%201:%20Backyard%20Birds%20https://biodash.github.io/codeclub/01_backyard-birds/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://biodash.github.io/codeclub/01_backyard-birds/&amp;title=Session%201:%20Backyard%20Birds" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>











  
  
    



  
  
  
  
  
  <div class="media author-card content-widget-hr">
    
      
      <a href="/authors/admin/"><img class="avatar mr-3 avatar-circle" src="/authors/admin/avatar_hu6d2513500ad4d038e15c8975d9d8565d_488180_270x270_fill_q90_lanczos_center.jpg" alt="Jelmer Poelstra"></a>
    

    <div class="media-body">
      <h5 class="card-title"><a href="/authors/admin/">Jelmer Poelstra</a></h5>
      <h6 class="card-subtitle">Bioinformatician at MCIC</h6>
      
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
    <li>
      <a href="mailto:poelstra.1@osu.edu" >
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  
    
    
    
    
    
    
    
      
    
    <li>
      <a href="https://scholar.google.com/citations?user=h_a6S0cAAAAJ" target="_blank" rel="noopener">
        <i class="ai ai-google-scholar"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/jelmerp" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>


  
















                    <br>

                    <hr class="separator">

                    <div class="clearfix">
                        
                        
                        <span class="previous-link">
                            <a href="https://biodash.github.io/codeclub/00_welcome-to-codeclub/">← Welcome to Code Club</a>
                        </span>
                        
                        
                        
                        
                        <span class="next-link">
                            <a href="https://biodash.github.io/codeclub/02_dplyr-core-verbs/">Session 2: dplyr core verbs →</a>
                        </span>
                        
                        
                    </div>

                    <hr class="separator">

                </div>

            </article>

            
        </main>
    </div>
</div>

  </div>

  <div class="page-footer">
    
    
    <div class="container">
      <footer class="site-footer">
    
    <p class="powered-by">
      <a href="/license/">LICENSE: CC-BY-SA <br> <i class="fab fa-creative-commons fa-2x"></i><i class="fab fa-creative-commons-by fa-2x"></i><i class="fab fa-creative-commons-sa fa-2x"></i> </a><br>
    </p>
    
</footer>
    </div>
    
  </div>

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/instant.page/5.1.0/instantpage.min.js" integrity="sha512-1+qUtKoh9XZW7j+6LhRMAyOrgSQKenQ4mluTR+cvxXjP1Z54RxZuzstR/H9kgPXQsVB8IW7DMDFUJpzLjvhGSQ==" crossorigin="anonymous"></script>

      
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      

      
      

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/highlight.min.js" integrity="sha512-TDKKr+IvoqZnPzc3l35hdjpHD0m+b2EC2SrLEgKDRWpxf2rFCxemkgvJ5kfU48ip+Y+m2XVKyOCD85ybtlZDmw==" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/languages/r.min.js"></script>
        
      

    

    
    
    
      <script async defer src="https://maps.googleapis.com/maps/api/js?key="></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/gmaps.js/0.4.25/gmaps.min.js" integrity="sha256-7vjlAeb8OaTrCXZkCNun9djzuB2owUsaO72kXaFDBJs=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
      
      <script id="search-hit-fuse-template" type="text/x-template">
        <div class="search-hit" id="summary-{{key}}">
          <div class="search-hit-content">
            <div class="search-hit-name">
              <a href="{{relpermalink}}">{{title}}</a>
              <div class="article-metadata search-hit-type">{{type}}</div>
              <p class="search-hit-description">{{snippet}}</p>
            </div>
          </div>
        </div>
      </script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
    

    
    <script src="/js/bootstrap.bundle.min.6aed84840afc03ab4d5750157f69c120.js"></script>

    
    
      
      
      
      
      
      
      
    

    
    
    
    
    
    
    
    
      
      
    
    
      
    
    <script src="/en/js/wowchemy.min.2f16af175a3ec51b41b36bb4934b1cf7.js"></script>

    






</body>
</html>
