<!DOCTYPE html><html lang="en-us" >

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  
  
  
    <meta name="generator" content="Wowchemy 5.0.0-beta.3 for Hugo" />
  

  

  
  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Jelmer Poelstra" />

  
  
  
    
  
  <meta name="description" content="During this second part of an introduction to ggplot2, we will get a better understanding of using geoms, with boxplots as an example, and will also learn about formatting and saving plots." />

  
  <link rel="alternate" hreflang="en-us" href="https://biodash.github.io/codeclub/s02e07_ggplot2_part2/" />

  







  




  
  
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
  

  
  
  
    <meta name="theme-color" content="#3f51b5" />
  

  
  
    
    <script src="/js/mathjax-config.js"></script>
  

  
  
  
  
    
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha256-FMvZuGapsJLjouA6k7Eo2lusoAX9i0ShlWFG6qt7SLc=" crossorigin="anonymous">

    
    

    
    
    
      
    
    
      
      
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-light" media="print" onload="this.media='all'">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-dark" media="print" onload="this.media='all'" disabled>
      
    

    
    
    

    

    
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      
      

      
      

      
    
      
      

      
      

      
    
  

  
  
  
    
      
      <link rel="preload" as="style" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap">
      <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media="print" onload="this.media='all'">
    
  

  
  
  
  
  
  <link rel="stylesheet" href="/css/wowchemy.80f9a276c6100914c616a06e45ba8d09.css" />

  




<script async src="https://www.googletagmanager.com/gtag/js?id=G-9F9CQ7P0DX"></script>
<script>
  window.dataLayer = window.dataLayer || [];

  function gtag() {
      dataLayer.push(arguments);
  }

  function trackOutboundLink(url, target) {
    gtag('event', 'click', {
         'event_category': 'outbound',
         'event_label': url,
         'transport_type': 'beacon',
         'event_callback': function () {
           if (target !== '_blank') {
             document.location = url;
           }
         }
    });
    console.debug("Outbound link clicked: " + url);
  }

  function onClickCallback(event) {
    if ((event.target.tagName !== 'A') || (event.target.host === window.location.host)) {
      return;
    }
    trackOutboundLink(event.target, event.target.getAttribute('target'));  
  }

  gtag('js', new Date());
  gtag('config', 'G-9F9CQ7P0DX', {});
  gtag('set', {'cookie_flags': 'SameSite=None;Secure'});

  
  document.addEventListener('click', onClickCallback, false);
</script>


  

  

  




  
  
  

  

  
    <link rel="manifest" href="/index.webmanifest" />
  

  <link rel="icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_32x32_fill_lanczos_center_2.png" />
  <link rel="apple-touch-icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_180x180_fill_lanczos_center_2.png" />

  <link rel="canonical" href="https://biodash.github.io/codeclub/s02e07_ggplot2_part2/" />

  
  
  
  
  
  
  
  
    
    
  
  
  <meta property="twitter:card" content="summary" />
  
  <meta property="og:site_name" content="BioDASH" />
  <meta property="og:url" content="https://biodash.github.io/codeclub/s02e07_ggplot2_part2/" />
  <meta property="og:title" content="S02E07: Intro to ggplot2 (part 2) | BioDASH" />
  <meta property="og:description" content="During this second part of an introduction to ggplot2, we will get a better understanding of using geoms, with boxplots as an example, and will also learn about formatting and saving plots." /><meta property="og:image" content="https://biodash.github.io/media/logo_hu6407506e646a281542ecb299640680ce_37988_300x300_fit_lanczos_2.png" />
    <meta property="twitter:image" content="https://biodash.github.io/media/logo_hu6407506e646a281542ecb299640680ce_37988_300x300_fit_lanczos_2.png" /><meta property="og:locale" content="en-us" />
  
    
      <meta
        property="article:published_time"
        content="2021-10-05T00:00:00&#43;00:00"
      />
    
    <meta property="article:modified_time" content="2021-10-07T13:33:38-04:00">
  

  



  

  

  <link rel='stylesheet' href='/css/highlight-light.css' title='hl-light'>
<link rel='stylesheet' href='/css/highlight-dark.css' title='hl-dark' disabled>



  <title>S02E07: Intro to ggplot2 (part 2) | BioDASH</title>
</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper   " data-wc-page-id="9447a0f876418349500f0f801001e795" >

  
  
  
  
  
  
  
  
  <script src="/js/wowchemy-init.min.7a224883d792570fd34d30c7134eace3.js"></script>

  


<aside class="search-modal" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#" aria-label="Close"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control"
        aria-label="Search...">
        
      </div>

      
      

      

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>



  <div class="page-header">
    











  


<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container-xl">

    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Code Club</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/codeclub-schedule/"><span>- <b>Schedule</b></span></a>
            
              <a class="dropdown-item" href="/codeclub-schedule/#previous-semesters"><span>- Past Sessions</span></a>
            
              <a class="dropdown-item" href="/codeclub-setup/"><span>- Computer Setup</span></a>
            
              <a class="dropdown-item" href="/"><span>------------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-suggest/"><span>- Suggest a Topic</span></a>
            
              <a class="dropdown-item" href="/"><span>-----------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-about/"><span>- <b>About Code Club & Join Us!</b></span></a>
            
          </div>
        </li>

        
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Tutorials & Resources</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/events/2023-02_aws/"><span>AWS Workshop (Feb '23)</span></a>
            
              <a class="dropdown-item" href="/"><span>---------------------</span></a>
            
              <a class="dropdown-item" href="/tutorials/r-resources-tips/"><span>- R Resources and Tips</span></a>
            
              <a class="dropdown-item" href="/"><span>--------------------</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/2020-12-microbiomics-workshop/"><span>- Metabarcoding workshop (Fall '20)</span></a>
            
              <a class="dropdown-item" href="/tutorials/2021-01_rnaseq/"><span>- RNA-seq intro meetings (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/pracs-sp21/"><span>- Computing Course (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/cl-workshop-22/"><span>- Command-line, shell scripting and OSC Workshop (Summer '22)</span></a>
            
          </div>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/support/"><span>Bioinformatics Support</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/about/"><span>About</span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="https://github.com/biodash/biodash.github.io" target="_blank" rel="noopener"><i class='fab fa-github'></i><span></span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/index.xml" target="_blank" rel="noopener"><i class='fas fa-rss'></i><span></span></a>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">

      
      
        
      
        
      
        
      
        
      

      
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      
      <li class="nav-item dropdown theme-dropdown">
        <a href="#" class="nav-link" data-toggle="dropdown" aria-haspopup="true" aria-label="Display preferences">
          <i class="fas fa-moon" aria-hidden="true"></i>
        </a>
        <div class="dropdown-menu">
          <a href="#" class="dropdown-item js-set-theme-light">
            <span>Light</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-dark">
            <span>Dark</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-auto">
            <span>Automatic</span>
          </a>
        </div>
      </li>
      

      
      

    </ul>

  </div>
</nav>


  </div>

  <div class="page-body">
    
<div class="container-fluid docs">
    <div class="row flex-xl-nowrap">
        <div class="d-none d-xl-block col-xl-2 docs-toc">
            <ul class="nav toc-top">
                <li><a href="#" id="back_to_top" class="docs-toc-title">Contents</a></li>
            </ul>
            <nav id="TableOfContents">
  <ul>
    <li><a href="#new-to-code-club">New to Code Club?</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#introduction">Introduction</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#1---a-geom-for-boxplots">1 - A geom for boxplots</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#2---adding-a-plot-layer">2 - Adding a plot layer</a></li>
    <li><a href="#breakout-rooms-i">Breakout Rooms I</a>
      <ul>
        <li><a href="#exercise-1">Exercise 1</a></li>
        <li><a href="#exercise-2">Exercise 2</a></li>
      </ul>
    </li>
    <li><a href="#3---intro-to-formatting-plots">3 - Intro to formatting plots</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#4---formatting-with-theme-and-more">4 - Formatting with <code>theme()</code> and more</a></li>
    <li><a href="#5---adding-labels-to-our-plot">5 - Adding labels to our plot</a></li>
    <li><a href="#breakout-rooms-ii">Breakout Rooms II</a>
      <ul>
        <li><a href="#exercise-3">Exercise 3</a></li>
        <li><a href="#exercise-4">Exercise 4</a></li>
        <li><a href="#exercise-5-bonus">Exercise 5 (bonus)</a></li>
      </ul>
    </li>
    <li><a href="#going-further">Going further</a></li>
  </ul>
</nav>
        </div>
        <main class="col-12 col-md-0 col-xl-10 py-md-3 pl-md-5 docs-content" role="main">
            

            <article class="article">

                












  

  
  
  
<div class="article-container pt-3">
  <h1>S02E07: Intro to ggplot2 (part 2)</h1>

  
  <p class="page-subtitle">Boxplots, layering, and formatting plots.</p>
  

  
    



<div class="article-metadata">

  
  
  
  
  <div>
    

  <span >
      <a href="/authors/admin/">Jelmer Poelstra</a></span>
  </div>
  
  

  
  <span class="article-date">
    
    
    Oct 7, 2021
  </span>
  

  

  

  
  
  
  
  

  
  

</div>

    





  
</div>



                <div class="article-container">
                    <div class="article-style">
                        <br>
<hr>
<h2 id="new-to-code-club">New to Code Club?</h2>
<ul>
<li>
<p>First, check out the <a href="/codeclub-setup/">Code Club Computer Setup</a> instructions, which also has pointers for if you&rsquo;re new to R or RStudio.</p>
</li>
<li>
<p>Please open RStudio before Code Club starts to test things out &ndash; and in case you run into issues, please join the Zoom call early and we&rsquo;ll troubleshoot.</p>
</li>
</ul>
<h4 id="new-to-ggplot2">New to <em>ggplot2</em>?</h4>
<ul>
<li>
<p>Check out the <a href="/codeclub/S02E06_ggplot2/">last Code Club Session</a>, which was the first of this two-part introduction to <em>ggplot2</em>.</p>
</li>
<li>
<p>You may find <a href="https://github.com/rstudio/cheatsheets/blob/master/data-visualization-2.1.pdf" target="_blank" rel="noopener">this <em>ggplot2</em> cheat sheet</a> useful!</p>
</li>
</ul>
<br>
<hr>
<h2 id="introduction">Introduction</h2>
<h4 id="session-goals">Session goals</h4>
<ul>
<li>
<p>Get more familiar with building and layering plots using geoms, using a new geom, <a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" target="_blank" rel="noopener"><code>geom_boxplot()</code></a>, as our starting point.</p>
</li>
<li>
<p>Learn to format plots using <a href="https://ggplot2.tidyverse.org/reference/theme.html" target="_blank" rel="noopener"><code>theme()</code></a> and <a href="https://ggplot2.tidyverse.org/reference/labs.html" target="_blank" rel="noopener"><code>labs()</code></a>.</p>
</li>
</ul>
<br>
<h4 id="getting-set-up">Getting set up</h4>
<p>We will continue to work with the data contained in the Palmer Penguins package. You only have to install it if you didn&rsquo;t do so in a previous session:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"palmerpenguins"</span><span class='o'>)</span></code></pre>
</div>
<p>If the package has been installed, you do need to always load it with the <a href="https://rdrr.io/r/base/library.html" target="_blank" rel="noopener"><code>library()</code></a> function &ndash; and we&rsquo;ll also load the <em>tidyverse</em>, which includes the <em>ggplot2</em> package.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://allisonhorst.github.io/palmerpenguins/'>palmerpenguins</a></span><span class='o'>)</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='o'>)</span></code></pre>
</div>
<br>
<h4 id="penguin-bill-length">Penguin bill length</h4>
<p>We are going to mostly be plotting <code>bill_length_mm</code>, which is the &ldquo;horizontal&rdquo; length of the bill: see the image below.</p>
<figure>
<p align="center">
<img src=figs/culmen_depth.png width="75%" alt="depiction of bill length protruding from the penguins face, and bill depth, the height of the bill">
<figcaption>
Artwork by Allison Horst
</figcaption>
</p>
</figure>
<br>
<hr>
<h2 id="1---a-geom-for-boxplots">1 - A geom for boxplots</h2>
<h4 id="geom-recap">Geom recap</h4>
<p><em>ggplot2</em>&rsquo;s &ldquo;geoms&rdquo; are basically plot types of which there are quite a few available (see the <a href="https://github.com/rstudio/cheatsheets/blob/master/data-visualization-2.1.pdf" target="_blank" rel="noopener">cheatsheet</a>). Last week, we saw two geoms: <code>geom_point()</code> to plot individual data points, and <code>geom_smooth()</code> to fit a line to data points.</p>
<p>While doing so, we also saw two other properties of <em>ggplot2</em> and its geoms:</p>
<ul>
<li>
<p>Geoms can be layered on top of each other.</p>
</li>
<li>
<p>Geoms can simply plot the data &ldquo;as is&rdquo; (<code>geom_point()</code>) or can perform computations under the hood, and show the results of those computations (<code>geom_smooth()</code>).</p>
</li>
</ul>
<p>Let&rsquo;s use a new geom to get a little more fluent with <em>ggplot2</em> basics.</p>
<h4 id="boxplots">Boxplots</h4>
<p>A boxplot is a very useful type of plot that shows you the median as well as the variation of a distribution. <em>ggplot2</em> has the geom <code>geom_boxplot()</code> to create boxplots &ndash; another example of a geom that does calculations for us prior to plotting.</p>
<p>Let&rsquo;s make a boxplot that shows the distribution of penguin bill length (column <code>bill_length_mm</code> in our <code>penguins</code> dataframe) along the y-axis &ndash; recall that we use <code>aes()</code> to refer to a column in the data frame from which the data should be taken:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-3-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<details>
<summary>
<b>Why do we get the warning shown above?</b> (click here)
</summary>
<br>
<p>We got the following warning:</p>
<blockquote>
<p><code>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</code></p>
</blockquote>
<p>We get this warning because 2 rows contain <code>NA</code>s for the variable we are plotting, <code>bill_length_mm</code>.</p>
<p>We could take a look at those rows as follows:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> <span class='nf'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='o'>(</span><span class='nf'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='o'>(</span><span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span>
<span class='c'>#&gt; <span style='color: #555555;'># A tibble: 2 × 8</span></span>
<span class='c'>#&gt;   species island bill_length_mm bill_depth_mm flipper_length_… body_mass_g sex  </span>
<span class='c'>#&gt;   <span style='color: #555555; font-style: italic;'>&lt;fct&gt;</span>   <span style='color: #555555; font-style: italic;'>&lt;fct&gt;</span>           <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>         <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>            <span style='color: #555555; font-style: italic;'>&lt;int&gt;</span>       <span style='color: #555555; font-style: italic;'>&lt;int&gt;</span> <span style='color: #555555; font-style: italic;'>&lt;fct&gt;</span></span>
<span class='c'>#&gt; <span style='color: #555555;'>1</span> Adelie  Torge…             <span style='color: #BB0000;'>NA</span>            <span style='color: #BB0000;'>NA</span>               <span style='color: #BB0000;'>NA</span>          <span style='color: #BB0000;'>NA</span> <span style='color: #BB0000;'>NA</span>   </span>
<span class='c'>#&gt; <span style='color: #555555;'>2</span> Gentoo  Biscoe             <span style='color: #BB0000;'>NA</span>            <span style='color: #BB0000;'>NA</span>               <span style='color: #BB0000;'>NA</span>          <span style='color: #BB0000;'>NA</span> <span style='color: #BB0000;'>NA</span>   </span>
<span class='c'>#&gt; <span style='color: #555555;'># … with 1 more variable: year &lt;int&gt;</span></span></code></pre>
</div>
<p>And we could <em>remove</em> those rows as follows, saving the results in a new dataframe:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'>## By negating the `is.na` condition:</span>
<span class='nv'>penguins_noNA</span> <span class='o'>&lt;-</span> <span class='nv'>penguins</span> <span class='o'>%&gt;%</span> <span class='nf'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='o'>(</span><span class='o'>!</span><span class='nf'><a href='https://rdrr.io/r/base/NA.html'>is.na</a></span><span class='o'>(</span><span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span>

<span class='c'>## Or using the specialized `drop_na` function:</span>
<span class='nv'>penguins_noNA</span> <span class='o'>&lt;-</span> <span class='nv'>penguins</span> <span class='o'>%&gt;%</span> <span class='nf'>drop_na</span><span class='o'>(</span><span class='nv'>bill_length_mm</span><span class='o'>)</span></code></pre>
</div>
</details>
<details>
<summary>
The nitty-gritty of boxplots: what is shown exactly? (click here)
</summary>
<br>
<ul>
<li>Lower whisker = smallest observation greater than or equal to lower hinge - 1.5 * IQR</li>
<li>Lower hinge/bottom line of box part of boxplot = 25% quantile</li>
<li>Middle line = median = 50% quantile</li>
<li>Upper hinge/top line of box part of boxplot = 75% quantile</li>
<li>Upper whisker = largest observation less than or equal to upper hinge + 1.5 * IQR</li>
</ul>
</details>
<br>
<p>That worked, but the plot shows the distribution of bill lengths across all 3 species together, which is not that informative. To separate species along the x-axis, we can map the <code>species</code> column to x:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-6-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Great! We can see, at a glance, that Adelie Penguins tend to have considerably shorter bills than the other two species. Chinstrap&rsquo;s bills are just a bit longer than those of Gentoos, but the longest-billed bird is a Gentoo.</p>
<br>
<hr>
<h2 id="2---adding-a-plot-layer">2 - Adding a plot layer</h2>
<p>To get an even better sense of the distribution of bill lengths, and also of our sample sizes, we may want to add the raw data points to our boxplot using <code>geom_point()</code>.</p>
<p>(You may have noticed that in the previous plot, a Gentoo Penguin outlier was shown as a point. To prevent plotting that point twice, we will add <code>outlier.shape = NA</code> to the boxplot call, a somewhat roundabout way of saying that we don&rsquo;t want to plot outliers.)</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span>,
               outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Error: geom_point requires the following missing aesthetics: x and y</span></code></pre>
</div>
<p>Why did this not work?</p>
<p>We had previously species the aesthetics mapping inside the <code>geom_boxplot()</code> call &ndash; that is, we set it for that geom only (&ldquo;local aesthetics&rdquo;) and not for the entire plot (&ldquo;global aesthetics&rdquo;). To add a <code>geom_point()</code> layer with the same aesthetics, we can do one of two things:</p>
<ul>
<li>Set the aesthetic mapping globally, i.e. inside the <code>ggplot()</code> call, or</li>
<li>Set a local aesthetic mapping <em>also</em> inside <code>geom_point()</code>.</li>
</ul>
<p>Let&rsquo;s do the former, so we are not repeating ourselves:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-8-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>This doesn&rsquo;t look too good because many of the points are plotted on top of each other. We can use a few arguments to <code>geom_point()</code> to make some changes:</p>
<ul>
<li>
<p>Add <code>position = &quot;jitter&quot;</code> to the <code>geom_point()</code> call to introduce a small amount of randomness to our points to make us able to see them better.</p>
</li>
<li>
<p>Add <code>size = 1</code> to make the point size a little smaller (1.5 is the default).</p>
</li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-9-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Note that <code>position = &quot;jitter&quot;</code> and <code>size = 1</code> are <em>not</em> specified as mappings (i.e., not inside <code>mapping = aes()</code>): here, we are not mapping data to the plot, but are just changing some &ldquo;settings&rdquo;.</p>
<div class="alert alert-note">
<div>
<p>Because jittering is so common, there is also a specialized jittering geom available: <code>geom_jitter()</code> is shorthand for <code>geom_point(position = &quot;jitter&quot;)</code>.</p>
<p>So, we could have also used the following code to create the same plot:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_jitter</span><span class='o'>(</span>size <span class='o'>=</span> <span class='m'>1</span><span class='o'>)</span></code></pre>
</div>
</div>
</div>
<br>
<hr>
<h2 id="breakout-rooms-i">Breakout Rooms I</h2>
<div class="puzzle">
<div>
<h3 id="exercise-1">Exercise 1</h3>
<ul>
<li>
<p>Run the code below and figure out what the problem is.<br>
(And why do you think <em>ggplot2</em> creates a legend with the item &ldquo;blue&rdquo;, instead of throwing an error?)</p>
</li>
<li>
<p>Modify the code to get the originally intended effect: blue points.</p>
</li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>color <span class='o'>=</span> <span class='s'>"blue"</span><span class='o'>)</span>,
             position <span class='o'>=</span> <span class='s'>"jitter"</span><span class='o'>)</span></code></pre>
</div>
<details>
<summary>
Hints (click here)
</summary>
<br>
<ul>
<li>Here is the botched plot:</li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>color <span class='o'>=</span> <span class='s'>"blue"</span><span class='o'>)</span>,
             position <span class='o'>=</span> <span class='s'>"jitter"</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-12-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<ul>
<li>Should <code>color = &quot;blue&quot;</code> be a mapping, that is, should it be part of the <code>mapping = aes()</code> argument?</li>
</ul>
</details>
<details>
<summary>
Solution (click here)
</summary>
<br>
<ul>
<li>
<p><strong>The problem with the original code</strong> is that <code>color = &quot;blue&quot;</code> should <em>not</em> be a mapping.</p>
</li>
<li>
<p><strong>Why <em>ggplot2</em> does not throw an error:</strong> the <code>mapping</code> argument is used to map data to an aesthetic like point color. Normally, that data is a column in the dataframe, but because the code quotes &ldquo;blue&rdquo; (<code>color = &quot;blue&quot;</code> instead of <code>color = blue</code>), <em>ggplot2</em> does not assume it is a column and instead creates a variable on the fly that just contains the value &ldquo;blue&rdquo;.</p>
</li>
<li>
<p><strong>The correct code to color points blue</strong>:</p>
</li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>color <span class='o'>=</span> <span class='s'>"blue"</span>,
             position <span class='o'>=</span> <span class='s'>"jitter"</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-13-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<br>
<div class="puzzle">
<div>
<h3 id="exercise-2">Exercise 2</h3>
<p>Violin plots are somewhat similar to boxplots, but show a density distribution. Using Google, find out which <em>ggplot2</em> geom creates a violin plot, and then make one plotting bill length by species like we have done for boxplots.</p>
<details>
<summary>
Hints (click here)
</summary>
<br>
<ul>
<li>
<p><code>geom_violin()</code> is the geom that creates violin plots.</p>
</li>
<li>
<p>Other than the geom function, you can leave the code the same as in the previous examples.</p>
</li>
</ul>
</details>
<details>
<summary>
Solution (click here)
</summary>
<br>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, x <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_violin</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_ydensity).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-14-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<br>
<hr>
<h2 id="3---intro-to-formatting-plots">3 - Intro to formatting plots</h2>
<p>So far, we have mostly been concerned with <em>what</em> we are plotting, and haven&rsquo;t paid much attention to how our plot looks. But I, for one, dislike that gray background to the plot, and perhaps the axis labels are a little small?</p>
<p><em>ggplot2</em> offers <em>many</em> options to modify the look of our plot. There are so many that it isn&rsquo;t really possible to remember even the majority of them. Therefore, even for daily users of <em>ggplot2</em>, creating a publication-ready figure will usually involve some Googling or checking the <a href="https://ggplot2.tidyverse.org/" target="_blank" rel="noopener">ggplot2 documentation</a>.</p>
<p>Let&rsquo;s have a look at some of the most commonly used options to change the look of <em>ggplot2</em> plots.</p>
<h4 id="a-starting-plot">A starting plot</h4>
<p>We&rsquo;ll start with the following plot, similar to one we have created before:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>, y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span>, color <span class='o'>=</span> <span class='s'>"grey70"</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-15-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>(Note the addition of <code>color = &quot;grey70&quot;</code> to make the points less dominant in the plot.<br>
See <a href="http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf" target="_blank" rel="noopener">this PDF for an overview of named colors in R</a>.)</p>
<br>
<hr>
<h2 id="4---formatting-with-theme-and-more">4 - Formatting with <code>theme()</code> and more</h2>
<p>The quickest way to modify the overall look and feel of our plot is by using a different &ldquo;complete theme&rdquo;. The default theme is <code>theme_gray()</code>, which comes with that gray background and many other settings that control what the plot looks like.</p>
<p>The <em>ggplot2</em> documentation has <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" target="_blank" rel="noopener">a list of complete themes</a> that shows you what they look like.</p>
<p>Let&rsquo;s switch to a different theme, <code>theme_classic()</code>, for our penguin boxplot:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>, y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span>, color <span class='o'>=</span> <span class='s'>"grey70"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-16-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>These complete theme functions (<code>theme_&lt;theme-name&gt;</code>) also take a few arguments &ndash;<br>
<code>base_size</code> is very useful if we want to simultaneously change the size of all text labels:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>, y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span>, color <span class='o'>=</span> <span class='s'>"grey70"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span>base_size <span class='o'>=</span> <span class='m'>14</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-17-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>This retains the relative sizes of different labels. For instance, note that in both plots, the &ldquo;axis titles&rdquo; (<code>species</code> on x, <code>bill_lenth_mm</code> on y) are larger than the &ldquo;axis text&rdquo; (the labels at the tick marks).</p>
<p>If we wanted to change individual theme components like those, we would need to use the <code>theme()</code> function (check <a href="https://ggplot2.tidyverse.org/reference/theme.html" target="_blank" rel="noopener">its documentation page</a> to see the many possible arguments).</p>
<p>For example, to make axis titles and axis text/labels the same size:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>, y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span>, color <span class='o'>=</span> <span class='s'>"grey70"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span>base_size <span class='o'>=</span> <span class='m'>14</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme</span><span class='o'>(</span>axis.text <span class='o'>=</span> <span class='nf'>element_text</span><span class='o'>(</span>size <span class='o'>=</span> <span class='m'>14</span><span class='o'>)</span>,
        axis.title <span class='o'>=</span> <span class='nf'>element_text</span><span class='o'>(</span>size <span class='o'>=</span> <span class='m'>14</span><span class='o'>)</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-18-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<br>
<hr>
<h2 id="5---adding-labels-to-our-plot">5 - Adding labels to our plot</h2>
<p>Right now, the axis titles are simply the names of the columns that we used in the mapping. The y-axis title in particular (<code>bill_length_mm</code>) could be improved. We might also want to add a title and even a subtitle to our plot.</p>
<p>We can do all of this with the <code>labs()</code> function as follows:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>, y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span>, color <span class='o'>=</span> <span class='s'>"grey70"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span>base_size <span class='o'>=</span> <span class='m'>14</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>labs</span><span class='o'>(</span>title <span class='o'>=</span> <span class='s'>"Penguin Bill Length by Species"</span>,
       subtitle <span class='o'>=</span> <span class='s'>"Collected at Palmer Station, Antarctica"</span>,
       x <span class='o'>=</span> <span class='s'>"Penguin species"</span>,     <span class='c'># x-axis label</span>
       y <span class='o'>=</span> <span class='s'>"Bill length (mm)"</span><span class='o'>)</span>    <span class='c'># y-axis label</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-19-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<br>
<hr>
<h2 id="breakout-rooms-ii">Breakout Rooms II</h2>
<div class="puzzle">
<div>
<h3 id="exercise-3">Exercise 3</h3>
<ul>
<li>
<p>Modify the code used to produce the last plot (just above this exercise) to try several of the themes from the <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" target="_blank" rel="noopener">list of complete themes</a>.</p>
<p>Do you have a preference?</p>
</li>
</ul>
<p>The <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" target="_blank" rel="noopener">list of complete themes</a> also shows that these functions have a few more arguments than the <code>base_size</code> one we explored.</p>
<ul>
<li>
<p>Bonus: Change the <code>base_line_size</code>. What does it do?</p>
</li>
<li>
<p>Bonus (<strong>may not work out of the box on Windows</strong>): Using a different font family can nicely shake things up &ndash; this is the <code>base_family</code> argument. Most standard font family names (e.g. see <a href="https://www.w3.org/Style/Examples/007/fonts.en.html" target="_blank" rel="noopener">this list</a>) should work. For instance, you can try <code>Optima</code>, <code>Verdana</code>, <code>Times New Roman</code>, <code>Courier</code>, or <code>cursive</code>.</p>
</li>
</ul>
<details>
<summary>
Example solution (click here)
</summary>
<br>
<p>With <code>theme_bw()</code> and:</p>
<ul>
<li><code>base_line_size = 1</code> (thicker axis lines)</li>
<li>the <code>cursive</code> font family using <code>base_family = cursive</code></li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>, y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span>, color <span class='o'>=</span> <span class='s'>"grey70"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_bw</span><span class='o'>(</span>base_size <span class='o'>=</span> <span class='m'>14</span>,
                base_line_size <span class='o'>=</span> <span class='m'>1</span>,
                base_family <span class='o'>=</span> <span class='s'>"cursive"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>labs</span><span class='o'>(</span>title <span class='o'>=</span> <span class='s'>"Penguin Bill Length by Species and Sex"</span>,
       subtitle <span class='o'>=</span> <span class='s'>"Collected at Palmer Station, Antarctica"</span>,
       x <span class='o'>=</span> <span class='s'>"Penguin Species"</span>, 
       y <span class='o'>=</span> <span class='s'>"Bill length (mm)"</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-20-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<br>
<div class="puzzle">
<div>
<h3 id="exercise-4">Exercise 4</h3>
<ul>
<li>Modify your code from Exercise 3 to color the jittered points, but not the boxplots, according to sex.</li>
</ul>
<p>As we also saw last week, a legend should have automatically appeared when mapping color to a variable. But what if we wanted to move the legend from the right to the top of the plot?</p>
<ul>
<li>Scroll through the <a href="https://ggplot2.tidyverse.org/reference/theme.html" target="_blank" rel="noopener"><code>theme()</code> documentation</a> and try and find the argument that controls the position of the legend. Then, use this argument to move the legend to the top.</li>
</ul>
<details>
<summary>
Hints (click here)
</summary>
<br>
<ul>
<li>
<p>To color points by sex without modifying the boxplots, add the mapping locally for <code>geom_point()</code> only.</p>
</li>
<li>
<p>To move the legend, use the <code>legend.position</code> argument of <code>theme()</code>.</p>
</li>
</ul>
</details>
<details>
<summary>
Solution (click here)
</summary>
<br>
<ul>
<li>Color points by sex:</li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>, y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span>,
             mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>color <span class='o'>=</span> <span class='nv'>sex</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span>base_size <span class='o'>=</span> <span class='m'>14</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>labs</span><span class='o'>(</span>title <span class='o'>=</span> <span class='s'>"Penguin Bill Length by Species and Sex"</span>,
       subtitle <span class='o'>=</span> <span class='s'>"Collected at Palmer Station, Antarctica"</span>,
       x <span class='o'>=</span> <span class='s'>"Penguin Species"</span>,
       y <span class='o'>=</span> <span class='s'>"Bill length (mm)"</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-21-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<ul>
<li>Move the legend to the top:</li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span>,
       mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>, y <span class='o'>=</span> <span class='nv'>bill_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span>outlier.shape <span class='o'>=</span> <span class='kc'>NA</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>position <span class='o'>=</span> <span class='s'>"jitter"</span>, size <span class='o'>=</span> <span class='m'>1</span>,
             mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>color <span class='o'>=</span> <span class='nv'>sex</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span>base_size <span class='o'>=</span> <span class='m'>14</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme</span><span class='o'>(</span>legend.position <span class='o'>=</span> <span class='s'>"top"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>labs</span><span class='o'>(</span>title <span class='o'>=</span> <span class='s'>"Penguin Bill Length by Species and Sex"</span>,
       subtitle <span class='o'>=</span> <span class='s'>"Collected at Palmer Station, Antarctica"</span>,
       x <span class='o'>=</span> <span class='s'>"Penguin Species"</span>,
       y <span class='o'>=</span> <span class='s'>"Bill length (mm)"</span><span class='o'>)</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing non-finite values (stat_boxplot).</span>
<span class='c'>#&gt; Warning: Removed 2 rows containing missing values (geom_point).</span>
</code></pre>
<p><img src="figs/unnamed-chunk-22-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<div class="puzzle">
<div>
<h3 id="exercise-5-bonus">Exercise 5 (bonus)</h3>
<ul>
<li>
<p>Try to modify one of the previous plots to get separate boxes for each combination of species and sex, and no jittered points (don&rsquo;t hesitate to look at the hints!).</p>
</li>
<li>
<p>Did you see the <code>NA</code> sex (i.e., missing data)? Recreate the plot without the missing data.</p>
</li>
</ul>
<details>
<summary>
Hints (click here)
</summary>
<br>
<ul>
<li>
<p>Map <code>sex</code> to <code>color</code> or <code>fill</code> (and continue to map <code>species</code> to <code>x</code>).</p>
</li>
<li>
<p>Use <code>drop_na(sex)</code> to remove rows with an <code>NA</code> in the <code>sex</code> column. You can save the result in a new dataframe and then plot that dataframe, or you can pipe (<code>%&gt;%</code>) the result straight into the <code>ggplot()</code> function.</p>
</li>
</ul>
</details>
<details>
<summary>
Solution (click here)
</summary>
<br>
<p>This example maps <code>sex</code> to <code>fill</code>, which will &ldquo;fill&rdquo; the box with colors (as we saw before, mapping to <code>color</code> colors the lines instead):</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span>
  <span class='nf'>drop_na</span><span class='o'>(</span><span class='nv'>sex</span><span class='o'>)</span> <span class='o'>%&gt;%</span>
  <span class='nf'>ggplot</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>species</span>,
                       y <span class='o'>=</span> <span class='nv'>bill_length_mm</span>,
                       fill <span class='o'>=</span> <span class='nv'>sex</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>   <span class='c'># Now mapping sex to "fill"</span>
  <span class='nf'>geom_boxplot</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>    <span class='c'># We now do want to see the outliers!</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span>base_size <span class='o'>=</span> <span class='m'>14</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>labs</span><span class='o'>(</span>title <span class='o'>=</span> <span class='s'>"Penguin Bill Length by Species and Sex"</span>,
       subtitle <span class='o'>=</span> <span class='s'>"Collected at Palmer Station, Antarctica"</span>,
       x <span class='o'>=</span> <span class='s'>"Penguin Species"</span>,
       y <span class='o'>=</span> <span class='s'>"Bill length (mm)"</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-23-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<br>
<hr>
<h2 id="going-further">Going further</h2>
<ul>
<li>
<p>One basic aspect of <em>ggplot2</em> that we have not touched upon in this or the previous session are &ldquo;<em>scales</em>&rdquo;. For instance, to change the colors used in mappings (like our boxplot color), we would need to use scales. For an introduction, see the <a href="https://r4ds.had.co.nz/graphics-for-communication.html#scales" target="_blank" rel="noopener">section on scales from the R for Data Science book</a>.</p>
</li>
<li>
<p>We also didn&rsquo;t talk about saving plots, which can be done with the <code>ggsave()</code> function (<a href="https://ggplot2.tidyverse.org/reference/ggsave.html" target="_blank" rel="noopener">documentation page</a>). Just be aware that if you don&rsquo;t specify dimensions with the <code>width</code> and <code>height</code> arguments, they will be taken from the current size of the RStudio plotting window, which is not necessarily what you want.</p>
</li>
</ul>

                    </div>

                    <br>

                    








<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://biodash.github.io/codeclub/s02e07_ggplot2_part2/&amp;text=S02E07:%20Intro%20to%20ggplot2%20%28part%202%29" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://biodash.github.io/codeclub/s02e07_ggplot2_part2/&amp;t=S02E07:%20Intro%20to%20ggplot2%20%28part%202%29" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=S02E07:%20Intro%20to%20ggplot2%20%28part%202%29&amp;body=https://biodash.github.io/codeclub/s02e07_ggplot2_part2/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://biodash.github.io/codeclub/s02e07_ggplot2_part2/&amp;title=S02E07:%20Intro%20to%20ggplot2%20%28part%202%29" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="whatsapp://send?text=S02E07:%20Intro%20to%20ggplot2%20%28part%202%29%20https://biodash.github.io/codeclub/s02e07_ggplot2_part2/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://biodash.github.io/codeclub/s02e07_ggplot2_part2/&amp;title=S02E07:%20Intro%20to%20ggplot2%20%28part%202%29" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>











  
  
    



  
  
  
  
  
  <div class="media author-card content-widget-hr">
    
      
      <a href="/authors/admin/"><img class="avatar mr-3 avatar-circle" src="/authors/admin/avatar_hu6d2513500ad4d038e15c8975d9d8565d_488180_270x270_fill_q90_lanczos_center.jpg" alt="Jelmer Poelstra"></a>
    

    <div class="media-body">
      <h5 class="card-title"><a href="/authors/admin/">Jelmer Poelstra</a></h5>
      <h6 class="card-subtitle">Bioinformatician at MCIC</h6>
      
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
    <li>
      <a href="mailto:poelstra.1@osu.edu" >
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  
    
    
    
    
    
    
    
      
    
    <li>
      <a href="https://scholar.google.com/citations?user=h_a6S0cAAAAJ" target="_blank" rel="noopener">
        <i class="ai ai-google-scholar"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/jelmerp" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>


  
















                    <br>

                    <hr class="separator">

                    <div class="clearfix">
                        
                        
                        <span class="previous-link">
                            <a href="https://biodash.github.io/codeclub/s02e06_ggplot2/">← Code Club S02E06: Visualizing Data</a>
                        </span>
                        
                        
                        
                        
                        <span class="next-link">
                            <a href="https://biodash.github.io/codeclub/s02e08_multiple_plots/">Code Club S02E08: Combining Plots →</a>
                        </span>
                        
                        
                    </div>

                    <hr class="separator">

                </div>

            </article>

            
        </main>
    </div>
</div>

  </div>

  <div class="page-footer">
    
    
    <div class="container">
      <footer class="site-footer">
    
    <p class="powered-by">
      <a href="/license/">LICENSE: CC-BY-SA <br> <i class="fab fa-creative-commons fa-2x"></i><i class="fab fa-creative-commons-by fa-2x"></i><i class="fab fa-creative-commons-sa fa-2x"></i> </a><br>
    </p>
    
</footer>
    </div>
    
  </div>

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/instant.page/5.1.0/instantpage.min.js" integrity="sha512-1+qUtKoh9XZW7j+6LhRMAyOrgSQKenQ4mluTR+cvxXjP1Z54RxZuzstR/H9kgPXQsVB8IW7DMDFUJpzLjvhGSQ==" crossorigin="anonymous"></script>

      
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      

      
      

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/highlight.min.js" integrity="sha512-TDKKr+IvoqZnPzc3l35hdjpHD0m+b2EC2SrLEgKDRWpxf2rFCxemkgvJ5kfU48ip+Y+m2XVKyOCD85ybtlZDmw==" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/languages/r.min.js"></script>
        
      

    

    
    
    
      <script async defer src="https://maps.googleapis.com/maps/api/js?key="></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/gmaps.js/0.4.25/gmaps.min.js" integrity="sha256-7vjlAeb8OaTrCXZkCNun9djzuB2owUsaO72kXaFDBJs=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
      
      <script id="search-hit-fuse-template" type="text/x-template">
        <div class="search-hit" id="summary-{{key}}">
          <div class="search-hit-content">
            <div class="search-hit-name">
              <a href="{{relpermalink}}">{{title}}</a>
              <div class="article-metadata search-hit-type">{{type}}</div>
              <p class="search-hit-description">{{snippet}}</p>
            </div>
          </div>
        </div>
      </script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
    

    
    <script src="/js/bootstrap.bundle.min.6aed84840afc03ab4d5750157f69c120.js"></script>

    
    
      
      
      
      
      
      
      
    

    
    
    
    
    
    
    
    
      
      
    
    
      
    
    <script src="/en/js/wowchemy.min.2f16af175a3ec51b41b36bb4934b1cf7.js"></script>

    






</body>
</html>
