<!DOCTYPE html><html lang="en-us" >

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  
  
  
    <meta name="generator" content="Wowchemy 5.0.0-beta.3 for Hugo" />
  

  

  
  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Jelmer Poelstra" />

  
  
  
    
  
  <meta name="description" content="Today, we&#39;ll discuss how you can run a Principal Component Analysis (PCA) in R with the `prcomp()` function and create three types of plots from the results: a score plot, a scree plot, and a biplot." />

  
  <link rel="alternate" hreflang="en-us" href="https://biodash.github.io/codeclub/s03e03_pca/" />

  







  




  
  
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
  

  
  
  
    <meta name="theme-color" content="#3f51b5" />
  

  
  
    
    <script src="/js/mathjax-config.js"></script>
  

  
  
  
  
    
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha256-FMvZuGapsJLjouA6k7Eo2lusoAX9i0ShlWFG6qt7SLc=" crossorigin="anonymous">

    
    

    
    
    
      
    
    
      
      
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-light" media="print" onload="this.media='all'">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-dark" media="print" onload="this.media='all'" disabled>
      
    

    
    
    

    

    
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      
      

      
      

      
    
      
      

      
      

      
    
  

  
  
  
    
      
      <link rel="preload" as="style" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap">
      <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media="print" onload="this.media='all'">
    
  

  
  
  
  
  
  <link rel="stylesheet" href="/css/wowchemy.80f9a276c6100914c616a06e45ba8d09.css" />

  




<script async src="https://www.googletagmanager.com/gtag/js?id=G-9F9CQ7P0DX"></script>
<script>
  window.dataLayer = window.dataLayer || [];

  function gtag() {
      dataLayer.push(arguments);
  }

  function trackOutboundLink(url, target) {
    gtag('event', 'click', {
         'event_category': 'outbound',
         'event_label': url,
         'transport_type': 'beacon',
         'event_callback': function () {
           if (target !== '_blank') {
             document.location = url;
           }
         }
    });
    console.debug("Outbound link clicked: " + url);
  }

  function onClickCallback(event) {
    if ((event.target.tagName !== 'A') || (event.target.host === window.location.host)) {
      return;
    }
    trackOutboundLink(event.target, event.target.getAttribute('target'));  
  }

  gtag('js', new Date());
  gtag('config', 'G-9F9CQ7P0DX', {});
  gtag('set', {'cookie_flags': 'SameSite=None;Secure'});

  
  document.addEventListener('click', onClickCallback, false);
</script>


  

  

  




  
  
  

  

  
    <link rel="manifest" href="/index.webmanifest" />
  

  <link rel="icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_32x32_fill_lanczos_center_2.png" />
  <link rel="apple-touch-icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_180x180_fill_lanczos_center_2.png" />

  <link rel="canonical" href="https://biodash.github.io/codeclub/s03e03_pca/" />

  
  
  
  
  
  
  
  
    
    
  
  
  <meta property="twitter:card" content="summary" />
  
  <meta property="og:site_name" content="BioDASH" />
  <meta property="og:url" content="https://biodash.github.io/codeclub/s03e03_pca/" />
  <meta property="og:title" content="S03E03: Principal Component Analysis (PCA) | BioDASH" />
  <meta property="og:description" content="Today, we&#39;ll discuss how you can run a Principal Component Analysis (PCA) in R with the `prcomp()` function and create three types of plots from the results: a score plot, a scree plot, and a biplot." /><meta property="og:image" content="https://biodash.github.io/media/logo_hu6407506e646a281542ecb299640680ce_37988_300x300_fit_lanczos_2.png" />
    <meta property="twitter:image" content="https://biodash.github.io/media/logo_hu6407506e646a281542ecb299640680ce_37988_300x300_fit_lanczos_2.png" /><meta property="og:locale" content="en-us" />
  
    
      <meta
        property="article:published_time"
        content="2022-01-27T00:00:00&#43;00:00"
      />
    
    <meta property="article:modified_time" content="2022-01-27T13:22:54-05:00">
  

  



  

  

  <link rel='stylesheet' href='/css/highlight-light.css' title='hl-light'>
<link rel='stylesheet' href='/css/highlight-dark.css' title='hl-dark' disabled>



  <title>S03E03: Principal Component Analysis (PCA) | BioDASH</title>
</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper   " data-wc-page-id="ac6e352bd77208f6035962a6e9211504" >

  
  
  
  
  
  
  
  
  <script src="/js/wowchemy-init.min.7a224883d792570fd34d30c7134eace3.js"></script>

  


<aside class="search-modal" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#" aria-label="Close"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control"
        aria-label="Search...">
        
      </div>

      
      

      

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>



  <div class="page-header">
    











  


<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container-xl">

    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Code Club</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/codeclub-schedule/"><span>- <b>Schedule</b></span></a>
            
              <a class="dropdown-item" href="/codeclub/"><span>- Past Sessions</span></a>
            
              <a class="dropdown-item" href="/codeclub-setup/"><span>- Computer Setup</span></a>
            
              <a class="dropdown-item" href="/"><span>------------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-suggest/"><span>- Suggest a Topic</span></a>
            
              <a class="dropdown-item" href="/codeclub-present/"><span>- Info for Presenters</span></a>
            
              <a class="dropdown-item" href="/"><span>-----------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-about/"><span>- <b>About Code Club & Join Us!</b></span></a>
            
          </div>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/events/"><span>Other Events</span></a>
        </li>

        
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Tutorials & Materials</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/material/"><span><b>Overview</b></span></a>
            
              <a class="dropdown-item" href="/"><span>---------------------</span></a>
            
              <a class="dropdown-item" href="/tutorials/r-resources-tips/"><span>- R Resources and Tips</span></a>
            
              <a class="dropdown-item" href="/"><span>--------------------</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/2020-12-microbiomics-workshop/"><span>- Metabarcoding workshop (Fall '20)</span></a>
            
              <a class="dropdown-item" href="/tutorials/2021-01_rnaseq/"><span>- RNA-seq intro meetings (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/pracs-sp21/"><span>- Computing Course (Spring '21)</span></a>
            
          </div>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/support/"><span>Need Support?</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/about/"><span>About</span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="https://github.com/biodash/biodash.github.io" target="_blank" rel="noopener"><i class='fab fa-github'></i><span></span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/index.xml" target="_blank" rel="noopener"><i class='fas fa-rss'></i><span></span></a>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">

      
      
        
      
        
      
        
      
        
      

      
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      
      <li class="nav-item dropdown theme-dropdown">
        <a href="#" class="nav-link" data-toggle="dropdown" aria-haspopup="true" aria-label="Display preferences">
          <i class="fas fa-moon" aria-hidden="true"></i>
        </a>
        <div class="dropdown-menu">
          <a href="#" class="dropdown-item js-set-theme-light">
            <span>Light</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-dark">
            <span>Dark</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-auto">
            <span>Automatic</span>
          </a>
        </div>
      </li>
      

      
      

    </ul>

  </div>
</nav>


  </div>

  <div class="page-body">
    
<div class="container-fluid docs">
    <div class="row flex-xl-nowrap">
        <div class="d-none d-xl-block col-xl-2 docs-toc">
            <ul class="nav toc-top">
                <li><a href="#" id="back_to_top" class="docs-toc-title">Contents</a></li>
            </ul>
            <nav id="TableOfContents">
  <ul>
    <li><a href="#housekeeping">Housekeeping</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#1---a-brief-intro-to-pca">1 - A brief intro to PCA</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#2---prcomp-scaling-and-centering">2 - <code>prcomp()</code>, scaling, and centering</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#3---our-first-pca">3 - Our first PCA</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#4---exploring-the-output-i">4 - Exploring the output I</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#breakout-rooms-i">Breakout Rooms I</a>
      <ul>
        <li><a href="#exercise-1">Exercise 1</a></li>
        <li><a href="#exercise-2">Exercise 2</a></li>
        <li><a href="#exercise-3-bonus">Exercise 3 (bonus)</a></li>
      </ul>
    </li>
    <li><a href="#5---exploring-the-output-ii">5 - Exploring the output II</a></li>
    <li><a href="#6---scree-plot">6 - Scree plot</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#7---score-classic-pca-plot">7 - Score (classic PCA) plot</a>
      <ul>
        <li></li>
        <li><a href="#a-better-aspect-ratio">A better aspect ratio</a></li>
      </ul>
    </li>
    <li><a href="#8---biplot">8 - Biplot</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#breakout-rooms-ii">Breakout Rooms II</a>
      <ul>
        <li><a href="#exercise-4">Exercise 4</a></li>
        <li><a href="#exercise-5">Exercise 5</a></li>
        <li><a href="#exercise-6-bonus">Exercise 6 (bonus)</a></li>
      </ul>
    </li>
    <li><a href="#further-watching--reading">Further watching &amp; reading</a></li>
  </ul>
</nav>
        </div>
        <main class="col-12 col-md-0 col-xl-10 py-md-3 pl-md-5 docs-content" role="main">
            

            <article class="article">

                












  

  
  
  
<div class="article-container pt-3">
  <h1>S03E03: Principal Component Analysis (PCA)</h1>

  
  <p class="page-subtitle">How to run a PCA in R and plot the results</p>
  

  
    



<div class="article-metadata">

  
  
  
  
  <div>
    

  <span >
      <a href="/authors/admin/">Jelmer Poelstra</a></span>
  </div>
  
  

  
  <span class="article-date">
    
    
    Jan 27, 2022
  </span>
  

  

  

  
  
  
  
  

  
  

</div>

    





  
</div>



                <div class="article-container">
                    <div class="article-style">
                        <figure>
<p align="center">
<img src=PCA.png width="75%">
<figcaption>
A PCA of genetic variation among Europeans, from <a href="https://www.nature.com/articles/nature07331">Novembre et al 2008</a>: "Genes mirror geography within Europe"
</figcaption>
</p>
</figure>
<hr>
<h2 id="housekeeping">Housekeeping</h2>
<h4 id="new-to-code-club">New to Code Club?</h4>
<p>Check out the <a href="/codeclub-setup/">Code Club Computer Setup</a> instructions, which also has pointers for if you&rsquo;re new to R or RStudio. A few related Code Club sessions include:</p>
<ul>
<li><a href="/codeclub/s03e01_ttests/">S03E01</a>: T-tests</li>
<li><a href="/codeclub/s03e02_anova/">S03E01</a>: ANOVA</li>
<li><a href="/codeclub/s02e06_ggplot2/">S02E06</a>: Intro to <em>ggplot2</em>, part I</li>
<li><a href="/codeclub/s02e07_ggplot2_part2/">S02E07</a>: Intro to <em>ggplot2</em>, part II</li>
</ul>
<h4 id="session-goals">Session goals</h4>
<ul>
<li>
<p>Learn <strong>how to perform a PCA</strong> in R using the <a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a> function.</p>
</li>
<li>
<p>Understand what is represented by the <strong>different components of the output</strong>.</p>
</li>
<li>
<p>Learn about <strong>three kinds of plots</strong> commonly used to visualize PCA results, and how to create them.</p>
</li>
</ul>
<div class="alert alert-note">
<div>
<h4 id="r-packages-we-will-use">R packages we will use</h4>
<ul>
<li><em>palmerpenguins</em> &ndash; A data package containing the data we will explore</li>
<li><em>tidyverse</em> &ndash; A metapackage that includes <em>ggplot2</em> which we&rsquo;ll use for plotting, access to the <code>%&gt;%</code> pipe, etc.</li>
<li><em>broom</em> &ndash; We&rsquo;ll again use the <a href="https://generics.r-lib.org/reference/tidy.html" target="_blank" rel="noopener"><code>tidy()</code></a> function in <em>broom</em> to create tidy dataframes from untidy statistical function output</li>
<li><em>glue</em> &ndash; For pasting strings with variables</li>
<li><em>factoextra</em> &ndash; For easily creating a PCA biplot (and other PCA plots)</li>
</ul>
</div>
</div>
<h4 id="getting-set-up">Getting set up</h4>
<p><em>If you plan to just listen during the first part,</em> <em>you can wait until the Breakout Rooms to do the following.</em> <em>Also, instead of copying-and-pasting code, you could download this</em> <em><a href="https://raw.githubusercontent.com/biodash/biodash.github.io/master/content/codeclub/S03E03_PCA/S03E03_PCA.R" target="_blank" rel="noopener">R script</a></em> <em>with today&rsquo;s code.</em></p>
<ul>
<li>
<p>Open a new R script in RStudio (<code>File</code> =&gt; <code>New File</code> =&gt; <code>R Script</code>)</p>
</li>
<li>
<p>Save the script, as something along the lines of <code>codeclub_S03E03_PCA.R</code></p>
</li>
<li>
<p>Copy the following code into the script, for installing packages where needed:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'>## `require(glue)` returns FALSE if glue isn't installed; therefore,</span>
  <span class='c'>## these lines will only try to install packages that aren't already installed.</span>
  <span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='https://allisonhorst.github.io/palmerpenguins/'>palmerpenguins</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"palmerpenguins"</span><span class='o'>)</span>
  <span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"tidyverse"</span><span class='o'>)</span>
  <span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='https://broom.tidymodels.org/'>broom</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"broom"</span><span class='o'>)</span>
  <span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='https://github.com/tidyverse/glue'>glue</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"glue"</span><span class='o'>)</span>
  <span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='http://www.sthda.com/english/rpkgs/factoextra'>factoextra</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"factoextra"</span><span class='o'>)</span></code></pre>
</div>
</li>
<li>
<p>Also copy the following code to load the packages into your R session:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://allisonhorst.github.io/palmerpenguins/'>palmerpenguins</a></span><span class='o'>)</span>
  <span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='o'>)</span>
  <span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://broom.tidymodels.org/'>broom</a></span><span class='o'>)</span>
  <span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://github.com/tidyverse/glue'>glue</a></span><span class='o'>)</span>
  <span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='http://www.sthda.com/english/rpkgs/factoextra'>factoextra</a></span><span class='o'>)</span></code></pre>
</div>
</li>
<li>
<p>Run the code above in the R console.</p>
</li>
</ul>
<br>
<hr>
<h2 id="1---a-brief-intro-to-pca">1 - A brief intro to PCA</h2>
<p><strong>Principal Component Analysis (PCA)</strong> is a popular method that creates <em>&ldquo;summary variables&rdquo; (Principal Components)</em> which represent as much of the information as possible from a high-dimensional dataset.</p>
<p>A <em>high-dimensional dataset</em> is a dataset with measurements for many variables, such as expression levels for thousands of genes.</p>
<p>PCA and similar methods like PCoA and nMDS (see box below) are also called &ldquo;<strong>dimension reduction</strong>&rdquo; or &ldquo;<strong>ordination</strong>&rdquo; methods, and can be classified as a type of <em>unsupervised learning</em>.</p>
<p>PCA is most commonly used for <strong>exploratory data visualization</strong> to see overall patterns in datasets, though you could also use the resulting Principal Components as response variables in a statistical model.</p>
<br>
<div class="alert alert-note">
<div>
<h4 id="glossary">Glossary</h4>
<ul>
<li><strong>Principal Components (PCs)</strong> &ndash; the summary variables that a PCA produces.</li>
<li><strong>Loadings</strong> (<em>rotations</em>) &ndash; Loadings apply to the <em>original variables</em>. They are the contributions of variables to PCs, which form the &ldquo;recipes&rdquo; used to create the PCs.</li>
<li><strong>Scores</strong> (<em>coordinates</em>) &ndash; Scores apply to the <em>samples</em>. These scores, for each PC, are coordinates that can be used to create a score plot which is the &ldquo;classic&rdquo; PCA plot.</li>
<li><strong>Eigenvalue</strong> &ndash; The variance (amount of variation) explained by a PC.</li>
</ul>
</div>
</div>
<br>
<div class="alert alert-note">
<div>
<h4 id="similar-ordination-methods">Similar ordination methods</h4>
<p>Besides PCA, other commonly used ordination methods that are also unconstrained (i.e., with no response variable) include the following:</p>
<ul>
<li>
<p><em>Principal Coordinate Analysis</em> (<strong>PCoA</strong>) is also known as <em>Metric Multidimensional Scaling</em> (<strong>MDS</strong> / mMDS). PCoA allows you to use distance measures other than Euclidean distance and can be run e.g. with <a href="https://rdrr.io/r/stats/cmdscale.html" target="_blank" rel="noopener"><code>stats::cmdscale()</code></a>.</p>
</li>
<li>
<p><em>Non-metric Multidimensional Scaling</em> (<strong>nMDS</strong>) is a non-metric method with quite different inner workings from PCA and PCoA that is especially suitable when your distance values are imprecise. It can be run e.g. with <a href="https://rdrr.io/pkg/vegan/man/metaMDS.html" target="_blank" rel="noopener"><code>vegan::metaMDS()</code></a>.</p>
</li>
</ul>
<p>If you&rsquo;re struggling to pick a suitable ordination approach for your data, take a look at <a href="https://journals.plos.org/ploscompbiol/article/figure?id=10.1371/journal.pcbi.1006907.t001" target="_blank" rel="noopener">Table 1</a> in <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006907" target="_blank" rel="noopener">Nguyen &amp; Holmes 2019</a>.</p>
</div>
</div>
<br>
<hr>
<h2 id="2---prcomp-scaling-and-centering">2 - <code>prcomp()</code>, scaling, and centering</h2>
<p>To perform a PCA analysis in R, there are two functions that can be used without the need to load any packages: <a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a> and <a href="https://rdrr.io/r/stats/princomp.html" target="_blank" rel="noopener"><code>princomp()</code></a>.</p>
<p>(Like last week&rsquo;s <a href="https://rdrr.io/r/stats/aov.html" target="_blank" rel="noopener"><code>aov()</code></a> function, these functions are in the <em>stats</em> package, which is loaded into your R session by default. More PCA functions are available in other packages but these tend to be very similar and/or simply wrap the two base R functions.)</p>
<p>We will use <strong><a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a></strong>, which is preferred among these two due to its slightly better accuracy<sup id="fnref:1"><a href="#fn:1" class="footnote-ref" role="doc-noteref">1</a></sup>.</p>
<h4 id="two-important-data-pre-processing-steps">Two important data pre-processing steps&hellip;</h4>
<p>&hellip;need to be done for many PCA analyses. Luckily, these can be done alongside the PCA computation in a single call to <a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a>:</p>
<p><strong>Centering the data</strong> &ndash; <em>Centering</em> the data around the origin (subtracting the mean of variables) is basically always advisable and is controlled by the <code>center</code> argument of <a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a>, which is <em>set to <code>TRUE</code> by default</em>.</p>
<p><strong>Scaling the data</strong> &ndash; Standardizing the standard deviation across the variables in the data (i.e., scaling) is advisable when variables are in different units or on different scales but is generally not recommended when all variables are of the same type and in the same units (e.g., gene counts<sup id="fnref:2"><a href="#fn:2" class="footnote-ref" role="doc-noteref">2</a></sup>). Whether or not to scale the data is controlled by the <code>scale.</code> argument of <a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a>, which is <em>set to <code>FALSE</code> by default</em>.</p>
<br>
<hr>
<h2 id="3---our-first-pca">3 - Our first PCA</h2>
<p>As a simple example, we want to <em>run a PCA summarizing the four numerical measurements taken for each penguin</em> (bill length, bill depth, flipper length, and body mass) in the <em>palmerpenguins</em> dataset.</p>
<p>First, we&rsquo;ll subset the <code>penguins</code> dataframe to:</p>
<ul>
<li>
<p>Remove rows with <code>NA</code>s (<a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a> will return an error if any of our variables contain <code>NA</code>s)</p>
</li>
<li>
<p>Select only the columns that we want to include in the PCA</p>
</li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'>## Remove rows with NAs</span>
<span class='nv'>penguins_noNA</span> <span class='o'>&lt;-</span> <span class='nf'>drop_na</span><span class='o'>(</span><span class='nv'>penguins</span><span class='o'>)</span>

<span class='c'>## Select columns</span>
<span class='nv'>penguins_for_pca</span> <span class='o'>&lt;-</span> <span class='nv'>penguins_noNA</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>select</span><span class='o'>(</span><span class='nv'>bill_length_mm</span>, <span class='nv'>bill_depth_mm</span>, <span class='nv'>flipper_length_mm</span>, <span class='nv'>body_mass_g</span><span class='o'>)</span></code></pre>
</div>
<p>Let&rsquo;s take a look at the resulting dataframe:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='o'>(</span><span class='nv'>penguins_for_pca</span><span class='o'>)</span>
<span class='c'>#&gt; <span style='color: #555555;'># A tibble: 6 × 4</span></span>
<span class='c'>#&gt;   bill_length_mm bill_depth_mm flipper_length_mm body_mass_g</span>
<span class='c'>#&gt;            <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>         <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>             <span style='color: #555555; font-style: italic;'>&lt;int&gt;</span>       <span style='color: #555555; font-style: italic;'>&lt;int&gt;</span></span>
<span class='c'>#&gt; <span style='color: #555555;'>1</span>           39.1          18.7               181        <span style='text-decoration: underline;'>3</span>750</span>
<span class='c'>#&gt; <span style='color: #555555;'>2</span>           39.5          17.4               186        <span style='text-decoration: underline;'>3</span>800</span>
<span class='c'>#&gt; <span style='color: #555555;'>3</span>           40.3          18                 195        <span style='text-decoration: underline;'>3</span>250</span>
<span class='c'>#&gt; <span style='color: #555555;'>4</span>           36.7          19.3               193        <span style='text-decoration: underline;'>3</span>450</span>
<span class='c'>#&gt; <span style='color: #555555;'>5</span>           39.3          20.6               190        <span style='text-decoration: underline;'>3</span>650</span>
<span class='c'>#&gt; <span style='color: #555555;'>6</span>           38.9          17.8               181        <span style='text-decoration: underline;'>3</span>625</span>

<span class='nf'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='o'>(</span><span class='nv'>penguins_for_pca</span><span class='o'>)</span>
<span class='c'>#&gt; [1] 333   4</span></code></pre>
</div>
<h4 id="run-the-pca">Run the PCA!</h4>
<p>Now, we are ready to run the PCA:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca</span> <span class='o'>&lt;-</span> <span class='nf'><a href='https://rdrr.io/r/stats/prcomp.html'>prcomp</a></span><span class='o'>(</span><span class='nv'>penguins_for_pca</span>, scale <span class='o'>=</span> <span class='kc'>TRUE</span><span class='o'>)</span>
<span class='c'># (Because `center = TRUE` by default, we don't have to include that.)</span></code></pre>
</div>
<p>Scaling is desirable here because as we saw above, the variables we use in our PCA are in different units (<code>mm</code> and <code>g</code>).</p>
<details>
<summary>
<b>More on scaling</b> (click here)
</summary>
<br>
<p>Because our variables are in different units, standard deviations for those variables may differ dramatically. This would lead the PCA to put more weight on variables with a higher standard deviation, which we don&rsquo;t want if those differences are merely a consequence of different units.</p>
<p>If we check the standard deviations in our dataset, we can indeed see large differences:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>map</span><span class='o'>(</span><span class='nv'>penguins_for_pca</span>, <span class='nv'>sd</span><span class='o'>)</span>
<span class='c'>#&gt; $bill_length_mm</span>
<span class='c'>#&gt; [1] 5.468668</span>
<span class='c'>#&gt; </span>
<span class='c'>#&gt; $bill_depth_mm</span>
<span class='c'>#&gt; [1] 1.969235</span>
<span class='c'>#&gt; </span>
<span class='c'>#&gt; $flipper_length_mm</span>
<span class='c'>#&gt; [1] 14.01577</span>
<span class='c'>#&gt; </span>
<span class='c'>#&gt; $body_mass_g</span>
<span class='c'>#&gt; [1] 805.2158</span></code></pre>
</div>
</details>
<br>
<hr>
<h2 id="4---exploring-the-output-i">4 - Exploring the output I</h2>
<p>Like with objects returned by the statistical tests we saw in the previous weeks, the object returned by <a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a> is not just a dataframe or even a regular list&hellip;</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/base/class.html'>class</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>)</span>
<span class='c'>#&gt; [1] "prcomp"</span></code></pre>
</div>
<p>&hellip; and trying to print the object to screen will only give you a summary of sorts:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca</span>
<span class='c'>#&gt; Standard deviations (1, .., p=4):</span>
<span class='c'>#&gt; [1] 1.6569115 0.8821095 0.6071594 0.3284579</span>
<span class='c'>#&gt; </span>
<span class='c'>#&gt; Rotation (n x k) = (4 x 4):</span>
<span class='c'>#&gt;                          PC1         PC2        PC3        PC4</span>
<span class='c'>#&gt; bill_length_mm     0.4537532 -0.60019490 -0.6424951  0.1451695</span>
<span class='c'>#&gt; bill_depth_mm     -0.3990472 -0.79616951  0.4258004 -0.1599044</span>
<span class='c'>#&gt; flipper_length_mm  0.5768250 -0.00578817  0.2360952 -0.7819837</span>
<span class='c'>#&gt; body_mass_g        0.5496747 -0.07646366  0.5917374  0.5846861</span></code></pre>
</div>
<p>Like we saw last week with <a href="https://rdrr.io/r/stats/aov.html" target="_blank" rel="noopener"><code>aov()</code></a>, we can get a more useful summary of the results with the <a href="https://rdrr.io/r/base/summary.html" target="_blank" rel="noopener"><code>summary()</code></a> function:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>)</span>
<span class='c'>#&gt; Importance of components:</span>
<span class='c'>#&gt;                           PC1    PC2     PC3     PC4</span>
<span class='c'>#&gt; Standard deviation     1.6569 0.8821 0.60716 0.32846</span>
<span class='c'>#&gt; Proportion of Variance 0.6863 0.1945 0.09216 0.02697</span>
<span class='c'>#&gt; Cumulative Proportion  0.6863 0.8809 0.97303 1.00000</span></code></pre>
</div>
<p>This shows us the &ldquo;importance&rdquo; of the 4 principal components that our PCA returned, i.e. the amount of variation they explain.</p>
<h4 id="seeing-all-elements-with-str">Seeing all elements with <code>str()</code></h4>
<p>These summaries are nice and all, but like we saw in previous weeks, they don&rsquo;t make it obvious where and how to access <em>all</em> the information contained in the object.</p>
<p>Running the <a href="https://rdrr.io/r/utils/str.html" target="_blank" rel="noopener"><code>str()</code></a> function is a good start for getting to the raw contents of the object, even though the information printed isn&rsquo;t easy to look at:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/utils/str.html'>str</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>)</span>
<span class='c'>#&gt; List of 5</span>
<span class='c'>#&gt;  $ sdev    : num [1:4] 1.657 0.882 0.607 0.328</span>
<span class='c'>#&gt;  $ rotation: num [1:4, 1:4] 0.454 -0.399 0.577 0.55 -0.6 ...</span>
<span class='c'>#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class='c'>#&gt;   .. ..$ : chr [1:4] "bill_length_mm" "bill_depth_mm" "flipper_length_mm" "body_mass_g"</span>
<span class='c'>#&gt;   .. ..$ : chr [1:4] "PC1" "PC2" "PC3" "PC4"</span>
<span class='c'>#&gt;  $ center  : Named num [1:4] 44 17.2 201 4207.1</span>
<span class='c'>#&gt;   ..- attr(*, "names")= chr [1:4] "bill_length_mm" "bill_depth_mm" "flipper_length_mm" "body_mass_g"</span>
<span class='c'>#&gt;  $ scale   : Named num [1:4] 5.47 1.97 14.02 805.22</span>
<span class='c'>#&gt;   ..- attr(*, "names")= chr [1:4] "bill_length_mm" "bill_depth_mm" "flipper_length_mm" "body_mass_g"</span>
<span class='c'>#&gt;  $ x       : num [1:333, 1:4] -1.85 -1.31 -1.37 -1.88 -1.92 ...</span>
<span class='c'>#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class='c'>#&gt;   .. ..$ : NULL</span>
<span class='c'>#&gt;   .. ..$ : chr [1:4] "PC1" "PC2" "PC3" "PC4"</span>
<span class='c'>#&gt;  - attr(*, "class")= chr "prcomp"</span></code></pre>
</div>
<p>In the first breakout room session, you&rsquo;ll explore the contents of our <code>pca</code> object a bit more.</p>
<br>
<hr>
<h2 id="breakout-rooms-i">Breakout Rooms I</h2>
<div class="puzzle">
<div>
<h3 id="exercise-1">Exercise 1</h3>
<p>If you didn&rsquo;t do so already, get set up for the remaining exercises. Either <a href="https://raw.githubusercontent.com/biodash/biodash.github.io/master/content/codeclub/S03E03_PCA/S03E03_PCA.R" target="_blank" rel="noopener">download this R script</a>, open it in RStudio, and run the code, or:</p>
<ul>
<li>
<p>Open a new R script in RStudio (<code>File</code> =&gt; <code>New File</code> =&gt; <code>R Script</code>)</p>
</li>
<li>
<p>Save the script, as something along the lines of <code>codeclub_S03E03_PCA.R</code></p>
</li>
<li>
<p>Copy the following code into the script and then run it in the R console:</p>
</li>
</ul>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'>## Install packages if needed</span>
<span class='c'>## (`require(glue)` returns FALSE if glue isn't installed; therefore,</span>
<span class='c'>##  these lines will only try to install packages that aren't already installed.)</span>
<span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='https://allisonhorst.github.io/palmerpenguins/'>palmerpenguins</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"palmerpenguins"</span><span class='o'>)</span>
<span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"tidyverse"</span><span class='o'>)</span>
<span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='https://broom.tidymodels.org/'>broom</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"broom"</span><span class='o'>)</span>
<span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='https://github.com/tidyverse/glue'>glue</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"glue"</span><span class='o'>)</span>
<span class='kr'>if</span> <span class='o'>(</span><span class='o'>!</span><span class='kr'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='o'>(</span><span class='nv'><a href='http://www.sthda.com/english/rpkgs/factoextra'>factoextra</a></span><span class='o'>)</span><span class='o'>)</span> <span class='nf'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='o'>(</span><span class='s'>"factoextra"</span><span class='o'>)</span>

<span class='c'>## Load the packages into your R session</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://allisonhorst.github.io/palmerpenguins/'>palmerpenguins</a></span><span class='o'>)</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='o'>)</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://broom.tidymodels.org/'>broom</a></span><span class='o'>)</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://github.com/tidyverse/glue'>glue</a></span><span class='o'>)</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='http://www.sthda.com/english/rpkgs/factoextra'>factoextra</a></span><span class='o'>)</span>

<span class='c'>## Prep the penguin data for the PCA</span>
<span class='nv'>penguins_noNA</span> <span class='o'>&lt;-</span> <span class='nf'>drop_na</span><span class='o'>(</span><span class='nv'>penguins</span><span class='o'>)</span>
<span class='nv'>penguins_for_pca</span> <span class='o'>&lt;-</span> <span class='nv'>penguins_noNA</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>select</span><span class='o'>(</span><span class='nv'>bill_length_mm</span>, <span class='nv'>bill_depth_mm</span>, <span class='nv'>flipper_length_mm</span>, <span class='nv'>body_mass_g</span><span class='o'>)</span>

<span class='c'>## Run the PCA</span>
<span class='nv'>pca</span> <span class='o'>&lt;-</span> <span class='nf'><a href='https://rdrr.io/r/stats/prcomp.html'>prcomp</a></span><span class='o'>(</span><span class='nv'>penguins_for_pca</span>, scale <span class='o'>=</span> <span class='kc'>TRUE</span><span class='o'>)</span></code></pre>
</div>
</div>
</div>
<br>
<div class="puzzle">
<div>
<h3 id="exercise-2">Exercise 2</h3>
<p>How can you access the different components in the <code>List of 5</code> that is summarized when running <a href="https://rdrr.io/r/utils/str.html" target="_blank" rel="noopener"><code>str(pca)</code></a>? For example, say you wanted to see the <code>rotation</code> element in its entirety, how could you do this?</p>
<details>
<summary>
<b>Hints</b> (click here)
</summary>
<br>
<p>The <strong><a href="https://rdrr.io/r/base/Extract.html" target="_blank" rel="noopener"><code>$</code></a></strong> (dollar sign) operator can be used to access the different elements (as implied by the dollar signs shown in front of the names of the elements).</p>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<br>
<p>To see the <code>rotation</code> element, type <strong><code>pca$rotation</code></strong>:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>rotation</span>
<span class='c'>#&gt;                          PC1         PC2        PC3        PC4</span>
<span class='c'>#&gt; bill_length_mm     0.4537532 -0.60019490 -0.6424951  0.1451695</span>
<span class='c'>#&gt; bill_depth_mm     -0.3990472 -0.79616951  0.4258004 -0.1599044</span>
<span class='c'>#&gt; flipper_length_mm  0.5768250 -0.00578817  0.2360952 -0.7819837</span>
<span class='c'>#&gt; body_mass_g        0.5496747 -0.07646366  0.5917374  0.5846861</span></code></pre>
</div>
</div>
</div>
<br>
<div class="puzzle">
<div>
<h3 id="exercise-3-bonus">Exercise 3 (bonus)</h3>
<p>Take a look at the contents of all five elements in the <code>pca</code> object. Do you have a (rough) understanding of what each represents?</p>
<details>
<summary>
<b>Hints</b> (click here)
</summary>
<br>
<p>Take a look at the <a href="#glossary">Glossary</a>.</p>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<br>
<p>All elements of the output are explained in the next section of this page.</p>
</details>
</div>
</div>
<br>
<hr>
<h2 id="5---exploring-the-output-ii">5 - Exploring the output II</h2>
<p>Let&rsquo;s take a quick look together at the three most important elements in the object returned by <a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a>, which we named <code>pca</code>:</p>
<ul>
<li>
<p><strong><code>pca$sdev</code></strong> is a vector of standard deviations associated with each principal component (PC), i.e. it is the <strong>amount of variation explained by each PC</strong>. We also saw this information when running <a href="https://rdrr.io/r/base/summary.html" target="_blank" rel="noopener"><code>summary(pca)</code></a> and we&rsquo;ll use it to create the <em>scree plot</em>.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>sdev</span>
  <span class='c'>#&gt; [1] 1.6569115 0.8821095 0.6071594 0.3284579</span></code></pre>
</div>
</li>
<li>
<p><strong><code>pca$x</code></strong> is the most-used part of the output: a matrix containing the <strong>scores (or coordinates)</strong> for each sample for each PC, used to create a <em>score plot</em> and part of the <em>biplot</em>.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>x</span><span class='o'>)</span>
  <span class='c'>#&gt;            PC1         PC2         PC3        PC4</span>
  <span class='c'>#&gt; [1,] -1.850808 -0.03202119  0.23454869  0.5276026</span>
  <span class='c'>#&gt; [2,] -1.314276  0.44286031  0.02742880  0.4011230</span>
  <span class='c'>#&gt; [3,] -1.374537  0.16098821 -0.18940423 -0.5278675</span>
  <span class='c'>#&gt; [4,] -1.882455  0.01233268  0.62792772 -0.4721826</span>
  <span class='c'>#&gt; [5,] -1.917096 -0.81636958  0.69999797 -0.1961213</span>
  <span class='c'>#&gt; [6,] -1.770356  0.36567266 -0.02841769  0.5046092</span></code></pre>
</div>
</li>
<li>
<p><strong><code>pca$rotation</code></strong> is a matrix that contains the <strong>loadings</strong> for each variable in each PC. These are the &ldquo;recipes&rdquo; for creating each PC, with <em>higher absolute values</em> indicating a larger effect of the variable on the PC. The <em>sign</em> (- or +) matters too: in PC1, larger values of <code>bill_depth_mm</code> lower the PC value, and vice versa for the other three variables. This matrix will be used in creating the <em>biplot</em>.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>rotation</span>
  <span class='c'>#&gt;                          PC1         PC2        PC3        PC4</span>
  <span class='c'>#&gt; bill_length_mm     0.4537532 -0.60019490 -0.6424951  0.1451695</span>
  <span class='c'>#&gt; bill_depth_mm     -0.3990472 -0.79616951  0.4258004 -0.1599044</span>
  <span class='c'>#&gt; flipper_length_mm  0.5768250 -0.00578817  0.2360952 -0.7819837</span>
  <span class='c'>#&gt; body_mass_g        0.5496747 -0.07646366  0.5917374  0.5846861</span></code></pre>
</div>
</li>
</ul>
<br>
<details>
<summary>
<b>...And the remaining two elements</b> (click here)
</summary>
<ul>
<li>
<p><strong><code>pca$center</code></strong> is a vector containing the means for each variable, which was subsequently used for centering the data (this would contain just <code>FALSE</code> if the data wasn&rsquo;t centered).</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>center</span>
  <span class='c'>#&gt;    bill_length_mm     bill_depth_mm flipper_length_mm       body_mass_g </span>
  <span class='c'>#&gt;          43.99279          17.16486         200.96697        4207.05706</span></code></pre>
</div>
</li>
<li>
<p><strong><code>pca$scale</code></strong> similarly is a vector containing the scaling constant for each variable (column) in the data, and would be <code>FALSE</code> if the data wasn&rsquo;t scaled.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>scale</span>
  <span class='c'>#&gt;    bill_length_mm     bill_depth_mm flipper_length_mm       body_mass_g </span>
  <span class='c'>#&gt;          5.468668          1.969235         14.015765        805.215802</span></code></pre>
</div>
</li>
</ul>
</details>
<br>
<hr>
<h2 id="6---scree-plot">6 - Scree plot</h2>
<p>A &ldquo;scree plot&rdquo;<sup id="fnref:3"><a href="#fn:3" class="footnote-ref" role="doc-noteref">3</a></sup> is a barplot that shows the <strong>amount of variation explained by each PC.</strong></p>
<p>We&rsquo;ll make a base R version of this plot (gasp!) because it is so quick to make, and we don&rsquo;t need this figure to be fancy:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-18-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>In this scree plot, we show the variance (i.e. the <em>eigenvalue</em>) associated with each PC (these are the square roots of the standard deviations in <code>pca$sdev</code>.)</p>
<div class="alert alert-note">
<div>
<h4 id="interpretation">Interpretation</h4>
<p>This gives us a quick visual overview of the importance of the PCs: PC1 is <em>by far</em> the most important, and PC4 doesn&rsquo;t do much at all. (PCs are always ordered by the amount of variation they explain, with PC1 explaining most.)</p>
</div>
</div>
<br>
<hr>
<h2 id="7---score-classic-pca-plot">7 - Score (classic PCA) plot</h2>
<p>A &ldquo;score plot&rdquo; shows <strong>the scores (coordinates) for each sample along two PCs</strong>, typically the first two.</p>
<p>We&rsquo;re going to need a dataframe to plot. But if we were to <a href="https://generics.r-lib.org/reference/tidy.html" target="_blank" rel="noopener"><code>broom::tidy()</code></a> the scores matrix (<code>pca$x</code>), akin to what we&rsquo;ve done with t-test and ANOVA output in previous weeks, we would get a dataframe with all PCs in one column that wouldn&rsquo;t be that easy to plot.</p>
<p>So in this case, we&rsquo;ll just manipulate <code>pca$x</code> ourselves &ndash; in particular, we want to add the source <code>penguins_noNA</code> dataframe back to it, which will allow us to color the points by, say, <code>species</code>.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'>## Column-bind (= put side-by-side) the scores and the source dataframe</span>
<span class='nv'>pca_scores</span> <span class='o'>&lt;-</span> <span class='nf'>bind_cols</span><span class='o'>(</span><span class='nf'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>x</span><span class='o'>)</span>, <span class='nv'>penguins_noNA</span><span class='o'>)</span>

<span class='nf'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='o'>(</span><span class='nv'>pca_scores</span><span class='o'>)</span>
<span class='c'>#&gt;         PC1         PC2         PC3        PC4 species    island bill_length_mm</span>
<span class='c'>#&gt; 1 -1.850808 -0.03202119  0.23454869  0.5276026  Adelie Torgersen           39.1</span>
<span class='c'>#&gt; 2 -1.314276  0.44286031  0.02742880  0.4011230  Adelie Torgersen           39.5</span>
<span class='c'>#&gt; 3 -1.374537  0.16098821 -0.18940423 -0.5278675  Adelie Torgersen           40.3</span>
<span class='c'>#&gt; 4 -1.882455  0.01233268  0.62792772 -0.4721826  Adelie Torgersen           36.7</span>
<span class='c'>#&gt; 5 -1.917096 -0.81636958  0.69999797 -0.1961213  Adelie Torgersen           39.3</span>
<span class='c'>#&gt; 6 -1.770356  0.36567266 -0.02841769  0.5046092  Adelie Torgersen           38.9</span>
<span class='c'>#&gt;   bill_depth_mm flipper_length_mm body_mass_g    sex year</span>
<span class='c'>#&gt; 1          18.7               181        3750   male 2007</span>
<span class='c'>#&gt; 2          17.4               186        3800 female 2007</span>
<span class='c'>#&gt; 3          18.0               195        3250 female 2007</span>
<span class='c'>#&gt; 4          19.3               193        3450 female 2007</span>
<span class='c'>#&gt; 5          20.6               190        3650   male 2007</span>
<span class='c'>#&gt; 6          17.8               181        3625 female 2007</span></code></pre>
</div>
<p>Now we&rsquo;re ready to create the plot:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>score_plot</span> <span class='o'>&lt;-</span> <span class='nf'>ggplot</span><span class='o'>(</span><span class='nv'>pca_scores</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>PC1</span>, y <span class='o'>=</span> <span class='nv'>PC2</span>, color <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span><span class='o'>)</span>

<span class='nv'>score_plot</span>
</code></pre>
<p><img src="figs/unnamed-chunk-20-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<div class="alert alert-note">
<div>
<h4 id="interpretation-1">Interpretation</h4>
<p>Across these four measurements, Gentoo Penguins can be very clearly distinguished from the other two species, whereas among Adelie and Chinstrap Penguins, there are average differences but they are not fully separable.</p>
</div>
</div>
<h3 id="a-better-aspect-ratio">A better aspect ratio</h3>
<p>One way to improve our plot is to set the aspect ratio (the proportional relationship between the height and the width) according to the relative percentages of variation explained by the two plotted PCs: because PC1 on the x-axis explains more variation, we want the plot to be wide.</p>
<p>To get the percentages in a dataframe, now we <em>will</em> use the <a href="https://generics.r-lib.org/reference/tidy.html" target="_blank" rel="noopener"><code>tidy()</code></a> function. But because the output of <a href="https://rdrr.io/r/stats/prcomp.html" target="_blank" rel="noopener"><code>prcomp()</code></a> contains multiple elements, we&rsquo;ll have to point <a href="https://generics.r-lib.org/reference/tidy.html" target="_blank" rel="noopener"><code>tidy()</code></a> to the <code>$sdev</code> element using the <code>matrix</code> argument (see the <a href="%28https://broom.tidymodels.org/reference/tidy.prcomp.html%29">docs</a>):</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca_eigen</span> <span class='o'>&lt;-</span> <span class='nf'><a href='https://generics.r-lib.org/reference/tidy.html'>tidy</a></span><span class='o'>(</span><span class='nv'>pca</span>, matrix <span class='o'>=</span> <span class='s'>"eigenvalues"</span><span class='o'>)</span>
<span class='nv'>pca_eigen</span>
<span class='c'>#&gt; <span style='color: #555555;'># A tibble: 4 × 4</span></span>
<span class='c'>#&gt;      PC std.dev percent cumulative</span>
<span class='c'>#&gt;   <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>   <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>   <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>      <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span></span>
<span class='c'>#&gt; <span style='color: #555555;'>1</span>     1   1.66   0.686       0.686</span>
<span class='c'>#&gt; <span style='color: #555555;'>2</span>     2   0.882  0.195       0.881</span>
<span class='c'>#&gt; <span style='color: #555555;'>3</span>     3   0.607  0.092<span style='text-decoration: underline;'>2</span>      0.973</span>
<span class='c'>#&gt; <span style='color: #555555;'>4</span>     4   0.328  0.027<span style='text-decoration: underline;'>0</span>      1</span></code></pre>
</div>
<p>Now, we&rsquo;ll store the percentages explained by the first two PCs (rounded to one decimal):</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># (Note: pca_eigen$percent contains proportions, not percentages...)</span>
<span class='nv'>PC1_percent</span> <span class='o'>&lt;-</span> <span class='nf'><a href='https://rdrr.io/r/base/Round.html'>round</a></span><span class='o'>(</span><span class='nv'>pca_eigen</span><span class='o'>$</span><span class='nv'>percent</span><span class='o'>[</span><span class='m'>1</span><span class='o'>]</span> <span class='o'>*</span> <span class='m'>100</span>, <span class='m'>1</span><span class='o'>)</span>
<span class='nv'>PC2_percent</span> <span class='o'>&lt;-</span> <span class='nf'><a href='https://rdrr.io/r/base/Round.html'>round</a></span><span class='o'>(</span><span class='nv'>pca_eigen</span><span class='o'>$</span><span class='nv'>percent</span><span class='o'>[</span><span class='m'>2</span><span class='o'>]</span> <span class='o'>*</span> <span class='m'>100</span>, <span class='m'>1</span><span class='o'>)</span>

<span class='nv'>PC1_percent</span>
<span class='c'>#&gt; [1] 68.6</span>
<span class='nv'>PC2_percent</span>
<span class='c'>#&gt; [1] 19.5</span></code></pre>
</div>
<p>Finally, we can modify the aspect ratio, which is expressed as <code>height / width</code> &ndash; and we&rsquo;ll also move the legend to the top, and add the percentages to the axis titles:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>score_plot</span> <span class='o'>&lt;-</span> <span class='nv'>score_plot</span> <span class='o'>+</span>
  <span class='nf'>theme</span><span class='o'>(</span>aspect.ratio <span class='o'>=</span> <span class='nv'>PC2_percent</span> <span class='o'>/</span> <span class='nv'>PC1_percent</span>,
        legend.position <span class='o'>=</span> <span class='s'>"top"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>labs</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nf'><a href='https://glue.tidyverse.org/reference/glue.html'>glue</a></span><span class='o'>(</span><span class='s'>"PC1 (&#123;PC1_percent&#125;%)"</span><span class='o'>)</span>,
       y <span class='o'>=</span> <span class='nf'><a href='https://glue.tidyverse.org/reference/glue.html'>glue</a></span><span class='o'>(</span><span class='s'>"PC2 (&#123;PC2_percent&#125;%)"</span><span class='o'>)</span><span class='o'>)</span>

<span class='nv'>score_plot</span>
</code></pre>
<p><img src="figs/unnamed-chunk-23-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<hr>
<h2 id="8---biplot">8 - Biplot</h2>
<p>A &ldquo;biplot&rdquo; shows the <strong>scores of samples for two PCs <em>and</em> the loadings for the original variables</strong> along the two PCs.</p>
<p>Because biplots are more complicated to make &ldquo;from scratch&rdquo; using <em>ggplot2</em>, we will turn to the package <em>factoextra</em>, which has a convenient function for making biplots, <a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html" target="_blank" rel="noopener"><code>fviz_pca()</code></a>:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/pkg/factoextra/man/fviz_pca.html'>fviz_pca</a></span><span class='o'>(</span><span class='nv'>pca</span>,
         label <span class='o'>=</span> <span class='s'>"var"</span>,                       <span class='c'># Show labels for variables only</span>
         habillage <span class='o'>=</span> <span class='nv'>penguins_noNA</span><span class='o'>$</span><span class='nv'>species</span><span class='o'>)</span> <span class='o'>+</span> <span class='c'># color by / shape by</span>
  <span class='nf'>theme</span><span class='o'>(</span>legend.position <span class='o'>=</span> <span class='s'>"top"</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-24-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>While this plot can certainly be improved upon, <em>biplots are by their nature a little unwieldy</em>.</p>
<div class="alert alert-note">
<div>
<h4 id="interpretation-2">Interpretation</h4>
<p>Biplots can be especially useful when you have a modest number of original variables, like here. Some information we can glean from this particular biplot:</p>
<ul>
<li>
<p>Flipper length and body mass are highly correlated among individuals, even across species. So flipper length relative to body mass is similar across species.</p>
</li>
<li>
<p>Gentoo penguins are larger and with narrower bills than the other two species.</p>
</li>
</ul>
</div>
</div>
<div class="alert alert-note">
<div>
<p>While we made a scree plot with base R and a score plot with &ldquo;base <em>ggplot2</em>&rdquo;, there are also <em>factoextra</em> functions for these and for other PCA plots:</p>
<ul>
<li><a href="https://rdrr.io/pkg/factoextra/man/eigenvalue.html" target="_blank" rel="noopener"><code>fviz_eig()</code></a> &ndash; scree plots</li>
<li><a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html" target="_blank" rel="noopener"><code>fviz_pca_ind()</code></a> &ndash; score plots</li>
<li><a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html" target="_blank" rel="noopener"><code>fviz_pca_var()</code></a> &ndash; &ldquo;correlation circles&rdquo;; plots showing loadings only (<a href="https://journals.plos.org/ploscompbiol/article/figure?id=10.1371/journal.pcbi.1006907.g003" target="_blank" rel="noopener">example</a>).</li>
<li><a href="https://rdrr.io/pkg/factoextra/man/fviz_contrib.html" target="_blank" rel="noopener"><code>fviz_contrib()</code></a> &ndash; a barplot with the contribution of variables to 1 PC (<a href="https://journals.plos.org/ploscompbiol/article/figure?id=10.1371/journal.pcbi.1006907.g003" target="_blank" rel="noopener">example</a>).</li>
</ul>
</div>
</div>
<details>
<summary>
<b>Or, to create a biplot from scratch...</b> (click here)
</summary>
<br>
<p>First, let&rsquo;s save the loadings in a dataframe:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>pca_loadings</span> <span class='o'>&lt;-</span> <span class='nf'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>rotation</span><span class='o'>)</span> <span class='o'>%&gt;%</span>
  <span class='nf'>rownames_to_column</span><span class='o'>(</span><span class='s'>"var"</span><span class='o'>)</span></code></pre>
</div>
<p>Next, we start with the score plot object <code>score_plot</code> we created above.</p>
<p>What we need to add are the variable loading, which we&rsquo;ll do with <code>geom_segment()</code> to draw arrows, and <code>geom_text()</code> to add text labels near the tips of the arrows:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'>## To make the arrows longer (all by the same amount),</span>
<span class='c'>## just to improve the visualization, we use a multiplication factor:</span>
<span class='nv'>mult</span> <span class='o'>&lt;-</span> <span class='m'>2.5</span>

<span class='nv'>score_plot</span> <span class='o'>+</span>
  <span class='c'>## geom_segment draws lines</span>
  <span class='nf'>geom_segment</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>pca_loadings</span>,
               <span class='c'>## The lines should start from the origin:</span>
               <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='m'>0</span>, y <span class='o'>=</span> <span class='m'>0</span>, xend <span class='o'>=</span> <span class='nv'>PC1</span> <span class='o'>*</span> <span class='nv'>mult</span>, yend <span class='o'>=</span> <span class='nv'>PC2</span> <span class='o'>*</span> <span class='nv'>mult</span><span class='o'>)</span>,
               <span class='c'>## We turn the line into an arrow:</span>
               arrow <span class='o'>=</span> <span class='nf'>arrow</span><span class='o'>(</span><span class='o'>)</span>,
               <span class='c'>## A gray-tone might work better than black:</span>
               color <span class='o'>=</span> <span class='s'>"grey40"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_text</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>pca_loadings</span>,
            <span class='c'>## The text labels go at the end of the arrows:</span>
            <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>PC1</span> <span class='o'>*</span> <span class='nv'>mult</span>, y <span class='o'>=</span> <span class='nv'>PC2</span> <span class='o'>*</span> <span class='nv'>mult</span>, label <span class='o'>=</span> <span class='nv'>var</span><span class='o'>)</span>,
            <span class='c'>## We left-align (hjust = 0) and lower (vjust = 1) the labels</span>
            hjust <span class='o'>=</span> <span class='m'>0</span>, vjust <span class='o'>=</span> <span class='m'>1</span>,
            <span class='c'>## Again, we use a gray color:</span>
            color <span class='o'>=</span> <span class='s'>"grey40"</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-26-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
<br>
<hr>
<h2 id="breakout-rooms-ii">Breakout Rooms II</h2>
<div class="puzzle">
<div>
<h3 id="exercise-4">Exercise 4</h3>
<p>Above, we plotted the scores for the first two PCs (PC1 and PC2) in our score plot and biplot. Now, create a biplot with another combination of two PCs.</p>
<p>Take a look at the help for the <a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html" target="_blank" rel="noopener"><code>fviz_pca()</code></a> function by typing<br>
<a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html" target="_blank" rel="noopener"><code>?fviz_pca</code></a> to find out how you might be able to plot different PCs.</p>
<details>
<summary>
<b>Hints</b> (click here)
</summary>
<br>
<ul>
<li>
<p>The <code>axes</code> argument to <a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html" target="_blank" rel="noopener"><code>fviz_pca()</code></a> controls which axes will be plotted; this argument accepts a vector of two numbers.</p>
</li>
<li>
<p>Do you think it would be worth plotting PC4, which explains &lt;3% of the variation? Would plotting PC3 with one of the PCs we already plotted be informative?</p>
</li>
</ul>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<br>
<p>To plot PC1 &amp; PC3 (which may be a better choice than including PC4 because it explains so little variation):</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/pkg/factoextra/man/fviz_pca.html'>fviz_pca</a></span><span class='o'>(</span><span class='nv'>pca</span>,
         axes <span class='o'>=</span> <span class='nf'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='o'>(</span><span class='m'>1</span>, <span class='m'>3</span><span class='o'>)</span>,
         label <span class='o'>=</span> <span class='s'>"var"</span>,
         habillage <span class='o'>=</span> <span class='nv'>penguins_noNA</span><span class='o'>$</span><span class='nv'>species</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme</span><span class='o'>(</span>legend.position <span class='o'>=</span> <span class='s'>"top"</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-27-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Behold, now we can distinguish much better between Adelie and Chinstrap Penguins!</p>
</details>
</div>
</div>
<br>
<div class="puzzle">
<div>
<h3 id="exercise-5">Exercise 5</h3>
<p>Run the PCA for just one of the three penguin species.</p>
<p>Then, make a biplot of the results, in which you color the points by something else than <code>species</code>, e.g. by <code>sex</code>. (If you want, also make a scree plot and/or a score plot.)</p>
<details>
<summary>
<b>Hints</b> (click here)
</summary>
<br>
<ul>
<li>
<p>Use the <em>dplyr</em> function <a href="https://rdrr.io/r/stats/filter.html" target="_blank" rel="noopener"><code>filter()</code></a> on the <code>penguins_noNA</code> object to select rows corresponding to one penguin species.</p>
</li>
<li>
<p>After that, the code will be nearly identical to that used before; just make sure to refer to the correct objects if you copy-and-paste code.</p>
</li>
</ul>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<br>
<p>This example solution runs a PCA for <em>Gentoo Penguins only</em>.</p>
<p>First, select rows corresponding to our focal penguin species, and run the PCA:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'>## (Save this object rather than using one pipeline,</span>
<span class='c'>## because you'll need it color the biplot by a factor)</span>
<span class='nv'>onepenguin_noNA</span> <span class='o'>&lt;-</span> <span class='nv'>penguins_noNA</span> <span class='o'>%&gt;%</span> <span class='nf'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='o'>(</span><span class='nv'>species</span> <span class='o'>==</span> <span class='s'>"Gentoo"</span><span class='o'>)</span>

<span class='nv'>pca</span> <span class='o'>&lt;-</span> <span class='nv'>onepenguin_noNA</span> <span class='o'>%&gt;%</span>
  <span class='nf'>select</span><span class='o'>(</span><span class='nv'>bill_length_mm</span>, <span class='nv'>bill_depth_mm</span>, <span class='nv'>flipper_length_mm</span>, <span class='nv'>body_mass_g</span><span class='o'>)</span> <span class='o'>%&gt;%</span> 
  <span class='nf'><a href='https://rdrr.io/r/stats/prcomp.html'>prcomp</a></span><span class='o'>(</span>scale <span class='o'>=</span> <span class='kc'>TRUE</span><span class='o'>)</span></code></pre>
</div>
<p>Next, we create the biplot:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/pkg/factoextra/man/fviz_pca.html'>fviz_pca</a></span><span class='o'>(</span><span class='nv'>pca</span>,
         label <span class='o'>=</span> <span class='s'>"var"</span>,
         habillage <span class='o'>=</span> <span class='nv'>onepenguin_noNA</span><span class='o'>$</span><span class='nv'>sex</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme</span><span class='o'>(</span>legend.position <span class='o'>=</span> <span class='s'>"top"</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-29-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>To create a scree plot:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-30-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Or a scree plot with <a href="https://rdrr.io/pkg/factoextra/man/eigenvalue.html" target="_blank" rel="noopener"><code>fviz_eig()</code></a>:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/pkg/factoextra/man/eigenvalue.html'>fviz_eig</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-31-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>To create a quick score plot (no aspect ratio manipulation):</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>bind_cols</span><span class='o'>(</span><span class='nf'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='o'>(</span><span class='nv'>pca</span><span class='o'>$</span><span class='nv'>x</span><span class='o'>)</span>, <span class='nv'>onepenguin_noNA</span><span class='o'>)</span> <span class='o'>%&gt;%</span>
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>PC1</span>, y <span class='o'>=</span> <span class='nv'>PC2</span>, color <span class='o'>=</span> <span class='nv'>sex</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_classic</span><span class='o'>(</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-32-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Or a score plot with <a href="https://rdrr.io/pkg/factoextra/man/fviz_pca.html" target="_blank" rel="noopener"><code>fviz_pca_ind()</code></a>:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'><a href='https://rdrr.io/pkg/factoextra/man/fviz_pca.html'>fviz_pca_ind</a></span><span class='o'>(</span><span class='nv'>pca</span>, geom <span class='o'>=</span> <span class='s'>"point"</span>, habillage <span class='o'>=</span> <span class='nv'>onepenguin_noNA</span><span class='o'>$</span><span class='nv'>sex</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-33-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
</div>
<br>
<div class="puzzle">
<div>
<h3 id="exercise-6-bonus">Exercise 6 (bonus)</h3>
<p>Make a scree plot of our original PCA results with <em>ggplot2</em> instead of base R.</p>
<details>
<summary>
<b>Hints</b> (click here)
</summary>
<br>
<ul>
<li>
<p>Use the <code>pca_eigen</code> dataframe that we created above for plotting, and use the geom <code>geom_col()</code>.</p>
</li>
<li>
<p>Think about what exactly you want to plot on the y-axis. The variance, like in the base R scree plot? Or the proportion/percentage of the variance explained?</p>
</li>
</ul>
</details>
<details>
<summary>
<b>Solution</b> (click here)
</summary>
<br>
<p>There are a couple of different things that could reasonably be put on the y-axis, but perhaps the clearest option is to put the proportion or percentage of variation (=variance) explained, like below:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span><span class='nv'>pca_eigen</span>, <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>PC</span>, y <span class='o'>=</span> <span class='nv'>percent</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_col</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>labs</span><span class='o'>(</span>y <span class='o'>=</span> <span class='s'>"Proportion of the variation explained"</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>theme_minimal</span><span class='o'>(</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-34-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>(Note once again that the column in <code>pca_eigen</code> is called <code>percent</code>, but it actually contains proportions.)</p>
</div>
</div>
<br>
<hr>
<h2 id="further-watching--reading">Further watching &amp; reading</h2>
<ul>
<li>&ldquo;StatQuest&rdquo; videos on:
<ul>
<li><a href="https://www.youtube.com/watch?v=FgakZw6K1QQ" target="_blank" rel="noopener">PCA (22 minutes)</a></li>
<li><a href="https://www.youtube.com/watch?v=oRvgq966yZg" target="_blank" rel="noopener">PCA follow-up: practical tips (8 minutes)</a></li>
<li><a href="https://www.youtube.com/watch?v=GEn-_dAyYME" target="_blank" rel="noopener">MDS and PCoA</a></li>
</ul>
</li>
<li><a href="http://web.stanford.edu/class/bios221/book/Chap-Multivariate.html" target="_blank" rel="noopener">Chapter on Multivariate Analysis from the book &ldquo;Modern Statistics for Modern Biology&rdquo;</a></li>
<li><a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006907" target="_blank" rel="noopener">Nguyen &amp; Holmes 2019: &ldquo;Ten quick tips for effective dimensionality reduction&rdquo;</a></li>
</ul>
<br>
<section class="footnotes" role="doc-endnotes">
<hr>
<ol>
<li id="fn:1" role="doc-endnote">
<p>Crawley 2012 &ndash; &ldquo;The R Book&rdquo; &ndash; <a href="https://www.cs.upc.edu/~robert/teaching/estadistica/TheRBook.pdf" target="_blank" rel="noopener">pdf</a> <a href="#fnref:1" class="footnote-backref" role="doc-backlink">&#x21a9;&#xfe0e;</a></p>
</li>
<li id="fn:2" role="doc-endnote">
<p>However, high-throughput sequencing results such as gene counts do need to be normalized by sample sequencing depth (&ldquo;library size&rdquo;) and subjected to a variance stabilizing normalization. <a href="#fnref:2" class="footnote-backref" role="doc-backlink">&#x21a9;&#xfe0e;</a></p>
</li>
<li id="fn:3" role="doc-endnote">
<p>As &ldquo;The R Book&rdquo; (Crawley 2012) explains: &ldquo;<em>This is called a scree plot in PCA because it is supposed to look like a cliff face on a mountainside (on the left), with a scree slope below it (the tail on the right).</em>&rdquo; <a href="#fnref:3" class="footnote-backref" role="doc-backlink">&#x21a9;&#xfe0e;</a></p>
</li>
</ol>
</section>

                    </div>

                    <br>

                    








<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://biodash.github.io/codeclub/s03e03_pca/&amp;text=S03E03:%20Principal%20Component%20Analysis%20%28PCA%29" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://biodash.github.io/codeclub/s03e03_pca/&amp;t=S03E03:%20Principal%20Component%20Analysis%20%28PCA%29" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=S03E03:%20Principal%20Component%20Analysis%20%28PCA%29&amp;body=https://biodash.github.io/codeclub/s03e03_pca/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://biodash.github.io/codeclub/s03e03_pca/&amp;title=S03E03:%20Principal%20Component%20Analysis%20%28PCA%29" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="whatsapp://send?text=S03E03:%20Principal%20Component%20Analysis%20%28PCA%29%20https://biodash.github.io/codeclub/s03e03_pca/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://biodash.github.io/codeclub/s03e03_pca/&amp;title=S03E03:%20Principal%20Component%20Analysis%20%28PCA%29" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>











  
  
    



  
  
  
  
  
  <div class="media author-card content-widget-hr">
    
      
      <a href="/authors/admin/"><img class="avatar mr-3 avatar-circle" src="/authors/admin/avatar_hu6d2513500ad4d038e15c8975d9d8565d_488180_270x270_fill_q90_lanczos_center.jpg" alt="Jelmer Poelstra"></a>
    

    <div class="media-body">
      <h5 class="card-title"><a href="/authors/admin/">Jelmer Poelstra</a></h5>
      <h6 class="card-subtitle">Bioinformatician at MCIC</h6>
      
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
    <li>
      <a href="mailto:poelstra.1@osu.edu" >
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  
    
    
    
    
    
    
    
      
    
    <li>
      <a href="https://scholar.google.com/citations?user=h_a6S0cAAAAJ" target="_blank" rel="noopener">
        <i class="ai ai-google-scholar"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/jelmerp" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>


  
















                    <br>

                    <hr class="separator">

                    <div class="clearfix">
                        
                        
                        <span class="previous-link">
                            <a href="https://biodash.github.io/codeclub/s03e02_anova/">← S03E02: ANOVA</a>
                        </span>
                        
                        
                        
                        
                        <span class="next-link">
                            <a href="https://biodash.github.io/codeclub/s03e04_anova2/">S03E04: ANOVA 2 →</a>
                        </span>
                        
                        
                    </div>

                    <hr class="separator">

                </div>

            </article>

            
        </main>
    </div>
</div>

  </div>

  <div class="page-footer">
    
    
    <div class="container">
      <footer class="site-footer">
    
    <p class="powered-by">
      <a href="/license/">LICENSE: CC-BY-SA <br> <i class="fab fa-creative-commons fa-2x"></i><i class="fab fa-creative-commons-by fa-2x"></i><i class="fab fa-creative-commons-sa fa-2x"></i> </a><br>
    </p>
    
</footer>
    </div>
    
  </div>

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/instant.page/5.1.0/instantpage.min.js" integrity="sha512-1+qUtKoh9XZW7j+6LhRMAyOrgSQKenQ4mluTR+cvxXjP1Z54RxZuzstR/H9kgPXQsVB8IW7DMDFUJpzLjvhGSQ==" crossorigin="anonymous"></script>

      
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      

      
      

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/highlight.min.js" integrity="sha512-TDKKr+IvoqZnPzc3l35hdjpHD0m+b2EC2SrLEgKDRWpxf2rFCxemkgvJ5kfU48ip+Y+m2XVKyOCD85ybtlZDmw==" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/languages/r.min.js"></script>
        
      

    

    
    
    
      <script async defer src="https://maps.googleapis.com/maps/api/js?key="></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/gmaps.js/0.4.25/gmaps.min.js" integrity="sha256-7vjlAeb8OaTrCXZkCNun9djzuB2owUsaO72kXaFDBJs=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
      
      <script id="search-hit-fuse-template" type="text/x-template">
        <div class="search-hit" id="summary-{{key}}">
          <div class="search-hit-content">
            <div class="search-hit-name">
              <a href="{{relpermalink}}">{{title}}</a>
              <div class="article-metadata search-hit-type">{{type}}</div>
              <p class="search-hit-description">{{snippet}}</p>
            </div>
          </div>
        </div>
      </script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
    

    
    <script src="/js/bootstrap.bundle.min.6aed84840afc03ab4d5750157f69c120.js"></script>

    
    
      
      
      
      
      
      
      
    

    
    
    
    
    
    
    
    
      
      
    
    
      
    
    <script src="/en/js/wowchemy.min.2f16af175a3ec51b41b36bb4934b1cf7.js"></script>

    






</body>
</html>
