<!DOCTYPE html><html lang="en-us" >

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  
  
  
    <meta name="generator" content="Wowchemy 5.0.0-beta.3 for Hugo" />
  

  

  
  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Jelmer Poelstra" />

  
  
  
    
  
  <meta name="description" content="In this session of Code Club, we&#39;ll look at how to visualize data in R using the tidyverse package **ggplot2**." />

  
  <link rel="alternate" hreflang="en-us" href="https://biodash.github.io/codeclub/s02e06_ggplot2/" />

  







  




  
  
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
  

  
  
  
    <meta name="theme-color" content="#3f51b5" />
  

  
  
    
    <script src="/js/mathjax-config.js"></script>
  

  
  
  
  
    
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" integrity="sha256-FMvZuGapsJLjouA6k7Eo2lusoAX9i0ShlWFG6qt7SLc=" crossorigin="anonymous">

    
    

    
    
    
      
    
    
      
      
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-light" media="print" onload="this.media='all'">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/styles/github.min.css" crossorigin="anonymous" title="hl-dark" media="print" onload="this.media='all'" disabled>
      
    

    
    
    

    

    
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      
      

      
      

      
    
      
      

      
      

      
    
  

  
  
  
    
      
      <link rel="preload" as="style" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap">
      <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media="print" onload="this.media='all'">
    
  

  
  
  
  
  
  <link rel="stylesheet" href="/css/wowchemy.80f9a276c6100914c616a06e45ba8d09.css" />

  




<script async src="https://www.googletagmanager.com/gtag/js?id=G-9F9CQ7P0DX"></script>
<script>
  window.dataLayer = window.dataLayer || [];

  function gtag() {
      dataLayer.push(arguments);
  }

  function trackOutboundLink(url, target) {
    gtag('event', 'click', {
         'event_category': 'outbound',
         'event_label': url,
         'transport_type': 'beacon',
         'event_callback': function () {
           if (target !== '_blank') {
             document.location = url;
           }
         }
    });
    console.debug("Outbound link clicked: " + url);
  }

  function onClickCallback(event) {
    if ((event.target.tagName !== 'A') || (event.target.host === window.location.host)) {
      return;
    }
    trackOutboundLink(event.target, event.target.getAttribute('target'));  
  }

  gtag('js', new Date());
  gtag('config', 'G-9F9CQ7P0DX', {});
  gtag('set', {'cookie_flags': 'SameSite=None;Secure'});

  
  document.addEventListener('click', onClickCallback, false);
</script>


  

  

  




  
  
  

  

  
    <link rel="manifest" href="/index.webmanifest" />
  

  <link rel="icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_32x32_fill_lanczos_center_2.png" />
  <link rel="apple-touch-icon" type="image/png" href="/media/icon_hucc2cf774636a7e386557c89fa75ff618_191491_180x180_fill_lanczos_center_2.png" />

  <link rel="canonical" href="https://biodash.github.io/codeclub/s02e06_ggplot2/" />

  
  
  
  
  
  
  
  
    
  
  
  <meta property="twitter:card" content="summary_large_image" />
  
  <meta property="og:site_name" content="BioDASH" />
  <meta property="og:url" content="https://biodash.github.io/codeclub/s02e06_ggplot2/" />
  <meta property="og:title" content="Code Club S02E06: Visualizing Data | BioDASH" />
  <meta property="og:description" content="In this session of Code Club, we&#39;ll look at how to visualize data in R using the tidyverse package **ggplot2**." /><meta property="og:image" content="https://biodash.github.io/codeclub/s02e06_ggplot2/featured.jpg" />
    <meta property="twitter:image" content="https://biodash.github.io/codeclub/s02e06_ggplot2/featured.jpg" /><meta property="og:locale" content="en-us" />
  
    
      <meta
        property="article:published_time"
        content="2021-09-28T00:00:00&#43;00:00"
      />
    
    <meta property="article:modified_time" content="2021-09-30T10:52:41-04:00">
  

  



  

  

  <link rel='stylesheet' href='/css/highlight-light.css' title='hl-light'>
<link rel='stylesheet' href='/css/highlight-dark.css' title='hl-dark' disabled>



  <title>Code Club S02E06: Visualizing Data | BioDASH</title>
</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper   " data-wc-page-id="11e737a4fdee25a356dbb58f23d00829" >

  
  
  
  
  
  
  
  
  <script src="/js/wowchemy-init.min.7a224883d792570fd34d30c7134eace3.js"></script>

  


<aside class="search-modal" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#" aria-label="Close"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control"
        aria-label="Search...">
        
      </div>

      
      

      

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>



  <div class="page-header">
    











  


<nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
  <div class="container-xl">

    
    <div class="d-none d-lg-inline-flex">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    <button type="button" class="navbar-toggler" data-toggle="collapse"
            data-target="#navbar-content" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span><i class="fas fa-bars"></i></span>
    </button>
    

    
    <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
      <a class="navbar-brand" href="/"><img src="/media/logo_hu6407506e646a281542ecb299640680ce_37988_0x70_resize_lanczos_2.png" alt="BioDASH"></a>
    </div>
    

    
    
    <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

      
      <ul class="navbar-nav d-md-inline-flex">
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Code Club</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/codeclub-schedule/"><span>- <b>Schedule</b></span></a>
            
              <a class="dropdown-item" href="/codeclub-schedule/#previous-semesters"><span>- Past Sessions</span></a>
            
              <a class="dropdown-item" href="/codeclub-setup/"><span>- Computer Setup</span></a>
            
              <a class="dropdown-item" href="/"><span>------------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-suggest/"><span>- Suggest a Topic</span></a>
            
              <a class="dropdown-item" href="/"><span>-----------------------</span></a>
            
              <a class="dropdown-item" href="/codeclub-about/"><span>- <b>About Code Club & Join Us!</b></span></a>
            
          </div>
        </li>

        
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true"><span>Tutorials & Resources</span><span class="caret"></span>
          </a>
          <div class="dropdown-menu">
            
              <a class="dropdown-item" href="/events/2023-02_aws/"><span>AWS Workshop (Feb '23)</span></a>
            
              <a class="dropdown-item" href="/"><span>---------------------</span></a>
            
              <a class="dropdown-item" href="/tutorials/r-resources-tips/"><span>- R Resources and Tips</span></a>
            
              <a class="dropdown-item" href="/"><span>--------------------</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/2020-12-microbiomics-workshop/"><span>- Metabarcoding workshop (Fall '20)</span></a>
            
              <a class="dropdown-item" href="/tutorials/2021-01_rnaseq/"><span>- RNA-seq intro meetings (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/pracs-sp21/"><span>- Computing Course (Spring '21)</span></a>
            
              <a class="dropdown-item" href="https://mcic-osu.github.io/cl-workshop-22/"><span>- Command-line, shell scripting and OSC Workshop (Summer '22)</span></a>
            
              <a class="dropdown-item" href="/tutorials/2023-04_aggenomics/presentation.html"><span>- Ag Genomics Lecture: Command-line computing (Apr '23)</span></a>
            
          </div>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/support/"><span>Bioinformatics Support</span></a>
        </li>

        
        

        

        
        
        
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/about/"><span>About</span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="https://github.com/biodash/biodash.github.io" target="_blank" rel="noopener"><i class='fab fa-github'></i><span></span></a>
        </li>

        
        

        

        
        
        
          
            
          
        

        

        
        
        
        

        
          
            
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/index.xml" target="_blank" rel="noopener"><i class='fas fa-rss'></i><span></span></a>
        </li>

        
        

      

        
      </ul>
    </div>

    <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">

      
      
        
      
        
      
        
      
        
      

      
      
      <li class="nav-item">
        <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
      </li>
      

      
      
      <li class="nav-item dropdown theme-dropdown">
        <a href="#" class="nav-link" data-toggle="dropdown" aria-haspopup="true" aria-label="Display preferences">
          <i class="fas fa-moon" aria-hidden="true"></i>
        </a>
        <div class="dropdown-menu">
          <a href="#" class="dropdown-item js-set-theme-light">
            <span>Light</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-dark">
            <span>Dark</span>
          </a>
          <a href="#" class="dropdown-item js-set-theme-auto">
            <span>Automatic</span>
          </a>
        </div>
      </li>
      

      
      

    </ul>

  </div>
</nav>


  </div>

  <div class="page-body">
    
<div class="container-fluid docs">
    <div class="row flex-xl-nowrap">
        <div class="d-none d-xl-block col-xl-2 docs-toc">
            <ul class="nav toc-top">
                <li><a href="#" id="back_to_top" class="docs-toc-title">Contents</a></li>
            </ul>
            <nav id="TableOfContents">
  <ul>
    <li><a href="#new-to-code-club">New To Code Club?</a></li>
    <li><a href="#session-goals">Session Goals</a></li>
    <li><a href="#intro-the-ggplot-philosophy">Intro: The ggplot philosophy</a></li>
    <li><a href="#a-first-example">A First Example</a>
      <ul>
        <li><a href="#exercise-1">Exercise 1</a></li>
      </ul>
    </li>
    <li><a href="#the-power-of-aesthetics">The power of aesthetics</a>
      <ul>
        <li><a href="#exercise-2">Exercise 2</a></li>
      </ul>
    </li>
    <li><a href="#the-fitted-line-geom-geom_smooth">The fitted-line geom: <code>geom_smooth()</code></a>
      <ul>
        <li><a href="#exercise-3">Exercise 3</a></li>
      </ul>
    </li>
  </ul>
</nav>
        </div>
        <main class="col-12 col-md-0 col-xl-10 py-md-3 pl-md-5 docs-content" role="main">
            

            <article class="article">

                




















  
  
    
  


<div class="article-container pt-3">
  <h1>Code Club S02E06: Visualizing Data</h1>

  
  <p class="page-subtitle">Introduction to data visualization with <strong>ggplot2</strong></p>
  

  



<div class="article-metadata">

  
  
  
  
  <div>
    

  <span >
      <a href="/authors/michael-broe/">Michael Broe</a></span>
  </div>
  
  

  
  <span class="article-date">
    
    
    Sep 30, 2021
  </span>
  

  

  

  
  
  
  
  

  
  

</div>

  





</div>


<div class="article-header article-container featured-image-wrapper mt-4 mb-4" style="max-width: 720px; max-height: 542px;">
  <div style="position: relative">
    <img src="/codeclub/s02e06_ggplot2/featured_hu6a190b872561b7085418f3f8d3024635_514847_720x0_resize_q90_lanczos.jpg" alt="" class="featured-image">
    <span class="article-header-caption">Image from Allison Horst&rsquo;s Twitter</span>
  </div>
</div>



                <div class="article-container">
                    <div class="article-style">
                        <p><br> <br> <br></p>
<hr>
<h2 id="new-to-code-club">New To Code Club?</h2>
<ul>
<li>
<p>First, check out the <a href="/codeclub-setup/">Code Club Computer Setup</a> instructions, which also has some pointers that might be helpful if you&rsquo;re new to R or RStudio.</p>
</li>
<li>
<p>Please open RStudio before Code Club to test things out &ndash; if you run into issues, join the Zoom call early and we&rsquo;ll troubleshoot.</p>
</li>
</ul>
<hr>
<h2 id="session-goals">Session Goals</h2>
<ul>
<li>Learn the philosophy of <strong>coding</strong> a graphic.</li>
<li>Learn the basic <strong>template</strong> of a <strong>ggplot</strong> graphic, so you can reuse it for multiple chart types.</li>
<li>Learn how you can quickly add visual information to a graphic using <strong>aesthetics</strong> and <strong>layers</strong>.</li>
</ul>
<hr>
<h2 id="intro-the-ggplot-philosophy">Intro: The ggplot philosophy</h2>
<p>We have seen that in R, instead of manually manipulating data frames with a mouse as you do when editing Excel spreadsheets, we <strong>code</strong> the operations we want to perform using <code>dplyr</code> verbs like <code>select()</code>, <a href="https://rdrr.io/r/stats/filter.html" target="_blank" rel="noopener"><code>filter()</code></a>, <code>arrange()</code>, and so on.</p>
<p>In a similar way when performing visualization, instead of clicking on a chart type in Excel and selecting options, we <strong>code the chart</strong> in R.</p>
<p>And just as <code>dplyr</code> gives us efficient ways to manipulate data frames, <code>ggplot2</code> (which is also part of the tidyverse) gives us efficient ways to manipulate charts/plots/graphics (we use these terms interchangeably).</p>
<p>The <strong>gg</strong> in <strong>gg</strong>plot2 stands for <em>grammar of graphics</em>, a systematic approach for designing statistical plots developed by Leland Wilkinson. The idea behind this is to think about &lsquo;pulling apart&rsquo; various plots into their component pieces, then provide code that could put those pieces together again in different ways to make new plots.</p>
<p>This notion of pulling apart a graphic leads to the idea of <em>layers</em>. You can build up a plot of any complexity by <em>overlaying</em> different views of the same data.</p>
<p>There&rsquo;s a learning curve here for sure, but there are many advantages.</p>
<ul>
<li>First, <em>every</em> graphic shares a <strong>common template</strong>. Once you understand this structure you can &ldquo;say&rdquo; a lot of different things.<br>
(And I mean a <em>lot</em>. The <a href="https://github.com/rstudio/cheatsheets/blob/master/data-visualization-2.1.pdf" target="_blank" rel="noopener">ggplot cheat sheet</a> lists over 40 plot-types, but because this is a language in its own right, users can create their own <a href="https://exts.ggplot2.tidyverse.org/gallery/" target="_blank" rel="noopener">extensions</a> that you can also utilize, adding over 80 more.)</li>
<li>Second, because the graphic is now a piece of code, it is <em>reusable</em>. Once you have a little library of graphs you&rsquo;ve made, you can tweak them in various ways to make more. You can also share your code with others, so they can apply the same approach to their data.</li>
<li>Third, the way we put layers together is identical to the way we use pipes. You can read <code>%&gt;%</code> as <em>and then</em>: &ldquo;<code>select()</code> <em>and then</em> <a href="https://rdrr.io/r/stats/filter.html" target="_blank" rel="noopener"><code>filter()</code></a> <em>and then</em> <code>mutate()</code>&rdquo;. In graphics, the &ldquo;and then&rdquo; symbol is <a href="https://rdrr.io/r/base/Arithmetic.html" target="_blank" rel="noopener"><code>+</code></a>: &ldquo;show this layer, <em>and then</em> add this layer, <em>and then</em> add this layer&rdquo;. In this way you can gradually build up a graphic of increasing complexity; or on the other hand take someone else&rsquo;s code and simplify it by deleting layers.</li>
</ul>
<br>
<hr>
<h2 id="a-first-example">A First Example</h2>
<p>So how does this work in practice? We&rsquo;ll work through visualizing aspects of the <a href="https://allisonhorst.github.io/palmerpenguins/" target="_blank" rel="noopener">palmerpenguins</a> package which was introduced in <a href="https://biodash.github.io/codeclub/s02e04_tidyverse-intro-part1/" target="_blank" rel="noopener">S02E04 Intro to the Tidyverse</a>.</p>
<p>In addition, <strong>ggplot2</strong> is part of the tidyverse package so we need to load that as well:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='c'># this assumes you've already installed tidyverse and palmerpenguins with install.packages()</span>

<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://tidyverse.tidyverse.org'>tidyverse</a></span><span class='o'>)</span>
<span class='kr'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='o'>(</span><span class='nv'><a href='https://allisonhorst.github.io/palmerpenguins/'>palmerpenguins</a></span><span class='o'>)</span></code></pre>
</div>
<p>And recall that the <strong>penguins</strong> dataset is included in that package:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>glimpse</span><span class='o'>(</span><span class='nv'>penguins</span><span class='o'>)</span>
<span class='c'>#&gt; Rows: 344</span>
<span class='c'>#&gt; Columns: 8</span>
<span class='c'>#&gt; $ species           <span style='color: #555555; font-style: italic;'>&lt;fct&gt;</span> Adelie, Adelie, Adelie, Adelie, Adelie, Adelie, Adel…</span>
<span class='c'>#&gt; $ island            <span style='color: #555555; font-style: italic;'>&lt;fct&gt;</span> Torgersen, Torgersen, Torgersen, Torgersen, Torgerse…</span>
<span class='c'>#&gt; $ bill_length_mm    <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span> 39.1, 39.5, 40.3, NA, 36.7, 39.3, 38.9, 39.2, 34.1, …</span>
<span class='c'>#&gt; $ bill_depth_mm     <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span> 18.7, 17.4, 18.0, NA, 19.3, 20.6, 17.8, 19.6, 18.1, …</span>
<span class='c'>#&gt; $ flipper_length_mm <span style='color: #555555; font-style: italic;'>&lt;int&gt;</span> 181, 186, 195, NA, 193, 190, 181, 195, 193, 190, 186…</span>
<span class='c'>#&gt; $ body_mass_g       <span style='color: #555555; font-style: italic;'>&lt;int&gt;</span> 3750, 3800, 3250, NA, 3450, 3650, 3625, 4675, 3475, …</span>
<span class='c'>#&gt; $ sex               <span style='color: #555555; font-style: italic;'>&lt;fct&gt;</span> male, female, female, NA, female, male, female, male…</span>
<span class='c'>#&gt; $ year              <span style='color: #555555; font-style: italic;'>&lt;int&gt;</span> 2007, 2007, 2007, 2007, 2007, 2007, 2007, 2007, 2007…</span></code></pre>
</div>
<p>What is the correlation between bill-length and bill-width in these species?</p>
<p><img src="culmen_depth.png" style="width:50.0%" /></p>
<p>Are longer bills also deeper? We can explore this with a <strong>scatterplot</strong>.</p>
<p>But first let&rsquo;s look at the <code>ggplot()</code> <strong>template</strong>.</p>
<blockquote>
<p><em>Note:</em> the <em>package</em> is <strong>ggplot2</strong>, the <em>command</em> is just <code>ggplot()</code>.</p>
</blockquote>
<pre><code>ggplot(data = &lt;DATA&gt;) + 
  &lt;GEOM_FUNCTION&gt;(mapping = aes(&lt;MAPPINGS&gt;))
</code></pre>
<p>These are the obligatory parts of any plot: <strong>data</strong>, <strong>geoms</strong>, and <strong>mappings</strong>. The first argument to <code>ggplot()</code> is the data frame:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-3-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>This is not very interesting! but it&rsquo;s notable that it is <em>something</em>. <code>ggplot()</code> has created a base coordinate system (a base layer) that we can add visual layers to. It&rsquo;s completely uninfomative because we haven&rsquo;t said how we want the data (or even which data) to be mapped to graphic elements.</p>
<p>The <em>add a layer</em> operator is &ldquo;<strong>+</strong>&rdquo;, which is the <code>ggplot</code> equivalent of the pipe symbol, and best practice is to <strong>place it at the end of the line</strong>, just like the pipe. This makes your layers more readable in the code (and can lead to problems if you put it some other places).</p>
<p>The next argument specifies the kind plot we want: scatterplot, bar chart, fitted line, boxplot, violin plot, etc, etc. <strong>ggplot2</strong> refers to these as <strong>geoms</strong>: each different plot-type uses a different <strong>geom</strong>etrical object to represent data.</p>
<p>You can see an overview of many of these geoms in the <a href="https://github.com/rstudio/cheatsheets/blob/master/data-visualization-2.1.pdf" target="_blank" rel="noopener">cheat sheet</a>.The geom for a scatterplot is <code>geom_point()</code>.</p>
<p>But we also require a <code>mapping</code> argument, which maps the <em>variables</em> or columns in the dataset we want to focus on to their <em>visual representation</em> in the plot.</p>
<p>And finally we need to specify <strong>aesthetics</strong> for the geometric objects in the plot, which will control things like shape, color, transparency, etc. Perhaps surprisingly, the <code>x</code> and <code>y</code> coordinates of the points are aesthetics, since these control, not the shape or color, but the relative <strong>position</strong> of the points in the plot. (After all, flipping the <code>x</code> for <code>y</code> in some plots will flip the plot 90 degrees.)</p>
<p>Sounds like a lot when you spell it all out, but it&rsquo;s actually pretty simple in practice. Here then is our complete first plot:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-4-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Hmmm&hellip; the relationship is not very clear, but I guess if you squint your eyes you can maybe see some local positive correlations? We&rsquo;ll explore this relationship in more detail below using extra aesthetics.</p>
<p>But let&rsquo;s get coding!</p>
<h3 id="exercise-1">Exercise 1</h3>
<div class="puzzle">
<p>Create a scatterplot comparing body mass and flipper length. Does it look like there is any correlation?</p>
<details>
<summary>
Hints (click here)
</summary>
<br> Check the output of the glimpse() function to see what the precise variable names for body mass and flipper length are. It's up to you which you assign to x and y. <br> <br>
</details>
<details>
<summary>
Solution (click here)
</summary>
<br>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nf'>ggplot</span><span class='o'>(</span>data <span class='o'>=</span> <span class='nv'>penguins</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>body_mass_g</span>, y <span class='o'>=</span> <span class='nv'>flipper_length_mm</span><span class='o'>)</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-5-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>
<br>
<hr>
<h2 id="the-power-of-aesthetics">The power of aesthetics</h2>
<p>We saw above that the relationship between bill length and bill depth is not particularly clear; but that there <em>might</em> be some local patterns. How can we explore this? Is it possible that these clusters correspond somehow to the three different species: <em>Adelie</em>, <em>Chinstrap</em>, and <em>Gentoo</em>? Part of the problem with our original plot may be that we are lumping all the data from all three species together. We can quickly explore this by adding an extra aesthetic.</p>
<p>Our current plot uses two <strong>numeric</strong> variables: <code>bill_length_mm</code> and <code>bill_depth_mm</code>. We can add a third <strong>categorical</strong> variable, like <code>species</code>, to a two dimensional scatterplot by mapping it to a different visual aesthetic. We&rsquo;ve mapped length and depth to <code>x,y</code> coordinates. Now we&rsquo;ll simultaneously map species to <code>color</code> by expanding our list of aesthetics.</p>
<p>But before we do that, let&rsquo;s improve our plotting commands a little.</p>
<p>First, our template includes a general <code>mapping =</code> directive, which in turn includes a bunch of <em>specific</em> mappings from variables to aesthetics. While I think it&rsquo;s quite useful to be explicit that this is a mapping component (which connects columns with graphical objects) it turns out we can drop the <code>mapping =</code> syntax completely (and many people do). The very fact we have specific mappings inside <code>aes()</code> makes the <code>mapping =</code> redundant. (I&rsquo;ve come to think of it as &lsquo;syntactic sugar&rsquo;).</p>
<p>Second, we can actually pipe the dataset into the <code>ggplot()</code> command! We then set the mappings, and then choose our geom in a new layer. This is very common. Our new syntax looks like this:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-6-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Now let&rsquo;s add an extra aesthetic for the <code>species</code> variable. For this example we&rsquo;ll use the <code>color</code> aesthetic:</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span>, color <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-7-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>We get a neat legend on the right hand side for free. It&rsquo;s much clearer now that, within each species, there does seem to be a positive correlation between length and depth. It&rsquo;s just that the absolute values for each species fall in different regions of &ldquo;bill space&rdquo;.</p>
<blockquote>
<p><em>Notice:</em> We are piping <em>the entire raw dataset</em> into the plot here, while only graphing part of it. The elegance of the pipe syntax is that we can first pass the dataset through a series of <code>dplyr</code> operations - filtering and mutating etc. - and then pipe that modified dataset directly into our plotting commands. We&rsquo;ll see more of that in coming Code Clubs (and Exercise 3)!</p>
</blockquote>
<h3 id="exercise-2">Exercise 2</h3>
<div class="puzzle">
<p>Go back to your body_mass/flipper data and color by species, also using our new syntax.</p>
<p>There are various aesthetics you can use for a categorical variable (see help for geom_point). One is <code>shape</code>. Try this instead of <code>color</code> and see which you prefer.</p>
<details>
<summary>
Solutions (click here)
</summary>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
<span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>body_mass_g</span>, y <span class='o'>=</span> <span class='nv'>flipper_length_mm</span>, color <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-8-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
<span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>body_mass_g</span>, y <span class='o'>=</span> <span class='nv'>flipper_length_mm</span>, shape <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span>
</code></pre>
<p><img src="figs/unnamed-chunk-9-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<br>
</details>
</div>
<br>
<hr>
<h2 id="the-fitted-line-geom-geom_smooth">The fitted-line geom: <code>geom_smooth()</code></h2>
<p>In the scatterplot for bill length vs. bill depth, there wasn&rsquo;t a very clean overall positive relationship. We can make this even more apparent visually by fitting a line to the data: <em>overlaying</em> another geom in the same plot.</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
<span class='nf'>ggplot</span><span class='o'>(</span>mapping <span class='o'>=</span> <span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; `geom_smooth()` using method = 'loess' and formula 'y ~ x'</span>
</code></pre>
<p><img src="figs/unnamed-chunk-10-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<blockquote>
<p><em>Technical note:</em> by default, the <code>geom_smooth()</code> function doesn&rsquo;t use classical linear regression to fit the data. Instead it uses <strong>lo</strong>cally <strong>e</strong>stimated <strong>s</strong>catterplot <strong>s</strong>moothing: <a href="https://en.m.wikipedia.org/wiki/Local_regression" target="_blank" rel="noopener">loess</a>. This snakes around and tries to find a more local best fit to the data. You can request different fit algorithms with the <code>method</code> option. See help for <code>geom_smooth</code> for details.</p>
</blockquote>
<p>But again, this is messy since we are trying to analyze all the data at once. What if we set the species aesthetic before we apply <em>any</em> geom, so that the aesthetic is inherited by <em>both</em> geoms?</p>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
<span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>bill_length_mm</span>, y <span class='o'>=</span> <span class='nv'>bill_depth_mm</span>, color <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; `geom_smooth()` using method = 'loess' and formula 'y ~ x'</span>
</code></pre>
<p><img src="figs/unnamed-chunk-11-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<p>Here we can see there does seem to be a strong positive correlation if we analyze the data species-by-species.</p>
<h3 id="exercise-3">Exercise 3</h3>
<div class="puzzle">
<p>Revisit your flipper/body_mass data, and fit lines to each species.</p>
<p>There is a lot of overlap between two of the species in the scatterplot, which makes it hard to see the overall patterns for those two species. Does the fitted line overlay help make sense of this?</p>
<details>
<summary>
Solution (click here)
</summary>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
<span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>body_mass_g</span>, y <span class='o'>=</span> <span class='nv'>flipper_length_mm</span>, color <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; `geom_smooth()` using method = 'loess' and formula 'y ~ x'</span>
</code></pre>
<p><img src="figs/unnamed-chunk-12-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
<p>Perhaps you think this plot is too noisy, with the scatterplot data messing up the basic linear relationships. Modify your plot so it just shows the linear layer without the scatterplot layer.</p>
<details>
<summary>
Hints (click here)
</summary>
<br> Remember you can delete layers just as easily as you can add layers. <br> <br>
</details>
<details>
<summary>
Solution (click here)
</summary>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
<span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>body_mass_g</span>, y <span class='o'>=</span> <span class='nv'>flipper_length_mm</span>, color <span class='o'>=</span> <span class='nv'>species</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_smooth</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; `geom_smooth()` using method = 'loess' and formula 'y ~ x'</span>
</code></pre>
<p><img src="figs/unnamed-chunk-13-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
<p>(Bonus: if you have time&hellip;)</p>
<p>Pipe your dataset into a dplyr <a href="https://rdrr.io/r/stats/filter.html" target="_blank" rel="noopener"><code>filter()</code></a> command which pulls out just the <code>Gentoo</code> data; then pipe that data into the <code>ggplot()</code> command, to create a scatterplot, together with a smoothed-line overlay, for just that one species.</p>
<details>
<summary>
Hints (click here)
</summary>
<br> Remember two important things about the filter() function: (i) the check for equality is a double equals sign == (ii) the species name is a character string, not a number, so has to be quoted with "...". <br> <br>
</details>
<details>
<summary>
Solution (click here)
</summary>
<div class="highlight">
<pre class='chroma'><code class='language-r' data-lang='r'><span class='nv'>penguins</span> <span class='o'>%&gt;%</span> 
  <span class='nf'><a href='https://rdrr.io/r/stats/filter.html'>filter</a></span><span class='o'>(</span><span class='nv'>species</span> <span class='o'>==</span> <span class='s'>"Gentoo"</span><span class='o'>)</span> <span class='o'>%&gt;%</span> 
  <span class='nf'>ggplot</span><span class='o'>(</span><span class='nf'>aes</span><span class='o'>(</span>x <span class='o'>=</span> <span class='nv'>body_mass_g</span>, y <span class='o'>=</span> <span class='nv'>flipper_length_mm</span><span class='o'>)</span><span class='o'>)</span> <span class='o'>+</span>
  <span class='nf'>geom_point</span><span class='o'>(</span><span class='o'>)</span> <span class='o'>+</span> 
  <span class='nf'>geom_smooth</span><span class='o'>(</span><span class='o'>)</span>
<span class='c'>#&gt; `geom_smooth()` using method = 'loess' and formula 'y ~ x'</span>
</code></pre>
<p><img src="figs/unnamed-chunk-14-1.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
</details>
</div>

                    </div>

                    <br>

                    








<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=https://biodash.github.io/codeclub/s02e06_ggplot2/&amp;text=Code%20Club%20S02E06:%20Visualizing%20Data" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=https://biodash.github.io/codeclub/s02e06_ggplot2/&amp;t=Code%20Club%20S02E06:%20Visualizing%20Data" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=Code%20Club%20S02E06:%20Visualizing%20Data&amp;body=https://biodash.github.io/codeclub/s02e06_ggplot2/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=https://biodash.github.io/codeclub/s02e06_ggplot2/&amp;title=Code%20Club%20S02E06:%20Visualizing%20Data" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="whatsapp://send?text=Code%20Club%20S02E06:%20Visualizing%20Data%20https://biodash.github.io/codeclub/s02e06_ggplot2/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=https://biodash.github.io/codeclub/s02e06_ggplot2/&amp;title=Code%20Club%20S02E06:%20Visualizing%20Data" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>











  
  
    



  
  
  
  
  
  <div class="media author-card content-widget-hr">
    
      
      <a href="/authors/michael-broe/"><img class="avatar mr-3 avatar-circle" src="/authors/michael-broe/avatar_hu13632d739ac05d9f7c100f25ef9127c7_976850_270x270_fill_q90_lanczos_center.jpg" alt="Michael Broe"></a>
    

    <div class="media-body">
      <h5 class="card-title"><a href="/authors/michael-broe/">Michael Broe</a></h5>
      <h6 class="card-subtitle">Bioinformatician at EEOB</h6>
      
      <ul class="network-icon" aria-hidden="true">
  
    
    
    
      
    
    
    
    
    
    <li>
      <a href="mailto:broe.1@osu.edu" >
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  
    
    
    
      
    
    
    
    
    
      
    
    <li>
      <a href="https://github.com/michaelbroe" target="_blank" rel="noopener">
        <i class="fab fa-github"></i>
      </a>
    </li>
  
</ul>

    </div>
  </div>


  
















                    <br>

                    <hr class="separator">

                    <div class="clearfix">
                        
                        
                        <span class="previous-link">
                            <a href="https://biodash.github.io/codeclub/s02e05_tidyverse-intro-part2/">← Code Club S02E05: Intro to the Tidyverse (Part 2)</a>
                        </span>
                        
                        
                        
                        
                        <span class="next-link">
                            <a href="https://biodash.github.io/codeclub/s02e07_ggplot2_part2/">S02E07: Intro to ggplot2 (part 2) →</a>
                        </span>
                        
                        
                    </div>

                    <hr class="separator">

                </div>

            </article>

            
        </main>
    </div>
</div>

  </div>

  <div class="page-footer">
    
    
    <div class="container">
      <footer class="site-footer">
    
    <p class="powered-by">
      <a href="/license/">LICENSE: CC-BY-SA <br> <i class="fab fa-creative-commons fa-2x"></i><i class="fab fa-creative-commons-by fa-2x"></i><i class="fab fa-creative-commons-sa fa-2x"></i> </a><br>
    </p>
    
</footer>
    </div>
    
  </div>

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

      

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/instant.page/5.1.0/instantpage.min.js" integrity="sha512-1+qUtKoh9XZW7j+6LhRMAyOrgSQKenQ4mluTR+cvxXjP1Z54RxZuzstR/H9kgPXQsVB8IW7DMDFUJpzLjvhGSQ==" crossorigin="anonymous"></script>

      
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      

      
      

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/highlight.min.js" integrity="sha512-TDKKr+IvoqZnPzc3l35hdjpHD0m+b2EC2SrLEgKDRWpxf2rFCxemkgvJ5kfU48ip+Y+m2XVKyOCD85ybtlZDmw==" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.2.0/languages/r.min.js"></script>
        
      

    

    
    
    
      <script async defer src="https://maps.googleapis.com/maps/api/js?key="></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/gmaps.js/0.4.25/gmaps.min.js" integrity="sha256-7vjlAeb8OaTrCXZkCNun9djzuB2owUsaO72kXaFDBJs=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
      
      <script id="search-hit-fuse-template" type="text/x-template">
        <div class="search-hit" id="summary-{{key}}">
          <div class="search-hit-content">
            <div class="search-hit-name">
              <a href="{{relpermalink}}">{{title}}</a>
              <div class="article-metadata search-hit-type">{{type}}</div>
              <p class="search-hit-description">{{snippet}}</p>
            </div>
          </div>
        </div>
      </script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
      
    

    
    

    
    
    
    

    
    <script src="/js/bootstrap.bundle.min.6aed84840afc03ab4d5750157f69c120.js"></script>

    
    
      
      
      
      
      
      
      
    

    
    
    
    
    
    
    
    
      
      
    
    
      
    
    <script src="/en/js/wowchemy.min.2f16af175a3ec51b41b36bb4934b1cf7.js"></script>

    






</body>
</html>
