<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Preprocessing FASTQ files</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="slides.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">

class:inverse middle center



# Preprocessing FASTQ files

----

### Making them read-only, and concatenating files from different lanes

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

### Jelmer Poelstra, MCIC Wooster
### 2021/02/17 (updated: 2021-02-16)
  
---
class: inverse middle center

# Making raw data as read-only:
# Viewing and modifying file permissions
----
&lt;br&gt;&lt;br&gt;&lt;br&gt;

---

## Showing file permissions

- To show file permissions, use **`ls`**  with the **`-l`** (*long* format)
  option.shows file permissions.
  
  (The command below also uses the **`-a`** option to show all files,
  including hidden ones, and **`-h`** shows file sizes in "human-readable"
  format.)

&lt;p align="center"&gt;
&lt;img src=img/long-ls.png width="700"&gt;
&lt;/p&gt;

---

## File permission notation in `ls -l`

&lt;br&gt;

&lt;p align="center"&gt;
&lt;img src=img/permissions.svg width="50%"&gt;
&lt;/p&gt;

---

## Changing file permissions

This can be done in several ways with the `chmod` command:

```sh
# chmod &lt;who&gt;=&lt;permission-to-set&gt;`:
$ chmod a=rwx  # a: all / u: user / g: group / o: others
   
# chmod &lt;who&gt;+&lt;permission-to-add&gt;:
$ chmod a+r  
# chmod &lt;who&gt;-&lt;permission-to-remove&gt;:
$ chmod o-w
```

--

- Using numbers:
  
.pull-left[
| Nr | Permission     |
|----|----------------|
| 1  | e**x**ecute    |
| 2  | **w**rite      |
| 4  | **r**ead       |
| 5  | r + x          |
| 6  | r + w          |
| 7  | r + w + x      |
]

.pull-right[
Three numbers are needed to specify permissions for user, group, and others:
```sh
# rwx permissions for all:
$ chmod 777 file.txt
# rwx for me, r for others:
$ chmod 744 file.txt
```
]

---

## File permissions &amp;ndash; Making your data read-only

- **To make files read-only:**
  ```sh
  $ chmod a=r data/raw/*       # *a*ll = *r*ead
  
  $ chmod u-w data/raw/*       # Remove write permission for *u*ser
  
  $ chmod 400 data/raw/*       # User: read // group &amp; others: none
  ```

- To operate on directories and all their contents, you need the `-R` flag.  
  (If you do this, you can't create new files in these directories either.)
  
  ```sh
  $ chmod -R a=r data/raw
  ```
  

&lt;br&gt;

.content-box-info[
As the owner of the file, you can always change file permission back.
]

---

## <i class="fas  fa-user-edit "></i> Changing file permissions: In practice

- First, let's create a file and look at its original permissions:
  
  ```sh
  $ touch testfile
  $ ls -l testfile
  #&gt; -rw-r--r-- 1 jelmer PAS0471 0 Feb 15 16:42 testfile
  ```

- Then, limit permissions for **a**ll to **r**ead, and check what we did:
  
  ```sh
  $ chmod a=r testfile
  $ ls -l testfile
  #&gt; -r--r--r-- 1 jelmer PAS0471 0 Feb 15 16:42 testfile
  ```

---

## <i class="fas  fa-user-edit "></i> Changing file permissions: In practice

- What happens when we try to overwrite write-protected files?
  
  ```sh
  $ echo "blabla" &gt; testfile
  #&gt; bash: testfile: Permission denied
  ```

- What happens when we try to remove write-protected files?
  ```sh
  # Answer "y"/"yes" to remove, "n"/"no" to cancel:
  $ rm testfile
  #&gt; rm: remove write-protected regular empty file ‘testfile’?
  
  # With the -f flag, the file will be removed without prompting:
  $ rm -f testfile
  ```

  .content-box-info[
  While *you* (the file owner) can still remove write-protected files,
  others really can't read/write/execute files without appropriate permissions.
  ]

---
class: inverse middle center

# Concatenating files from different lanes
----
&lt;br&gt;&lt;br&gt;&lt;br&gt;

---

## Concatenating files from different lanes

- For each sample, we have 4 files: 2 read directions (`R1` and `R2`),
  and 2 lanes (`L001` and `L002`).

```sh
$ tree data/fastq
#&gt;data/fastq
#&gt;├── md5.txt
#&gt;    ├── nreads.txt
#&gt;    ├── README.txt
#&gt;    ├── X6465_Benitez-PonceM_C6_myb2_V1N_1
#&gt;    │   ├── checksum_files.md5
#&gt;    │   ├── X6465_Benitez-PonceM_C6_myb2_V1N_1_S1_L001_R1_001.fastq.gz
#&gt;    │   ├── X6465_Benitez-PonceM_C6_myb2_V1N_1_S1_L001_R2_001.fastq.gz
#&gt;    │   ├── X6465_Benitez-PonceM_C6_myb2_V1N_1_S1_L002_R1_001.fastq.gz
#&gt;    │   └── X6465_Benitez-PonceM_C6_myb2_V1N_1_S1_L002_R2_001.fastq.gz
#&gt;    ├── X6466_Benitez-PonceM_C6_myb3_V1N_1
#&gt;    │   ├── checksum_files.md5
#&gt;    │   ├── X6466_Benitez-PonceM_C6_myb3_V1N_1_S2_L001_R1_001.fastq.gz
#&gt;    │   ├── X6466_Benitez-PonceM_C6_myb3_V1N_1_S2_L001_R2_001.fastq.gz
#&gt;    │   ├── X6466_Benitez-PonceM_C6_myb3_V1N_1_S2_L002_R1_001.fastq.gz
#&gt;    │   └── X6466_Benitez-PonceM_C6_myb3_V1N_1_S2_L002_R2_001.fastq.gz
```

---

## Concatenating files from different lanes (cont.)

Since our end goal is to quantify gene expression on a per-sample basis,
we can concatenate the files from the two different lanes.

(After mapping, we will also combine R1 and R2, so we will go down to 1 file
per sample.)

This is extremely simple &amp;ndash; to concatenate two gzipped FASTQ file,
we can use `cat`:

```sh
$ cat file1.fastq.gz file2.fastq.gz &gt; concat.fastq.gz
```

---

## Concatenating files from different lanes (cont.)

This also gives us an opportunity to get rid of the non-informative parts
of our file names &amp;ndash; for instance, only `C6_myb3` is the sample name
in these file names:

```sh
X6466_Benitez-PonceM_C6_myb3_V1N_1_S2_L001_R1_001.fastq.gz
X6466_Benitez-PonceM_C6_myb3_V1N_1_S2_L002_R1_001.fastq.gz
```

Therefore, we can name the concatenated file:
```sh
C6_myb3_R1.fastq.gz
```

---

## Code to concatenate all files

```sh
outdir=results/fastq_concat

for fastq_dir in data/fastq/X646[5-9]* data/fastq/X647[0-9]*; do
  
  for R in R1 R2; do

    L1=$(ls "$fastq_dir"/*_L001_"$R"_*fastq.gz) # File from lane 1
    L2=${L1/_L001_/_L002_}                      # File from lane 2

    sample_id=$(basename "$L1" | \
        sed -E 's/X64[6-7][0-9]_Benitez-PonceM_(.*)_V1N.*fastq.gz/\1/')
    
    concat="$outdir"/"$sample_id"_"$R".fastq.gz # File with all seqs

    cat "$L1" "$L2" &gt; "$concat" # Concatenate the L1 and L2 seqs
    
    ls -lh "$L1" # List the input and output files as a check
    ls -lh "$L2"
    ls -lh "$concat"
    echo "----"
  done

done
```

---

## Sample info

| Sample ID	|  AMF	      | Treatment
|-----------|-----------|-------------------
| C6_myb2	  |  Ri	      | Agrobacterium_noexp
| C6_myb3	  |  Ri	      | Agrobacterium_noexp
| C6_myb4	  |  Ri	      | Agrobacterium_noexp
| CT_05	    |  Ri	      | mock
| CT_06	    |  Ri	      | mock
| CT_09	    |  Ri	      | mock
| T_26	    |  Ri	      | Agrobacterium_myb
| T_29	    |  Ri	      | Agrobacterium_myb
| T_30	    |  Ri	      | Agrobacterium_myb
| cnt_1	    |  control	| mock
| cnt_2	    |  control	| mock
| cnt_3	    |  control	| mock
| myb_1	    |  control	| Agrobacterium_myb
| myb_2	    |  control	| Agrobacterium_myb
| myb_3	    |  control	| Agrobacterium_myb


---

## Sample info (cont.)

All samples are from tomato roots.

| Column     | descriptions	
|------------|-----------------------------------
| **mock**	     | Mock *Agrobacterium* inoculation
| **myb	**       | Inoculated with *Agrobacterium*, expressing *myb* reporter
| **no_exp**	   | Inoculated with *Agrobacterium*, root does not express *myb*
| **Ri**	       | *Rhizogenes intraradices* inoculation and observed colonization
| **control**	   | No AMF (Ri) inoculation
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "rainbow",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
