<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>MultiQC at OSC</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="slides.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">

class:inverse middle center



# Running MultiQC at OSC

----

&lt;br&gt; &lt;br&gt; &lt;br&gt;

### Jelmer Poelstra, MCIC Wooster
### 2021/02/12 (updated: 2021-02-11)

  
---

## MultiQC

MultiQC is a program that *aggregates and summarizes QC log files* into a single
nicely formatted HTML page.

It is most often used to summarize FastQC results, but can also be used to
summarize STAR and featureCounts output logs, as we'll do later.

&lt;br&gt;

&lt;p align="center"&gt;
&lt;img src=img/multiqc_front.png width="90%"&gt;
&lt;/p&gt;

---

## MultiQC

- For instance, this figure shows mean quality scores across 132 samples:

&lt;p align="center"&gt;
&lt;img src=img/multiqc_qual.png width="100%"&gt;
&lt;/p&gt;

---

## MultiQC and Conda at OSC

MultiQC is *not* available as a module at OSC (i.e., no `module load`).

The best way to use it is by installing it for yourself
**with the software manager Conda**, which:
  
  - Creates *environments* with one (best-practice) or more software packages.

  - Lots of bioinformatics software is available as a Conda package.
  
  - Handles dependencies but does not require admin rights.
  
  - Environments are activated and deactivated like with the `module` system.


---

## Conda at OSC

- To use Conda at OSC, you first need to load a module:

  ```sh
  $ module load python/3.6-conda5.2
  ```
  
  .content-box-info[
  Like with other software, this module needs to be loaded every time you
  want to use Conda.
  ]

---

## One-time installation of MultiQC &lt;br&gt; into a Conda environment

- Now, we are ready to create a new Conda environment with Python: 
 
  ```sh
  $ conda create -y -n multiqc-env python=3.7
  ```

  - `create` is the Conda command to create a new environment.
  
  - `-y` avoids a confirmation prompt.
  
  - `-n multiqc-env` is the name for the new environment, which is arbitrary.
  
  - `python=3.7` is how specify that we want to install Python and specifically,
     Python version 3.7.
  
---

## One-time installation of MultiQC &lt;br&gt; into a Conda environment

- Next, we *activate* the environment and install MultiQC into it:

  ```sh
  $ source activate multiqc-env
  
  $ conda install -y -c bioconda -c conda-forge multiqc
  ```

  - `-c bioconda -c conda-forge` is how we indicate which so-called
    *channels*, i.e. repositories, we want to use.
    (These sorts of details are provided in software installation instructions.)

---

## Running MultiQC from now on

- Now that we have MultiQC installed into a Conda environment at OSC,  
  we just need to take the following steps whenever we want to use it:
  
  ```sh
  $ module load python/3.6-conda5.2
  $ source activate multiqc-env
  ```

- Having done that, you should be able to run MultiQC:

  ```sh
  $ multiqc --help
  ```

&lt;br&gt;

.content-box-info[
If you didn't manage to create your own conda environment,  
you can also use mine:

```sh
$ source activate /users/PAS0471/jelmer/.conda/envs/multiqc
```
]

---

## Running MultiQC

Actually running MultiQC is straightforward:

```sh
# This will put the MultiQC report in the working dir:
$ multiqc &lt;dir-with-fastqc-reports&gt;

# You can also specify the output dir for the MultiQC report: 
$ multiqc &lt;dir-with-fastqc-reports&gt; -o &lt;output-dir&gt;
```

&lt;br&gt;

--

For example, to start an interactive job and run MultiQC there:

```sh
$ cd /fs/project/PAS0471/teach/misc/2021-02_rnaseq/$USER

# Start an interactive job at OSC:
$ sinteractive -A PAS0471 # No time specified: will be 30 minutes

$ module load python/3.6-conda5.2  # Load the conda module
$ source activate multiqc          # Activate the MultiQC env.

$ multiqc results/QC_fastq/ -o results/QC_fastq/ # Run MultiQC

$ ls -lh results/QC_fastq/multiqc_report.html    # Main output file
```

---

## MultiQC output

&lt;br&gt;

.content-box-diy[
Let's take a look at the output file via OnDemand's file browser.
]

---

## MultiQC script

We could also run MultiQC in a script (`scripts/QC_fastq/multiqc.sh`):

```sh
#!/bin/bash
#SBATCH --account=PAS0471
#SBATCH --time=60
#SBATCH --out=slurm-multiqc-%j.out
set -e -u -o pipefail

# Save command-line arguments in named variables:
input_dir="$1"
output_dir="$2" 

mkdir -p "$output_dir" # If necessary, create the output dir

# Report before starting the pogram:
echo "Running MultiQC for dir: $input_dir"
echo "Output dir: $output_dir"

# Load the conda module and the MultiQC conda environment:
module load python/3.6-conda5.2
source activate /users/PAS0471/jelmer/.conda/envs/multiqc

# Run fastqc:
multiqc "$input_dir" -o "$output_dir"
```

---

## Running our MultiQC script

- Submitting the script as a job:
  ```sh
  $ dir=results/QC_fastq/
  $ sbatch multiqc.sh $dir $dir
  ```

&lt;br&gt;

- Checking the queue:
  ```sh
  $ squeue -u $USER
  ```

- Checking the output:
  ```sh
  $ ls
  
  $ less slurm*
  
  $ ls $dir/multiqc_report.html
  ```

---
class: inverse middle center

# Questions?

----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "rainbow",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
