<!DOCTYPE html>
<html lang="en"><head>
<script src="presentation_files/libs/clipboard/clipboard.min.js"></script>
<script src="presentation_files/libs/quarto-html/tabby.min.js"></script>
<script src="presentation_files/libs/quarto-html/popper.min.js"></script>
<script src="presentation_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="presentation_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="presentation_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="presentation_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="presentation_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<link href="presentation_files/libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="presentation_files/libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.302">

  <meta name="author" content="Jelmer Poelstra">
  <meta name="dcterms.date" content="2023-04-04">
  <title>Command-line genomics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #007020; background-color: #f3f5f6; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #007020; } /* Normal */
    code span.al { color: #ff0000; font-weight: bold; } /* Alert */
    code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
    code span.at { color: #7d9029; } /* Attribute */
    code span.bn { color: #40a070; } /* BaseN */
    code span.bu { color: #008000; } /* BuiltIn */
    code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #4070a0; } /* Char */
    code span.cn { color: #880000; } /* Constant */
    code span.co { color: #60a0b0; font-style: italic; } /* Comment */
    code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
    code span.do { color: #ba2121; font-style: italic; } /* Documentation */
    code span.dt { color: #902000; } /* DataType */
    code span.dv { color: #40a070; } /* DecVal */
    code span.er { color: #ff0000; font-weight: bold; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #40a070; } /* Float */
    code span.fu { color: #06287e; } /* Function */
    code span.im { color: #008000; font-weight: bold; } /* Import */
    code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
    code span.kw { color: #007020; font-weight: bold; } /* Keyword */
    code span.op { color: #666666; } /* Operator */
    code span.ot { color: #007020; } /* Other */
    code span.pp { color: #bc7a00; } /* Preprocessor */
    code span.sc { color: #4070a0; } /* SpecialChar */
    code span.ss { color: #bb6688; } /* SpecialString */
    code span.st { color: #4070a0; } /* String */
    code span.va { color: #19177c; } /* Variable */
    code span.vs { color: #4070a0; } /* VerbatimString */
    code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link rel="stylesheet" href="slides.css">
  <link href="presentation_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Command-line genomics</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Jelmer Poelstra 
</div>
        <p class="quarto-title-affiliation">
            MCIC Wooster, OSU
          </p>
    </div>
</div>

  <p class="date">2023-04-04</p>
</section>
<section class="slide level2">

<h3 id="what-ill-talk-about-today">What I’ll talk about today</h3>
<p>Today, I will give an overview of a <strong>typical computational environment/</strong> <strong>infrastructure used for all kinds of genomics projects</strong>.</p>
<p><br></p>
<p>Of course, I won’t have time to teach you the different components, but I would like to orient you on this topic so that it’s not as much of a black box as it may seem now.</p>
<p>This will hopefully give you a starting point for learning more – and I will point you to some specific resources as well.</p>
</section>
<section class="slide level2">

<h3 id="what-kind-of-research-is-this-talk-relevant-for">What kind of research is this talk relevant for?</h3>
<div>
<ul>
<li class="fragment"><p>Any project in which you generate <em>high-throughput sequencing data</em>, e.g.:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Whole-genome sequencing</strong> – de novo assembly, pangenomics, “resequencing”</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Reduced-representation sequencing</strong> (GBS, etc) for population genomics</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Microbiomics</strong> – both shotgun metagenomics and amplicon metabarcoding</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> <strong>Transcriptomics</strong> with RNAseq</p></li>
</ul>
<p><br></p>
<ul>
<li class="fragment">Other genomics projects like <strong>comparative genomics</strong> with publicly available genomes.</li>
</ul>
<p><br></p>
<ul>
<li class="fragment">But only to some extent to <em>proteomics</em> and <em>metabolomics</em>.</li>
</ul>
</div>
</section>
<section class="slide level2">

<h3 id="overview">Overview</h3>
<ol type="1">
<li><p>Introduction</p></li>
<li><p><strong>Command-line genomics?</strong></p></li>
<li><p>Ohio Supercomputer Center (OSC) overview</p></li>
<li><p>Command-line software</p></li>
<li><p>The VS Code editor and the whole game</p></li>
</ol>
</section>
<section class="slide level2">

<h3 id="a-typical-computational-infrastructure-for-genomics">A typical computational infrastructure for genomics</h3>
<div>
<ul>
<li class="fragment"><p>A <strong>supercomputer</strong> – in our case, the Ohio Supercomputer Center (OSC)<br>
<em>(Cloud computing is an alternative, but won’t be covered today)</em></p></li>
<li class="fragment"><p>The <strong>Unix shell</strong> (AKA the terminal)</p></li>
<li class="fragment"><p>A <strong>text editor</strong> – I recommend and will demonstrate <em>VS Code</em></p></li>
</ul>
</div>
</section>
<section class="slide level2">

<h3 id="a-typical-computational-infrastructure-for-genomics-1">A typical computational infrastructure for genomics</h3>
<ul>
<li><p>A <strong>supercomputer</strong> – in our case, the Ohio Supercomputer Center (OSC)<br>
<em>(Cloud computing is an alternative, but won’t be covered today)</em></p></li>
<li><p>The <strong>Unix shell</strong> (AKA the terminal)</p></li>
<li><p>A <strong>text editor</strong> – I recommend and will demonstrate <em>VS Code</em></p></li>
</ul>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Command-line genomics</strong></p>
</div>
<div class="callout-content">
<p>For the purposes of this talk, I will refer to working with the above elements by running command-line programs as “batch jobs” (non-interactively) as <strong>command-line genomics</strong>.</p>
</div>
</div>
</div>
</section>
<section class="slide level2">

<h3 id="a-typical-computational-infrastructure-for-genomics-2">A typical computational infrastructure for genomics</h3>
<ul>
<li><p>A <strong>supercomputer</strong> – in our case, the Ohio Supercomputer Center (OSC)<br>
<em>(Cloud computing is an alternative, but won’t be covered today)</em></p></li>
<li><p>The <strong>Unix shell</strong> (AKA the terminal)</p></li>
<li><p>A <strong>text editor</strong> – I recommend and will demonstrate <em>VS Code</em></p></li>
</ul>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Command-line genomics</strong></p>
</div>
<div class="callout-content">
<p>For the purposes of this talk, I will refer to working with the above elements by running command-line programs as “batch jobs” (non-interactively) as <strong>command-line genomics</strong>.</p>
</div>
</div>
</div>
<ul>
<li><strong>R</strong> or perhaps <em>Python</em> for interactive statistical analysis and visualization.<br>
<em>(I won’t talk more about this other than showing you how you can use R at OSC)</em></li>
</ul>
</section>
<section class="slide level2">

<h3 id="what-is-a-supercomputer">What is a supercomputer?</h3>
<p>A highly interconnected set of many computer processors and storage units. You can think of it simply as <em>a network of computers</em>.</p>
<p>Supercomputers are also commonly referred to as High-Performance Computing (HPC) clusters or simply compute clusters.</p>
</section>
<section class="slide level2">

<h3 id="why-do-i-need-a-supercomputer">Why do I need a supercomputer?</h3>
<ul>
<li><p>Your genomics dataset is likely to be too large to be handled efficiently, or even at all, by a laptop or desktop computer.</p></li>
<li><p>To speed up long-running analyses by using more computing power.</p></li>
<li><p>To speed up analyses where some part needs be repeated many times, like the independent mapping of reads for different samples to a reference genome.</p></li>
<li><p>It’s also a great place to store large amounts of data</p></li>
</ul>
</section>
<section class="slide level2">

<h3 id="what-is-the-unix-shell">What is the Unix shell?</h3>
<p>A computer’s <strong>shell</strong> is also referred to as a <em>Terminal</em> or “<em>the command line</em>”, and allows you to interact with your computer by <strong>typing commands</strong> rather than pointing-and-clicking.</p>
<p><br></p>
<p>The <em>Unix</em> shell is the shell of Unix-based computers, which include Mac and Linux (but not Windows) operating systems.<sup>1</sup></p>
<p><br></p>
<p align="center">
<img src="img/shell.png" width="70%">
</p>
<aside><ol class="aside-footnotes"><li id="fn1"><p>And the most common Unix shell is the <strong>Bash shell</strong>, which runs the Bash language.</p></li></ol></aside></section>
<section class="slide level2">

<h3 id="why-do-i-need-to-work-in-the-unix-shell">Why do I need to work in the Unix shell?</h3>
<ul>
<li><p>A genomics project usually involves sequentially running a whole array of bioinformatics programs (or “tools”).</p>
<p>For instance, an RNAseq project may include:<br>
raw read QC =&gt; raw read trimming =&gt; trimmed read mapping =&gt; gene counting</p></li>
</ul>
<p><br></p>
<ul>
<li>Many of these tools can only be run through a “command-line interface” (CLI)</li>
</ul>
</section>
<section class="slide level2">

<h3 id="why-do-i-need-to-work-in-the-unix-shell-1">Why do I need to work in the Unix shell?</h3>
<ul>
<li><p>A genomics project usually involves sequentially running a whole array of bioinformatics programs (or “tools”).</p>
<p>For instance, an RNAseq project may include:<br>
raw read QC =&gt; raw read trimming =&gt; trimmed read mapping =&gt; gene counting</p></li>
</ul>
<p><br></p>
<ul>
<li><p>Many of these tools can only be run through a “command-line interface” (CLI)</p></li>
<li><p>Even those that have a “graphical user interface” (GUI) are more efficiently and reproducibly run through a CLI:</p>
<ul>
<li><p>Efficiency: A CLI allows you to write a simple loop to run it in the same way for many samples. (In combination with the computing power of a supercomputer, this in turn allows you to process those hundreds of samples <em>in parallel</em>.)</p></li>
<li><p>Reproducibility: You can easily save all commands and scripts which would allow you to rerun a project rather straightforwardly.</p></li>
</ul></li>
</ul>
</section>
<section class="slide level2">

<h3 id="howwhen-can-i-avoid-all-of-this">How/when can I <em>avoid</em> all of this?</h3>
<p>If you will often do genomics projects like the ones mentioned above, it’s hard to avoid command line genomics as described.</p>
<p>But here are some conditions in which you might reasonably avoid it:</p>
<ul>
<li><p>You’re doing a single genomics project, your main research focus is elsewhere</p></li>
<li><p>You have data which can be analyzed with no or a relatively small command-line-based part, such as proteomics/metabolomics/metabarcoding/RNAseq.</p></li>
</ul>
<p>In such cases, you might be able to get someone else to do the command line work, or you could try <strong><em>Galaxy</em></strong>, a cloud-based bioinformatics platform with a web browser interface and no coding.</p>
</section>
<section class="slide level2">

<h3 id="overview-1">Overview</h3>
<ol type="1">
<li><p>Introduction</p></li>
<li><p>Command-line genomics?</p></li>
<li><p><strong>Ohio Supercomputer Center (OSC) overview</strong></p></li>
<li><p>Command-line software</p></li>
<li><p>The VS Code editor and the whole game</p></li>
</ol>
</section>
<section class="slide level2">

<h3 id="the-ohio-supercomputer-center-osc">The Ohio Supercomputer Center (OSC)</h3>
<div>
<ul>
<li class="fragment"><p>The <strong>Ohio Supercomputer Center (OSC)</strong> provides computing resources to researchers (and others) across Ohio.</p>
<p>OSC has two individual supercomputers/clusters (named Owens and Pitzer), and lots of infrastructure for their usage.</p></li>
<li class="fragment"><p>Research usage is charged but at <a href="https://www.osc.edu/content/academic_fee_model_faq">heavily subsidized rates</a>, and most or all OSU colleges absorb these costs at the college level (!)</p>
<p>Educational usage is entirely free, like for the <code>PAS2250</code> project you have been added to for this lecture.</p></li>
</ul>
</div>
<p align="center">
<img src="img/osc_logo.png" width="560">
</p>
</section>
<section class="slide level2">

<h3 id="supercomputer-overview">Supercomputer overview</h3>
<p align="center">
<img src="img/cluster_login.jpg" width="80%">
</p>
</section>
<section class="slide level2">

<h3 id="osc-ondemand">OSC OnDemand</h3>
<p>The OSC OnDemand web portal allows you to access OSC resources through a web browser, such as:</p>
<ul>
<li><p>A file browser where you can also create and rename folders and files, and download and upload files. <sup>1</sup></p></li>
<li><p>A Unix shell</p></li>
<li><p>More than a dozen different “Interactive Apps”, or programs with a GUI, such as RStudio, Jupyter, QGIS, and more.</p></li>
</ul>
<p><br></p>
<p>(<em>Project and user management goes through a separate website, <a href="https://my.osc.edu" class="uri">https://my.osc.edu</a>.</em>)</p>
<aside><ol class="aside-footnotes"><li id="fn2"><p>But for large transfers, use a different interface: <em>Globus</em></p></li></ol></aside></section>
<section class="slide level2">

<h3 id="connecting-to-osc-with-ondemand">Connecting to OSC with OnDemand</h3>
<ul>
<li>Go to <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>, and log in with your OSC credentials.</li>
</ul>
<p align="center">
<img src="img/ondemand_homepage.png" width="80%">
</p>
</section>
<section class="slide level2">

<h3 id="ondemand-files-menu">OnDemand “Files” menu</h3>
<p>Choose a folder as a starting point for the file browser:</p>
<p align="center">
<img src="img/ondemand_files.png" width="60%">
</p>
</section>
<section class="slide level2">

<h3 id="ondemand-files-menu-1">OnDemand “Files” menu</h3>
<p>Here you can view, create and rename folders and files, and download and upload files:</p>
<p align="center">
<img src="img/ondemand_files2.png" width="85%">
</p>
</section>
<section class="slide level2">

<h3 id="ondemand-clusters-menu">OnDemand “Clusters” menu</h3>
<p>Here you can access a Unix shell on either of the two clusters:</p>
<p><br></p>
<p align="center">
<img src="img/ondemand_clusters.png" width="40%">
</p>
<p><br></p>
<p><em>(Since the two clusters share the file systems, and they have fairly similar capabilities, it generally doesn’t matter which cluster you connect to).</em></p>
</section>
<section class="slide level2">

<h3 id="ondemand-clusters-menu-shell">OnDemand “Clusters” menu: Shell</h3>
<p>When you click on one of the shell options, a new browser tab with a shell will open. There are some welcome messages, and some storage usage/quota info, and then you get a prompt to type commands:</p>
<p align="center">
<img src="img/ondemand_terminal.png" width="95%">
</p>
</section>
<section class="slide level2">

<h3 id="ondemand-clusters-menu-system-status">OnDemand “Clusters” menu: System Status</h3>
<p>If you click on “System status”, you’ll get an overview of the current usage of the two clusters:</p>
<p><br></p>
<p align="center">
<img src="img/ondemand_systemstatus.png" width="100%">
</p>
</section>
<section class="slide level2">

<h3 id="ondemand-interactive-apps-menu">OnDemand “Interactive Apps” menu</h3>
<p>Here you can mostly access programs with a GUI that will run on a compute node. We’ll try <code>Code Server</code> and <code>RStudio Server</code>.</p>
<p align="center">
<img src="img/ondemand_interactiveapps.png" width="29%">
</p>
</section>
<section class="slide level2">

<h3 id="ondemand-interactive-apps-menu-rstudio-form">OnDemand “Interactive Apps” menu: RStudio Form</h3>
<p>Fill out this form to start an RStudio session. This will run on a compute node and is therefore charged: for that reason, it needs the OSC account number so as to bill the correct account.</p>
<p align="center">
<img src="img/ondemand_rstudioform.png" width="43%">
</p>
</section>
<section class="slide level2">

<h3 id="ondemand-interactive-apps-rstudio">OnDemand Interactive Apps: RStudio</h3>
<p>Once that top bar below is green and says “Running”, you can click “Connect to RStudio Server” way at the bottom:</p>
<p align="center">
<img src="img/ondemand_rstudiocard.png" width="70%">
</p>
</section>
<section class="slide level2">

<h3 id="ondemand-interactive-apps-rstudio-1">OnDemand Interactive Apps: RStudio</h3>
<p>Now, you’ll have RStudio running in your browser!<br>
It looks identical to the Desktop App version:</p>
<p align="center">
<img src="img/ondemand_rstudio.png" width="61%">
</p>
</section>
<section class="slide level2">

<h3 id="what-is-different-at-osc-on-a-supercomputer">What is different at OSC / on a supercomputer?</h3>
<p><em>(That is, even compared to command-line computing on a personal Linux</em> <em>computer.)</em></p>
<div>
<ul>
<li class="fragment"><p><strong>Software</strong><br>
Because you don’t have administrator rights, and because the system is shared by so many people, you can’t install and use software “the regular way”</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> For system-wide installed software, <strong>load it</strong> with <code>module</code> commands.</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> If something is not installed, ask OSC or use <em>Conda</em> or containers.</p></li>
</ul>
<p><br></p>
<ul>
<li class="fragment"><strong>Login versus compute nodes</strong> (node ≈ computer)<br>
“Login nodes”, the nodes you end up on after logging in, are not meant for heavy computing and you have to <em>request access to “compute nodes”</em> to run most analyses.</li>
</ul>
<p><br></p>
<ul>
<li class="fragment"><strong>“Non-interactive” usage is common, using a job scheduler (SLURM)</strong><br>
You submit your scripts to the SLURM queue and monitor the resulting jobs.</li>
</ul>
</div>
</section>
<section class="slide level2">

<h3 id="overview-2">Overview</h3>
<ol type="1">
<li><p>Introduction</p></li>
<li><p>Command-line genomics?</p></li>
<li><p>Ohio Supercomputer Center (OSC) overview</p></li>
<li><p><strong>Command-line software</strong></p></li>
<li><p>The VS Code editor and the whole game</p></li>
</ol>
</section>
<section class="slide level2">

<h3 id="an-example-of-a-command-line-program">An example of a command-line program</h3>
<p>A useful example of a genomics tool with a CLI is FastQC, a program for quality control of FASTQ files.</p>
<p>It is ubiquitous because nearly all high-throughput sequencing data comes in FASTQ files, and your first step is always to check the quality of the reads.</p>
<p>FastQC produces visualizations and assessments of aspects of your reads such as adapter content, and, as shown below, per-base quality:</p>
<div class="columns">
<div class="column" style="width:50%;">
<p><img src="img/fastqc_good.png" width="100%"></p>
</div><div class="column" style="width:50%;">
<p><img src="img/fastqc_bad.png" width="100%"></p>
</div>
</div>
</section>
<section class="slide level2">

<h3 id="running-fastqc">Running FastQC</h3>
<div>
<ul>
<li class="fragment"><p>To run FastQC, you use the command <code>fastqc</code>.</p>
<p>Command-line programs are typically run non-interactively, so we don’t fire up the program first, and tell it what to do later, like we would with a program with a GUI.</p>
<p>Instead, we at once issue a complete set of instructions for the program to do what we would like it to.</p></li>
</ul>
<p><br></p>
<ul>
<li class="fragment"><p>For example, say we want to analyze one of the FASTQ files that I put in <code>/fs/scratch/PAS2250/data</code>, with default FastQC settings. In that case, our complete FastQC command would be:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="ex">[jelmer@owens-login04</span> ~]$ fastqc /fs/scratch/PAS2250/data/sample1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>So, it is simply <code>fastqc</code> followed by a space and the name of the file!</p></li>
</ul>
</div>
</section>
<section class="slide level2">

<h3 id="loading-the-fastqc-module-at-osc">Loading the FastQC “module” at OSC</h3>
<div>
<ul>
<li class="fragment"><p>FastQC is available to us at OSC<sup>1</sup>, but we first have to <strong>load it</strong>. Here is what happens when we try to run the program in a fresh shell session at OSC:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="ex">[jelmer@owens-login04</span> ~]$ fastqc /fs/scratch/PAS2250/data/sample1.fastq.gz</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="co">#&gt; -bash: fastqc: command not found</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li class="fragment"><p>We can load the software as follows:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="ex">[jelmer@owens-login04</span> ~]$ module load fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li class="fragment"><p>Now, let’s try again:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="ex">[jelmer@owens-login04</span> ~]$ fastqc /fs/scratch/PAS2250/data/sample1.fastq.gz</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co">#&gt; Started analysis of sample1.fastq.gz</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">#&gt; Approx 5% complete for sample1.fastq.gz</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">#&gt; Approx 10% complete for sample1.fastq.gz</span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co">#&gt; Approx 15% complete for sample1.fastq.gz</span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">#&gt; [truncated]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Success!</p></li>
</ul>
</div>
<aside><ol class="aside-footnotes"><li id="fn3"><p>full list of installed software: <a href="https://www.osc.edu/resources/available_software/software_list" class="uri">https://www.osc.edu/resources/available_software/software_list</a></p></li></ol></aside></section>
<section class="slide level2">

<h3 id="something-is-missing-here">Something is missing here</h3>
<p>I mentioned earlier that one benefit of running programs at the command-line is reproducibility – but how do we save the commands that we run?</p>
<ul>
<li><p>We need to not just save them, but to keep a detailed digital notebook that will enable us to redo our analysis.</p></li>
<li><p>We also need to wrap these commands in little scripts, so that we can run programs non-interactively and in parallel.</p></li>
</ul>
<p>For all of this, we will need a good <strong>text editor</strong>.</p>
</section>
<section class="slide level2">

<h3 id="overview-3">Overview</h3>
<ol type="1">
<li><p>Introduction</p></li>
<li><p>Command-line genomics?</p></li>
<li><p>Ohio Supercomputer Center (OSC) overview</p></li>
<li><p>Command-line software</p></li>
<li><p><strong>The VS Code editor and the whole game</strong></p></li>
</ol>
</section>
<section class="slide level2">

<h3 id="the-vs-code-text-editor">The VS Code text editor</h3>
<p>VS Code (in full, “Visual Studio Code”) is a great modern GUI-based text editor. <sup>1</sup><br>
We can use a version of this editor (often referred to as <em>Code Server</em>) directly in our browser via the OSC OnDemand website.</p>
<p>I do the vast majority of my bioinformatics work using VS Code at OSC. Because it also has an integrated terminal to access a Unix shell, this effectively combines 3 of the 4 mentioned aspects of the genomics computational environment:</p>
<ul>
<li>Supercomputer</li>
<li>Unix shell</li>
<li>Text editor</li>
</ul>
<p>That is, all except R, for which I prefer RStudio. <sup>2</sup></p>
<aside><ol class="aside-footnotes"><li id="fn4"><p>While there are also command-line text editors, these make the learning curve for command-line genomics even steeper.</p></li><li id="fn5"><p>Though VS Code does have quite good R functionality!</p></li></ol></aside></section>
<section class="slide level2">

<h3 id="starting-up-vs-code-in-ondemand">Starting up VS Code in OnDemand</h3>
<ol type="1">
<li>Log in to OSC’s OnDemand portal at <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>.</li>
<li>In the blue top bar, select <code>Interactive Apps</code> and then near the bottom of the dropdown menu, click <code>Code Server</code>.</li>
<li>In the form that appears on a new page:
<ul>
<li>Select project <code>PAS2250</code></li>
<li>No need to change “Number of hours” and “Working Directory”</li>
<li>Make sure the “Codeserver Version” is <code>4.8</code>.</li>
</ul></li>
<li>On the next page, once the top bar of the box has turned green and says <code>Runnning</code>, click <code>Connect to VS Code</code>.</li>
</ol>
</section>
<section class="slide level2">

<h3 id="starting-up-vs-code-in-ondemand-1">Starting up VS Code in OnDemand</h3>
<p align="center">
<img src="img/vscode_form.png" width="62%">
</p>
</section>
<section class="slide level2">

<h3 id="starting-up-vs-code-in-ondemand-2">Starting up VS Code in OnDemand</h3>
<p>Once the session is running, you can click “Connect to VS Code”:</p>
<p align="center">
<img src="img/vscode_card.png" width="80%">
</p>
</section>
<section class="slide level2">

<h3 id="vs-code-basics-side-bar-with-file-browser">VS Code Basics: Side bar with file browser</h3>
<p align="center">
<img src="img/vscode-welcome_ed.png" width="60%">
</p>
</section>
<section class="slide level2">

<h3 id="vs-code-basics-editor-pane">VS Code Basics: Editor pane</h3>
<p>The main part of the VS Code is the <em>editor pane</em>.</p>
<p>Whenever you open VS Code, a editor pane tab with a <code>Get Started</code> document is automatically opened.</p>
<p>We can use this <code>Get Started</code> document to open a new text file by clicking <code>New file</code> below <code>Start</code>, which opens as a second tab in the editor pane.</p>
<p align="center">
<img src="img/vscode_editor.png" width="80%">
</p>
</section>
<section class="slide level2">

<h3 id="vs-code-basics-terminal">VS Code Basics: Terminal</h3>
<p>You can open a terminal with a Unix shell by clicking &nbsp; <i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; =&gt; <code>Terminal</code> =&gt; <code>New Terminal</code>. The great thing with this setup is that we can keep notes and write shell scripts in the same window as our shell and a file browser!</p>
<p align="center">
<img src="img/vscode_terminal.png" width="60%">
</p>
</section>
<section class="slide level2">

<h3 id="the-whole-game-building-on-the-fastqc-example">The whole game: building on the FastQC example</h3>
<p>I’ve shown you the main pieces of the computational infrastructure for<br>
“<strong>command-line genomics</strong>”:</p>
<ul>
<li>A supercomputer</li>
<li>The Unix shell,</li>
<li>A text editor – which actually ties all three together</li>
</ul>
<p>We’ve seen a very basic example of loading and running a command-line tool at OSC.</p>
</section>
<section class="slide level2">

<h3 id="the-whole-game-building-on-the-fastqc-example-1">The whole game: building on the FastQC example</h3>
<p>The missing pieces for a fuller example of how such tools are run in the context of an actual genomics project are (if we stay with FastQC):</p>
<div>
<ul>
<li class="fragment">Putting the command to run FastQC in a <strong>“shell script”</strong>.<br>
<em>(This used the same language (Bash) as the commands you’d type interactively,</em> <em>so at its most basic this involves pasting those commands into a text file.)</em></li>
</ul>
<p><br></p>
<ul>
<li class="fragment"><strong>Submitting the script</strong> to the SLURM scheduler queue as a “batch job”.<br>
<em>(At its most basic, this involves putting <code>sbatch</code> in front of the script name.)</em></li>
</ul>
<p><br></p>
<ul>
<li class="fragment">To make use of the capabilities of the supercomputer and speeding up our analysis, we can <strong>submit multiple jobs <em>in parallel</em> using a loop</strong>.</li>
</ul>
</div>
</section>
<section class="slide level2">

<h3 id="so-what-computational-skills-should-i-learn">So what computational skills should I learn?</h3>
<div>
<ul>
<li class="fragment"><p><strong>The core skills</strong>:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> Unix shell basics – the commonly used commands</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i>Some shell <em>scripting</em> basics</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> SLURM basics to submit and manage your batch jobs</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> R for “downstream”, statistical and visualization tasks</p></li>
<li class="fragment"><p><strong>When you start doing genomics projects more often</strong>:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> Using <code>conda</code> or containers for software</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> Unix data tools (<code>grep</code>, <code>sed</code>, <code>awk</code>, etc)</p></li>
<li class="fragment"><p><strong>When you want to become proficient in applied bioinformatics</strong>:</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> Version control with <code>git</code></p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> More advanced: formal workflow/pipeline management tools (e.g.&nbsp;<code>Nextflow</code>)</p>
<p><i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> More advanced: Python (or advanced R) for custom data processing</p></li>
</ul>
</div>
</section>
<section class="slide level2">

<h3 id="resources-for-further-learning-osc">Resources for further learning: OSC</h3>
<ul>
<li>Tutorials and courses!
<ul>
<li><a href="https://khill42.github.io/OSC_IntroHPC/" class="uri">https://khill42.github.io/OSC_IntroHPC/</a></li>
<li><a href="https://osc.catalog.instructure.com/" class="uri">https://osc.catalog.instructure.com/</a></li>
</ul></li>
</ul>
<p><br></p>
<ul>
<li>OSC regularly has online introductory sessions, both overviews and more hands-on sessions – see the OSC Events page: <a href="https://www.osc.edu/events" class="uri">https://www.osc.edu/events</a>.<br>
They also have weekly office hours.</li>
</ul>
<p><br></p>
<ul>
<li>There is also some good introductory material at their <em>Getting Started</em> pages (<a href="https://www.osc.edu/resources/getting_started" class="uri">https://www.osc.edu/resources/getting_started</a>), as well as more specific <em>“HOWTO” pages</em> (<a href="https://www.osc.edu/resources/getting_started/howto" class="uri">https://www.osc.edu/resources/getting_started/howto</a>).</li>
</ul>
</section>
<section class="slide level2">

<h3 id="resources-for-further-learning-osu">Resources for further learning: OSU</h3>
<ul>
<li><p>OSU courses and workshops</p>
<ul>
<li><p>Jonathan Fresnedo Ramirez’s “Genome Analytics” course (<code>HCS 7004</code>)</p></li>
<li><p>Microbiome Informatics (<code>MICRBIO 8161</code>)</p></li>
<li><p>The online materials for the workshop “Command line basics for genomic analysis at OSC” that myself and Mike Sovic gave last August:<br>
<a href="https://mcic-osu.github.io/cl-workshop-22/" class="uri">https://mcic-osu.github.io/cl-workshop-22/</a></p></li>
<li><p>I have a course “Practical Computing Skills for Omics Data” (<code>PLNTPTH 5006</code>) that I am planning to teach in in Spring 2024. All materials for the 2021 version of this course (“Practical Computing Skills for Biologists”) are at: <a href="https://mcic-osu.github.io/pracs-sp21/" class="uri">https://mcic-osu.github.io/pracs-sp21/</a></p></li>
</ul></li>
</ul>
</section>
<section class="slide level2">

<h3 id="resources-for-further-learning-osu-1">Resources for further learning: OSU</h3>
<p>Some particularly useful books:</p>
<ul>
<li><p>The Linux Command Line (William Shotts, 2019)</p></li>
<li><p>Bioinformatics Data Skills (Vince Buffalo, 2015)</p></li>
<li><p>Computing Skills for Biologists: A Toolbox (Wilmes &amp; Allesino, 2019)</p></li>
<li><p>A Primer for Computational Biology (Shawn T. O’ Neil, 2019) <a href="https://open.oregonstate.education/computationalbiology/" class="uri">https://open.oregonstate.education/computationalbiology/</a></p></li>
</ul>
</section>
<section class="slide level2">

<h3 id="mcic-bioinformatics-assistance">MCIC bioinformatics assistance</h3>
</section>
<section class="slide level2">

<h3 id="any-questions">Any questions?</h3>
</section>
<section class="slide level2">

<h3 id="submitting-fastqc-batch-jobs">Submitting FastQC batch jobs</h3>
<p>Due to time constraints, I have skipped over the details of these steps. But an example shell script to run FastQC can be found at <code>/fs/scratch/PAS2250/scripts/fastqc.sh</code>, which contains the following code:</p>
<div class="columns">
<div class="column" style="width:60%;">
<div class="sourceCode" id="cb5"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="co">#SBATCH --account=PAS2250</span></span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co"># Load the software</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="ex">module</span> load fastqc</span>
<span id="cb5-6"><a href="#cb5-6"></a></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="co"># Bash strict settings</span></span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span>
<span id="cb5-9"><a href="#cb5-9"></a></span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="co"># Copy the placeholder variables</span></span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="va">input_file</span><span class="op">=</span><span class="va">$1</span></span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="va">output_dir</span><span class="op">=</span><span class="va">$2</span></span>
<span id="cb5-13"><a href="#cb5-13"></a></span>
<span id="cb5-14"><a href="#cb5-14"></a><span class="co"># Create the output dir if needed</span></span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="st">"</span><span class="va">$output_dir</span><span class="st">"</span></span>
<span id="cb5-16"><a href="#cb5-16"></a></span>
<span id="cb5-17"><a href="#cb5-17"></a><span class="co"># Run FastQC</span></span>
<span id="cb5-18"><a href="#cb5-18"></a><span class="ex">fastqc</span> <span class="at">--outdir</span><span class="op">=</span><span class="st">"</span><span class="va">$output_dir</span><span class="st">"</span> <span class="st">"</span><span class="va">$input_file</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column" style="width:40%;">
<ul>
<li><p>Indicate it’s a Bash script and tell SLURM which OSC account to use</p></li>
<li><p>Strict settings make the script stop on failure</p></li>
<li><p>The script takes “arguments”, which are stored as placeholder variables <code>$1</code> and <code>$2</code>. This allows us to run the script for different files</p></li>
</ul>
</div>
</div>
</section>
<section class="slide level2">

<h3 id="submitting-fastqc-batch-jobs-1">Submitting FastQC batch jobs</h3>
<p>Here is how I would submit that script as a batch job to analyze one FASTQ file:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="va">fastq_file</span><span class="op">=</span>/fs/scratch/PAS2250/data/sample1.fastq.gz</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="ex">sbatch</span> /fs/scratch/PAS2250/scripts/fastqc.sh <span class="st">"</span><span class="va">$fastq_file</span><span class="st">"</span> results_jelmer</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">#&gt; Submitted batch job 16323144</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br></p>
<p>And how you can loop over all FASTQ files to submit as many jobs in parallel as you have FASTQ files:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="cf">for</span> fastq_file <span class="kw">in</span> /fs/scratch/PAS2250/data/<span class="pp">*</span>.fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>    <span class="ex">sbatch</span> /fs/scratch/PAS2250/scripts/fastqc.sh <span class="st">"</span><span class="va">$fastq_file</span><span class="st">"</span> results_jelmer</span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="cf">done</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">#&gt; Submitted batch job 16323145</span></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co">#&gt; Submitted batch job 16323146</span></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">#&gt; Submitted batch job 16323147</span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co">#&gt; Submitted batch job 16323148</span></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="co">#&gt; Submitted batch job 16323149</span></span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="co">#&gt; Submitted batch job 16323110</span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="co">#&gt; Submitted batch job 16323111</span></span>
<span id="cb7-11"><a href="#cb7-11"></a><span class="co">#&gt; Submitted batch job 16323112</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="footer footer-default">

</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="presentation_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="presentation_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="presentation_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': true,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>