---
title: "Untitled"
format: revealjs
---

------

### The MCIC

- The MCIC has three main sections:

  - **Microscopy**  
    We have confocal, TEM, and SEM microscopes
  
  - **Wetlab genetics and genomics**    
    E.g. qPCR,
    HTS (high-throughput sequencing) library preparation,
    some in-house HTS with an Illumina MiSeq and an Oxford Nanopore MinION.

  - **Applied bioinformatics**  
    Providing assistance with the analysis of genomics datasets.
    This is my role -- you can contact me if you need help!
    (<poelstra.1@osu.edu>)

<br>

- The most common type of genomics projects (both wetlab and analysis)
  are _RNAseq_, _microbial metabarcoding_, and _whole-genome sequencing_.
  
------

### About me

- Since the summer of '20, 
  I have been working as a bioinformatician at the
  Molecular and Cellular Imaging Center (MCIC) in Selby Hall
  on the Wooster Campus.

- I am a biologist by training,
  with a background in evolutionary biology / evolutionary genomics.

- For my PhD, I worked on the genomics of coloration and speciation in two very
  closely related species of European crows, carrion and hooded crows.

<p align="center">
<img src=img/crows2.png width="70%">
</p>

## Other FastQC commands

It is also possible to run FastQC for multiple FASTQ files at once.
In that case, you simply keep adding filenames to the command:

```bash
fastqc sample1_R1.fastq.gz sample1_R2.fastq.gz
```

------

## Other FastQC commands

Finally, we could modify the _way_ in which FastQC runs.
For instance, we may not want its output files to end up in our
"working directory", i.e. the folder that we are currently in,
and instead want to send these to separate folder.

We can do that with the FastQC's `--outdir` option:

```bash
fastqc --outdir=results/fastqc sample1_R1.fastq.gz
```
